{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os import path\n",
    "import gtfparse\n",
    "import pyensembl\n",
    "pyensembl.EnsemblRelease(release=87)\n",
    "import gspread\n",
    "from  oauth2client.service_account  import ServiceAccountCredentials\n",
    "import gspread\n",
    "import sys\n",
    "if sys.version_info[0] < 3: \n",
    "    from StringIO import StringIO\n",
    "else:\n",
    "    from io import StringIO\n",
    "from fish_helpers import *"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.twilio.com/blog/2017/02/an-easy-way-to-read-and-write-to-a-google-spreadsheet-in-python.html Go to the Google APIs Console. Create a new project. Click Enable API. Search for and enable the Google Drive API. Create credentials for a Web Server to access Application Data. Name the service account and grant it a Project Role of Editor. Download the JSON file. Copy the JSON file to your code directory Open JSON and find \"client_email\" share google sheet with this email"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:oauth2client.client:Refreshing access_token\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "      <th>Description</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Isoform %</th>\n",
       "      <th>Length</th>\n",
       "      <th>Oligos</th>\n",
       "      <th>Status</th>\n",
       "      <th>Transcript_ID</th>\n",
       "      <th>Type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>100000000101000000010</td>\n",
       "      <td>v-abl Abelson murine leukemia viral oncogene 2...</td>\n",
       "      <td>4.66</td>\n",
       "      <td>Abl2</td>\n",
       "      <td>ENSMUSG00000026596</td>\n",
       "      <td>66.33%</td>\n",
       "      <td>10,151</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000166172</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>11000000000000101</td>\n",
       "      <td>atypical chemokine receptor 3</td>\n",
       "      <td>7.71</td>\n",
       "      <td>Ackr3</td>\n",
       "      <td>ENSMUSG00000044337</td>\n",
       "      <td>48.94%</td>\n",
       "      <td>3,065</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000065587</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td>1000000000000001011000</td>\n",
       "      <td>actin, alpha 2, smooth muscle, aorta</td>\n",
       "      <td>309.73</td>\n",
       "      <td>Acta2</td>\n",
       "      <td>ENSMUSG00000035783</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>2,035</td>\n",
       "      <td>87</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000039631</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td>1000010000011000000000</td>\n",
       "      <td>ARP3 actin-related protein 3</td>\n",
       "      <td>40.98</td>\n",
       "      <td>Actr3</td>\n",
       "      <td>ENSMUSG00000026341</td>\n",
       "      <td>57.58%</td>\n",
       "      <td>2,565</td>\n",
       "      <td>90</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000178474</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td>1000000000001100001000</td>\n",
       "      <td>a disintegrin and metallopeptidase domain 17</td>\n",
       "      <td>7.18</td>\n",
       "      <td>Adam17</td>\n",
       "      <td>ENSMUSG00000052593</td>\n",
       "      <td>39.51%</td>\n",
       "      <td>4,443</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000064536</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td></td>\n",
       "      <td>1010001000000000100</td>\n",
       "      <td>a disintegrin and metallopeptidase domain 8</td>\n",
       "      <td>0.28</td>\n",
       "      <td>Adam8</td>\n",
       "      <td>ENSMUSG00000025473</td>\n",
       "      <td>53.81%</td>\n",
       "      <td>3,117</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000106069</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td></td>\n",
       "      <td>10100110000000000000000</td>\n",
       "      <td>adrenergic receptor, beta 2</td>\n",
       "      <td>0.43</td>\n",
       "      <td>Adrb2</td>\n",
       "      <td>ENSMUSG00000045730</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>2,144</td>\n",
       "      <td>92</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000053640</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td></td>\n",
       "      <td>10011000010000000000000</td>\n",
       "      <td>angiotensinogen (serpin peptidase inhibitor, c...</td>\n",
       "      <td>0.56</td>\n",
       "      <td>Agt</td>\n",
       "      <td>ENSMUSG00000031980</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>1,891</td>\n",
       "      <td>93</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000063278</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td></td>\n",
       "      <td>10001100100000000000000</td>\n",
       "      <td>angiotensin II receptor, type 2</td>\n",
       "      <td>0.24</td>\n",
       "      <td>Agtr2</td>\n",
       "      <td>ENSMUSG00000068122</td>\n",
       "      <td>22.01%</td>\n",
       "      <td>3,283</td>\n",
       "      <td>104</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000089188</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td></td>\n",
       "      <td>10000010000010010</td>\n",
       "      <td>ajuba LIM protein</td>\n",
       "      <td>8.16</td>\n",
       "      <td>Ajuba</td>\n",
       "      <td>ENSMUSG00000022178</td>\n",
       "      <td>97.51%</td>\n",
       "      <td>3,397</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000054487</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td></td>\n",
       "      <td>10000000101000000100</td>\n",
       "      <td>A kinase (PRKA) anchor protein (gravin) 12</td>\n",
       "      <td>4.57</td>\n",
       "      <td>Akap12</td>\n",
       "      <td>ENSMUSG00000038587</td>\n",
       "      <td>68.47%</td>\n",
       "      <td>6,195</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000045730</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td></td>\n",
       "      <td>100000001100000100000000</td>\n",
       "      <td>thymoma viral proto-oncogene 1</td>\n",
       "      <td>68.19</td>\n",
       "      <td>Akt1</td>\n",
       "      <td>ENSMUSG00000001729</td>\n",
       "      <td>30.86%</td>\n",
       "      <td>2,690</td>\n",
       "      <td>115</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000001780</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>S</td>\n",
       "      <td>110001000010</td>\n",
       "      <td>aldehyde dehydrogenase 9, subfamily A1 [Source...</td>\n",
       "      <td>13.50</td>\n",
       "      <td>Aldh9a1</td>\n",
       "      <td>ENSMUSG00000026687</td>\n",
       "      <td>82.94%</td>\n",
       "      <td>2,835</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000028004</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td></td>\n",
       "      <td>1000001011</td>\n",
       "      <td>angiopoietin 1</td>\n",
       "      <td>2.49</td>\n",
       "      <td>Angpt1</td>\n",
       "      <td>ENSMUSG00000022309</td>\n",
       "      <td>50.00%</td>\n",
       "      <td>4,260</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000022921</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td></td>\n",
       "      <td>10000010100000000000010</td>\n",
       "      <td>angiopoietin 2</td>\n",
       "      <td>0.20</td>\n",
       "      <td>Angpt2</td>\n",
       "      <td>ENSMUSG00000031465</td>\n",
       "      <td>23.78%</td>\n",
       "      <td>3,560</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000033846</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td></td>\n",
       "      <td>10000001000000000100001</td>\n",
       "      <td>angiopoietin 4</td>\n",
       "      <td>0.17</td>\n",
       "      <td>Angpt4</td>\n",
       "      <td>ENSMUSG00000027460</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>2,335</td>\n",
       "      <td>113</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000028955</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td></td>\n",
       "      <td>10000000000000010100100</td>\n",
       "      <td>ankyrin repeat domain 26</td>\n",
       "      <td>1.23</td>\n",
       "      <td>Ankrd26</td>\n",
       "      <td>ENSMUSG00000007827</td>\n",
       "      <td>63.97%</td>\n",
       "      <td>6,930</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000112830</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td></td>\n",
       "      <td>10110100000</td>\n",
       "      <td>annexin A2</td>\n",
       "      <td>242.85</td>\n",
       "      <td>Anxa2</td>\n",
       "      <td>ENSMUSG00000032231</td>\n",
       "      <td>87.69%</td>\n",
       "      <td>1,447</td>\n",
       "      <td>81</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000034756</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td></td>\n",
       "      <td>110000010100000</td>\n",
       "      <td>annexin A9</td>\n",
       "      <td>0.37</td>\n",
       "      <td>Anxa9</td>\n",
       "      <td>ENSMUSG00000015702</td>\n",
       "      <td>92.25%</td>\n",
       "      <td>1,851</td>\n",
       "      <td>100</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000015846</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td></td>\n",
       "      <td>100000001001000000000001</td>\n",
       "      <td>apoptosis inhibitor 5</td>\n",
       "      <td>17.72</td>\n",
       "      <td>Api5</td>\n",
       "      <td>ENSMUSG00000027193</td>\n",
       "      <td>49.95%</td>\n",
       "      <td>3,736</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000028617</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td></td>\n",
       "      <td>1001000000101000</td>\n",
       "      <td>apelin</td>\n",
       "      <td>4.38</td>\n",
       "      <td>Apln</td>\n",
       "      <td>ENSMUSG00000037010</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>3,149</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000039026</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td></td>\n",
       "      <td>1000000000001100010</td>\n",
       "      <td>adaptor protein, phosphotyrosine interaction, ...</td>\n",
       "      <td>3.01</td>\n",
       "      <td>Appl1</td>\n",
       "      <td>ENSMUSG00000040760</td>\n",
       "      <td>71.91%</td>\n",
       "      <td>7,642</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000036570</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td></td>\n",
       "      <td>100000010000000100000100</td>\n",
       "      <td>adaptor protein, phosphotyrosine interaction, ...</td>\n",
       "      <td>4.51</td>\n",
       "      <td>Appl2</td>\n",
       "      <td>ENSMUSG00000020263</td>\n",
       "      <td>46.39%</td>\n",
       "      <td>3,020</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000020500</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td></td>\n",
       "      <td>10010000000010000001000</td>\n",
       "      <td>aquaporin 1</td>\n",
       "      <td>31.66</td>\n",
       "      <td>Aqp1</td>\n",
       "      <td>ENSMUSG00000004655</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>2,627</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000004774</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td></td>\n",
       "      <td>10001000011000000</td>\n",
       "      <td>Araf proto-oncogene, serine/threonine kinase</td>\n",
       "      <td>6.19</td>\n",
       "      <td>Araf</td>\n",
       "      <td>ENSMUSG00000001127</td>\n",
       "      <td>36.42%</td>\n",
       "      <td>2,358</td>\n",
       "      <td>95</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000001155</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>G2M</td>\n",
       "      <td>100001000000000000110</td>\n",
       "      <td>Rho GTPase activating protein 19 [Source:MGI S...</td>\n",
       "      <td>1.70</td>\n",
       "      <td>Arhgap19</td>\n",
       "      <td>ENSMUSG00000025154</td>\n",
       "      <td>95.55%</td>\n",
       "      <td>5,103</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000026150</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td></td>\n",
       "      <td>1000000010100000100000</td>\n",
       "      <td>Rho guanine nucleotide exchange factor (GEF) 5</td>\n",
       "      <td>1.37</td>\n",
       "      <td>Arhgef5</td>\n",
       "      <td>ENSMUSG00000033542</td>\n",
       "      <td>58.86%</td>\n",
       "      <td>5,469</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000031750</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td></td>\n",
       "      <td>1010000100010</td>\n",
       "      <td>AT rich interactive domain 5B (MRF1-like)</td>\n",
       "      <td>5.79</td>\n",
       "      <td>Arid5b</td>\n",
       "      <td>ENSMUSG00000019947</td>\n",
       "      <td>85.89%</td>\n",
       "      <td>7,581</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000219238</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td></td>\n",
       "      <td>100100100000010000</td>\n",
       "      <td>arrestin, beta 1</td>\n",
       "      <td>0.91</td>\n",
       "      <td>Arrb1</td>\n",
       "      <td>ENSMUSG00000018909</td>\n",
       "      <td>57.52%</td>\n",
       "      <td>7,132</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000098266</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td></td>\n",
       "      <td>11000000000000000100010</td>\n",
       "      <td>ASH1 like histone lysine methyltransferase</td>\n",
       "      <td>3.03</td>\n",
       "      <td>Ash1l</td>\n",
       "      <td>ENSMUSG00000028053</td>\n",
       "      <td>66.12%</td>\n",
       "      <td>11,268</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000090933</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>442</th>\n",
       "      <td></td>\n",
       "      <td>11000001001000</td>\n",
       "      <td>TAO kinase 3</td>\n",
       "      <td>3.33</td>\n",
       "      <td>Taok3</td>\n",
       "      <td>ENSMUSG00000061288</td>\n",
       "      <td>55.39%</td>\n",
       "      <td>4,467</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000111978</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>443</th>\n",
       "      <td>G1</td>\n",
       "      <td>1000010000000110000000</td>\n",
       "      <td>t-complex 11 (mouse) like 2 [Source:MGI Symbol...</td>\n",
       "      <td>3.00</td>\n",
       "      <td>Tcp11l2</td>\n",
       "      <td>ENSMUSG00000020034</td>\n",
       "      <td>99.41%</td>\n",
       "      <td>2,725</td>\n",
       "      <td>118</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000020223</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>444</th>\n",
       "      <td></td>\n",
       "      <td>1000000000001001100000</td>\n",
       "      <td>teratocarcinoma-derived growth factor 1</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Tdgf1</td>\n",
       "      <td>ENSMUSG00000032494</td>\n",
       "      <td>84.69%</td>\n",
       "      <td>1,932</td>\n",
       "      <td>87</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000035075</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>445</th>\n",
       "      <td></td>\n",
       "      <td>1000011001000000000</td>\n",
       "      <td>transforming growth factor, beta 2</td>\n",
       "      <td>0.66</td>\n",
       "      <td>Tgfb2</td>\n",
       "      <td>ENSMUSG00000039239</td>\n",
       "      <td>54.92%</td>\n",
       "      <td>5,116</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000045288</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>446</th>\n",
       "      <td></td>\n",
       "      <td>100100010100</td>\n",
       "      <td>transforming growth factor, beta 3</td>\n",
       "      <td>13.58</td>\n",
       "      <td>Tgfb3</td>\n",
       "      <td>ENSMUSG00000021253</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>3,383</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000003687</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>447</th>\n",
       "      <td>Receptor</td>\n",
       "      <td>1001000000010000000100</td>\n",
       "      <td>transforming growth factor, beta receptor I [S...</td>\n",
       "      <td>5.30</td>\n",
       "      <td>Tgfbr1</td>\n",
       "      <td>ENSMUSG00000007613</td>\n",
       "      <td>90.81%</td>\n",
       "      <td>5756</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000007757</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>448</th>\n",
       "      <td>Receptor</td>\n",
       "      <td>1000011000100000</td>\n",
       "      <td>transforming growth factor, beta receptor II [...</td>\n",
       "      <td>2.85</td>\n",
       "      <td>Tgfbr2</td>\n",
       "      <td>ENSMUSG00000032440</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>3476</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000035014</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>449</th>\n",
       "      <td></td>\n",
       "      <td>1000100000100000000010</td>\n",
       "      <td>transforming growth factor, beta receptor III</td>\n",
       "      <td>12.40</td>\n",
       "      <td>Tgfbr3</td>\n",
       "      <td>ENSMUSG00000029287</td>\n",
       "      <td>94.72%</td>\n",
       "      <td>6,087</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000031224</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>450</th>\n",
       "      <td></td>\n",
       "      <td>100000110100000000000000</td>\n",
       "      <td>thrombospondin 1</td>\n",
       "      <td>75.02</td>\n",
       "      <td>Thbs1</td>\n",
       "      <td>ENSMUSG00000040152</td>\n",
       "      <td>49.89%</td>\n",
       "      <td>5,848</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000039559</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>451</th>\n",
       "      <td>S</td>\n",
       "      <td>10000000101000000100000</td>\n",
       "      <td>transmembrane protein 109 [Source:MGI Symbol;A...</td>\n",
       "      <td>29.73</td>\n",
       "      <td>Tmem109</td>\n",
       "      <td>ENSMUSG00000034659</td>\n",
       "      <td>62.64%</td>\n",
       "      <td>2,188</td>\n",
       "      <td>115</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000038128</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>452</th>\n",
       "      <td></td>\n",
       "      <td>1000000010001000010000</td>\n",
       "      <td>transmembrane protein 201</td>\n",
       "      <td>2.77</td>\n",
       "      <td>Tmem201</td>\n",
       "      <td>ENSMUSG00000044700</td>\n",
       "      <td>35.04%</td>\n",
       "      <td>3,620</td>\n",
       "      <td>115</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000103208</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>453</th>\n",
       "      <td></td>\n",
       "      <td>10000100000000110</td>\n",
       "      <td>tumor necrosis factor receptor superfamily, me...</td>\n",
       "      <td>39.15</td>\n",
       "      <td>Tnfrsf1a</td>\n",
       "      <td>ENSMUSG00000030341</td>\n",
       "      <td>58.23%</td>\n",
       "      <td>2,156</td>\n",
       "      <td>100</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000032491</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>454</th>\n",
       "      <td></td>\n",
       "      <td>1100010000000000010</td>\n",
       "      <td>TNF receptor-associated factor 6</td>\n",
       "      <td>1.81</td>\n",
       "      <td>Traf6</td>\n",
       "      <td>ENSMUSG00000027164</td>\n",
       "      <td>48.03%</td>\n",
       "      <td>6,169</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000004949</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>455</th>\n",
       "      <td></td>\n",
       "      <td>100100001000010000000</td>\n",
       "      <td>tripartite motif-containing 32</td>\n",
       "      <td>8.94</td>\n",
       "      <td>Trim32</td>\n",
       "      <td>ENSMUSG00000051675</td>\n",
       "      <td>51.47%</td>\n",
       "      <td>3,194</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000050850</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>456</th>\n",
       "      <td></td>\n",
       "      <td>10000001000110000</td>\n",
       "      <td>tripartite motif-containing 71</td>\n",
       "      <td>0.02</td>\n",
       "      <td>Trim71</td>\n",
       "      <td>ENSMUSG00000079259</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>9,332</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000111816</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>457</th>\n",
       "      <td></td>\n",
       "      <td>10000011001</td>\n",
       "      <td>transformation related protein 53</td>\n",
       "      <td>45.30</td>\n",
       "      <td>Trp53</td>\n",
       "      <td>ENSMUSG00000059552</td>\n",
       "      <td>76.56%</td>\n",
       "      <td>1,771</td>\n",
       "      <td>91</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000108658</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>458</th>\n",
       "      <td></td>\n",
       "      <td>10001000000000000001010</td>\n",
       "      <td>transformation related protein 53 inducible nu...</td>\n",
       "      <td>1.42</td>\n",
       "      <td>Trp53inp1</td>\n",
       "      <td>ENSMUSG00000028211</td>\n",
       "      <td>40.17%</td>\n",
       "      <td>5,404</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000029865</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>459</th>\n",
       "      <td></td>\n",
       "      <td>100101010000000000</td>\n",
       "      <td>TSC complex subunit 2</td>\n",
       "      <td>0.14</td>\n",
       "      <td>Tsc2</td>\n",
       "      <td>ENSMUSG00000002496</td>\n",
       "      <td>0.41%</td>\n",
       "      <td>6,018</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000097373</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>460</th>\n",
       "      <td></td>\n",
       "      <td>11000000000001010</td>\n",
       "      <td>vascular endothelial growth factor C</td>\n",
       "      <td>0.89</td>\n",
       "      <td>Vegfc</td>\n",
       "      <td>ENSMUSG00000031520</td>\n",
       "      <td>68.08%</td>\n",
       "      <td>2,449</td>\n",
       "      <td>80</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000033919</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>461</th>\n",
       "      <td></td>\n",
       "      <td>100000010000000000101</td>\n",
       "      <td>vascular endothelial growth factor D</td>\n",
       "      <td>13.15</td>\n",
       "      <td>Vegfd</td>\n",
       "      <td>ENSMUSG00000031380</td>\n",
       "      <td>48.95%</td>\n",
       "      <td>2,049</td>\n",
       "      <td>94</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000033751</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>462</th>\n",
       "      <td></td>\n",
       "      <td>100010000000011</td>\n",
       "      <td>WAPL cohesin release factor</td>\n",
       "      <td>3.97</td>\n",
       "      <td>Wapl</td>\n",
       "      <td>ENSMUSG00000041408</td>\n",
       "      <td>50.80%</td>\n",
       "      <td>6,351</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000048263</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>463</th>\n",
       "      <td></td>\n",
       "      <td>10000000001000101000000</td>\n",
       "      <td>WD repeat containing planar cell polarity effe...</td>\n",
       "      <td>0.96</td>\n",
       "      <td>Wdpcp</td>\n",
       "      <td>ENSMUSG00000020319</td>\n",
       "      <td>51.90%</td>\n",
       "      <td>2,792</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000020568</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>464</th>\n",
       "      <td>S</td>\n",
       "      <td>110100001000</td>\n",
       "      <td>WD repeat domain 18 [Source:MGI Symbol;Acc:MGI...</td>\n",
       "      <td>6.24</td>\n",
       "      <td>Wdr18</td>\n",
       "      <td>ENSMUSG00000035754</td>\n",
       "      <td>51.99%</td>\n",
       "      <td>3,911</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000045247</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>465</th>\n",
       "      <td></td>\n",
       "      <td>100000000100100001000000</td>\n",
       "      <td>wingless-type MMTV integration site family, me...</td>\n",
       "      <td>0.01</td>\n",
       "      <td>Wnt1</td>\n",
       "      <td>ENSMUSG00000022997</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>2,378</td>\n",
       "      <td>86</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000023734</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>466</th>\n",
       "      <td></td>\n",
       "      <td>10000101001000000</td>\n",
       "      <td>wingless-type MMTV integration site family, me...</td>\n",
       "      <td>0.03</td>\n",
       "      <td>Wnt2</td>\n",
       "      <td>ENSMUSG00000010797</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>2,255</td>\n",
       "      <td>99</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000010941</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>467</th>\n",
       "      <td></td>\n",
       "      <td>100000000001001010</td>\n",
       "      <td>wingless-type MMTV integration site family, me...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>Wnt4</td>\n",
       "      <td>ENSMUSG00000036856</td>\n",
       "      <td>100.00%</td>\n",
       "      <td>4,194</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000045747</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>468</th>\n",
       "      <td></td>\n",
       "      <td>1001100000010</td>\n",
       "      <td>X-ray repair complementing defective repair in...</td>\n",
       "      <td>1.83</td>\n",
       "      <td>Xrcc2</td>\n",
       "      <td>ENSMUSG00000028933</td>\n",
       "      <td>96.41%</td>\n",
       "      <td>3,247</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000030773</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>469</th>\n",
       "      <td>G1</td>\n",
       "      <td>1000001000001000000010</td>\n",
       "      <td>zinc fingers and homeoboxes 1 [Source:MGI Symb...</td>\n",
       "      <td>3.49</td>\n",
       "      <td>Zhx1</td>\n",
       "      <td>ENSMUSG00000022361</td>\n",
       "      <td>74.82%</td>\n",
       "      <td>4,764</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000070143</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>470</th>\n",
       "      <td></td>\n",
       "      <td>11001000100000000</td>\n",
       "      <td>zinc finger, MIZ-type containing 1</td>\n",
       "      <td>11.53</td>\n",
       "      <td>Zmiz1</td>\n",
       "      <td>ENSMUSG00000007817</td>\n",
       "      <td>81.60%</td>\n",
       "      <td>7,482</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000162645</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>471</th>\n",
       "      <td></td>\n",
       "      <td>11100000000010000000</td>\n",
       "      <td>zinc metallopeptidase, STE24</td>\n",
       "      <td>11.33</td>\n",
       "      <td>Zmpste24</td>\n",
       "      <td>ENSMUSG00000043207</td>\n",
       "      <td>87.90%</td>\n",
       "      <td>3,429</td>\n",
       "      <td>120</td>\n",
       "      <td>Passed</td>\n",
       "      <td>ENSMUST00000058754</td>\n",
       "      <td>Passed</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>472 rows × 12 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                Barcode  \\\n",
       "0                 100000000101000000010   \n",
       "1                     11000000000000101   \n",
       "2                1000000000000001011000   \n",
       "3                1000010000011000000000   \n",
       "4                1000000000001100001000   \n",
       "5                   1010001000000000100   \n",
       "6               10100110000000000000000   \n",
       "7               10011000010000000000000   \n",
       "8               10001100100000000000000   \n",
       "9                     10000010000010010   \n",
       "10                 10000000101000000100   \n",
       "11             100000001100000100000000   \n",
       "12          S              110001000010   \n",
       "13                           1000001011   \n",
       "14              10000010100000000000010   \n",
       "15              10000001000000000100001   \n",
       "16              10000000000000010100100   \n",
       "17                          10110100000   \n",
       "18                      110000010100000   \n",
       "19             100000001001000000000001   \n",
       "20                     1001000000101000   \n",
       "21                  1000000000001100010   \n",
       "22             100000010000000100000100   \n",
       "23              10010000000010000001000   \n",
       "24                    10001000011000000   \n",
       "25        G2M     100001000000000000110   \n",
       "26               1000000010100000100000   \n",
       "27                        1010000100010   \n",
       "28                   100100100000010000   \n",
       "29              11000000000000000100010   \n",
       "..        ...                       ...   \n",
       "442                      11000001001000   \n",
       "443        G1    1000010000000110000000   \n",
       "444              1000000000001001100000   \n",
       "445                 1000011001000000000   \n",
       "446                        100100010100   \n",
       "447  Receptor    1001000000010000000100   \n",
       "448  Receptor          1000011000100000   \n",
       "449              1000100000100000000010   \n",
       "450            100000110100000000000000   \n",
       "451         S   10000000101000000100000   \n",
       "452              1000000010001000010000   \n",
       "453                   10000100000000110   \n",
       "454                 1100010000000000010   \n",
       "455               100100001000010000000   \n",
       "456                   10000001000110000   \n",
       "457                         10000011001   \n",
       "458             10001000000000000001010   \n",
       "459                  100101010000000000   \n",
       "460                   11000000000001010   \n",
       "461               100000010000000000101   \n",
       "462                     100010000000011   \n",
       "463             10000000001000101000000   \n",
       "464         S              110100001000   \n",
       "465            100000000100100001000000   \n",
       "466                   10000101001000000   \n",
       "467                  100000000001001010   \n",
       "468                       1001100000010   \n",
       "469        G1    1000001000001000000010   \n",
       "470                   11001000100000000   \n",
       "471                11100000000010000000   \n",
       "\n",
       "                                           Description    FPKM       Gene  \\\n",
       "0    v-abl Abelson murine leukemia viral oncogene 2...    4.66       Abl2   \n",
       "1                        atypical chemokine receptor 3    7.71      Ackr3   \n",
       "2                 actin, alpha 2, smooth muscle, aorta  309.73      Acta2   \n",
       "3                         ARP3 actin-related protein 3   40.98      Actr3   \n",
       "4         a disintegrin and metallopeptidase domain 17    7.18     Adam17   \n",
       "5          a disintegrin and metallopeptidase domain 8    0.28      Adam8   \n",
       "6                          adrenergic receptor, beta 2    0.43      Adrb2   \n",
       "7    angiotensinogen (serpin peptidase inhibitor, c...    0.56        Agt   \n",
       "8                      angiotensin II receptor, type 2    0.24      Agtr2   \n",
       "9                                    ajuba LIM protein    8.16      Ajuba   \n",
       "10          A kinase (PRKA) anchor protein (gravin) 12    4.57     Akap12   \n",
       "11                      thymoma viral proto-oncogene 1   68.19       Akt1   \n",
       "12   aldehyde dehydrogenase 9, subfamily A1 [Source...   13.50    Aldh9a1   \n",
       "13                                      angiopoietin 1    2.49     Angpt1   \n",
       "14                                      angiopoietin 2    0.20     Angpt2   \n",
       "15                                      angiopoietin 4    0.17     Angpt4   \n",
       "16                            ankyrin repeat domain 26    1.23    Ankrd26   \n",
       "17                                          annexin A2  242.85      Anxa2   \n",
       "18                                          annexin A9    0.37      Anxa9   \n",
       "19                               apoptosis inhibitor 5   17.72       Api5   \n",
       "20                                              apelin    4.38       Apln   \n",
       "21   adaptor protein, phosphotyrosine interaction, ...    3.01      Appl1   \n",
       "22   adaptor protein, phosphotyrosine interaction, ...    4.51      Appl2   \n",
       "23                                         aquaporin 1   31.66       Aqp1   \n",
       "24        Araf proto-oncogene, serine/threonine kinase    6.19       Araf   \n",
       "25   Rho GTPase activating protein 19 [Source:MGI S...    1.70   Arhgap19   \n",
       "26      Rho guanine nucleotide exchange factor (GEF) 5    1.37    Arhgef5   \n",
       "27           AT rich interactive domain 5B (MRF1-like)    5.79     Arid5b   \n",
       "28                                    arrestin, beta 1    0.91      Arrb1   \n",
       "29          ASH1 like histone lysine methyltransferase    3.03      Ash1l   \n",
       "..                                                 ...     ...        ...   \n",
       "442                                       TAO kinase 3    3.33      Taok3   \n",
       "443  t-complex 11 (mouse) like 2 [Source:MGI Symbol...    3.00    Tcp11l2   \n",
       "444            teratocarcinoma-derived growth factor 1    0.01      Tdgf1   \n",
       "445                 transforming growth factor, beta 2    0.66      Tgfb2   \n",
       "446                 transforming growth factor, beta 3   13.58      Tgfb3   \n",
       "447  transforming growth factor, beta receptor I [S...    5.30     Tgfbr1   \n",
       "448  transforming growth factor, beta receptor II [...    2.85     Tgfbr2   \n",
       "449      transforming growth factor, beta receptor III   12.40     Tgfbr3   \n",
       "450                                   thrombospondin 1   75.02      Thbs1   \n",
       "451  transmembrane protein 109 [Source:MGI Symbol;A...   29.73    Tmem109   \n",
       "452                          transmembrane protein 201    2.77    Tmem201   \n",
       "453  tumor necrosis factor receptor superfamily, me...   39.15   Tnfrsf1a   \n",
       "454                   TNF receptor-associated factor 6    1.81      Traf6   \n",
       "455                     tripartite motif-containing 32    8.94     Trim32   \n",
       "456                     tripartite motif-containing 71    0.02     Trim71   \n",
       "457                  transformation related protein 53   45.30      Trp53   \n",
       "458  transformation related protein 53 inducible nu...    1.42  Trp53inp1   \n",
       "459                              TSC complex subunit 2    0.14       Tsc2   \n",
       "460               vascular endothelial growth factor C    0.89      Vegfc   \n",
       "461               vascular endothelial growth factor D   13.15      Vegfd   \n",
       "462                        WAPL cohesin release factor    3.97       Wapl   \n",
       "463  WD repeat containing planar cell polarity effe...    0.96      Wdpcp   \n",
       "464  WD repeat domain 18 [Source:MGI Symbol;Acc:MGI...    6.24      Wdr18   \n",
       "465  wingless-type MMTV integration site family, me...    0.01       Wnt1   \n",
       "466  wingless-type MMTV integration site family, me...    0.03       Wnt2   \n",
       "467  wingless-type MMTV integration site family, me...    0.00       Wnt4   \n",
       "468  X-ray repair complementing defective repair in...    1.83      Xrcc2   \n",
       "469  zinc fingers and homeoboxes 1 [Source:MGI Symb...    3.49       Zhx1   \n",
       "470                 zinc finger, MIZ-type containing 1   11.53      Zmiz1   \n",
       "471                       zinc metallopeptidase, STE24   11.33   Zmpste24   \n",
       "\n",
       "                Gene_ID Isoform %  Length  Oligos  Status       Transcript_ID  \\\n",
       "0    ENSMUSG00000026596    66.33%  10,151     120  Passed  ENSMUST00000166172   \n",
       "1    ENSMUSG00000044337    48.94%   3,065     120  Passed  ENSMUST00000065587   \n",
       "2    ENSMUSG00000035783   100.00%   2,035      87  Passed  ENSMUST00000039631   \n",
       "3    ENSMUSG00000026341    57.58%   2,565      90  Passed  ENSMUST00000178474   \n",
       "4    ENSMUSG00000052593    39.51%   4,443     120  Passed  ENSMUST00000064536   \n",
       "5    ENSMUSG00000025473    53.81%   3,117     120  Passed  ENSMUST00000106069   \n",
       "6    ENSMUSG00000045730   100.00%   2,144      92  Passed  ENSMUST00000053640   \n",
       "7    ENSMUSG00000031980   100.00%   1,891      93  Passed  ENSMUST00000063278   \n",
       "8    ENSMUSG00000068122    22.01%   3,283     104  Passed  ENSMUST00000089188   \n",
       "9    ENSMUSG00000022178    97.51%   3,397     120  Passed  ENSMUST00000054487   \n",
       "10   ENSMUSG00000038587    68.47%   6,195     120  Passed  ENSMUST00000045730   \n",
       "11   ENSMUSG00000001729    30.86%   2,690     115  Passed  ENSMUST00000001780   \n",
       "12   ENSMUSG00000026687    82.94%   2,835     120  Passed  ENSMUST00000028004   \n",
       "13   ENSMUSG00000022309    50.00%   4,260     120  Passed  ENSMUST00000022921   \n",
       "14   ENSMUSG00000031465    23.78%   3,560     120  Passed  ENSMUST00000033846   \n",
       "15   ENSMUSG00000027460   100.00%   2,335     113  Passed  ENSMUST00000028955   \n",
       "16   ENSMUSG00000007827    63.97%   6,930     120  Passed  ENSMUST00000112830   \n",
       "17   ENSMUSG00000032231    87.69%   1,447      81  Passed  ENSMUST00000034756   \n",
       "18   ENSMUSG00000015702    92.25%   1,851     100  Passed  ENSMUST00000015846   \n",
       "19   ENSMUSG00000027193    49.95%   3,736     120  Passed  ENSMUST00000028617   \n",
       "20   ENSMUSG00000037010   100.00%   3,149     120  Passed  ENSMUST00000039026   \n",
       "21   ENSMUSG00000040760    71.91%   7,642     120  Passed  ENSMUST00000036570   \n",
       "22   ENSMUSG00000020263    46.39%   3,020     120  Passed  ENSMUST00000020500   \n",
       "23   ENSMUSG00000004655   100.00%   2,627     120  Passed  ENSMUST00000004774   \n",
       "24   ENSMUSG00000001127    36.42%   2,358      95  Passed  ENSMUST00000001155   \n",
       "25   ENSMUSG00000025154    95.55%   5,103     120  Passed  ENSMUST00000026150   \n",
       "26   ENSMUSG00000033542    58.86%   5,469     120  Passed  ENSMUST00000031750   \n",
       "27   ENSMUSG00000019947    85.89%   7,581     120  Passed  ENSMUST00000219238   \n",
       "28   ENSMUSG00000018909    57.52%   7,132     120  Passed  ENSMUST00000098266   \n",
       "29   ENSMUSG00000028053    66.12%  11,268     120  Passed  ENSMUST00000090933   \n",
       "..                  ...       ...     ...     ...     ...                 ...   \n",
       "442  ENSMUSG00000061288    55.39%   4,467     120  Passed  ENSMUST00000111978   \n",
       "443  ENSMUSG00000020034    99.41%   2,725     118  Passed  ENSMUST00000020223   \n",
       "444  ENSMUSG00000032494    84.69%   1,932      87  Passed  ENSMUST00000035075   \n",
       "445  ENSMUSG00000039239    54.92%   5,116     120  Passed  ENSMUST00000045288   \n",
       "446  ENSMUSG00000021253   100.00%   3,383     120  Passed  ENSMUST00000003687   \n",
       "447  ENSMUSG00000007613    90.81%    5756     120  Passed  ENSMUST00000007757   \n",
       "448  ENSMUSG00000032440   100.00%    3476     120  Passed  ENSMUST00000035014   \n",
       "449  ENSMUSG00000029287    94.72%   6,087     120  Passed  ENSMUST00000031224   \n",
       "450  ENSMUSG00000040152    49.89%   5,848     120  Passed  ENSMUST00000039559   \n",
       "451  ENSMUSG00000034659    62.64%   2,188     115  Passed  ENSMUST00000038128   \n",
       "452  ENSMUSG00000044700    35.04%   3,620     115  Passed  ENSMUST00000103208   \n",
       "453  ENSMUSG00000030341    58.23%   2,156     100  Passed  ENSMUST00000032491   \n",
       "454  ENSMUSG00000027164    48.03%   6,169     120  Passed  ENSMUST00000004949   \n",
       "455  ENSMUSG00000051675    51.47%   3,194     120  Passed  ENSMUST00000050850   \n",
       "456  ENSMUSG00000079259   100.00%   9,332     120  Passed  ENSMUST00000111816   \n",
       "457  ENSMUSG00000059552    76.56%   1,771      91  Passed  ENSMUST00000108658   \n",
       "458  ENSMUSG00000028211    40.17%   5,404     120  Passed  ENSMUST00000029865   \n",
       "459  ENSMUSG00000002496     0.41%   6,018     120  Passed  ENSMUST00000097373   \n",
       "460  ENSMUSG00000031520    68.08%   2,449      80  Passed  ENSMUST00000033919   \n",
       "461  ENSMUSG00000031380    48.95%   2,049      94  Passed  ENSMUST00000033751   \n",
       "462  ENSMUSG00000041408    50.80%   6,351     120  Passed  ENSMUST00000048263   \n",
       "463  ENSMUSG00000020319    51.90%   2,792     120  Passed  ENSMUST00000020568   \n",
       "464  ENSMUSG00000035754    51.99%   3,911     120  Passed  ENSMUST00000045247   \n",
       "465  ENSMUSG00000022997   100.00%   2,378      86  Passed  ENSMUST00000023734   \n",
       "466  ENSMUSG00000010797   100.00%   2,255      99  Passed  ENSMUST00000010941   \n",
       "467  ENSMUSG00000036856   100.00%   4,194     120  Passed  ENSMUST00000045747   \n",
       "468  ENSMUSG00000028933    96.41%   3,247     120  Passed  ENSMUST00000030773   \n",
       "469  ENSMUSG00000022361    74.82%   4,764     120  Passed  ENSMUST00000070143   \n",
       "470  ENSMUSG00000007817    81.60%   7,482     120  Passed  ENSMUST00000162645   \n",
       "471  ENSMUSG00000043207    87.90%   3,429     120  Passed  ENSMUST00000058754   \n",
       "\n",
       "       Type  \n",
       "0    Passed  \n",
       "1    Passed  \n",
       "2    Passed  \n",
       "3    Passed  \n",
       "4    Passed  \n",
       "5    Passed  \n",
       "6    Passed  \n",
       "7    Passed  \n",
       "8    Passed  \n",
       "9    Passed  \n",
       "10   Passed  \n",
       "11   Passed  \n",
       "12   Passed  \n",
       "13   Passed  \n",
       "14   Passed  \n",
       "15   Passed  \n",
       "16   Passed  \n",
       "17   Passed  \n",
       "18   Passed  \n",
       "19   Passed  \n",
       "20   Passed  \n",
       "21   Passed  \n",
       "22   Passed  \n",
       "23   Passed  \n",
       "24   Passed  \n",
       "25   Passed  \n",
       "26   Passed  \n",
       "27   Passed  \n",
       "28   Passed  \n",
       "29   Passed  \n",
       "..      ...  \n",
       "442  Passed  \n",
       "443  Passed  \n",
       "444  Passed  \n",
       "445  Passed  \n",
       "446  Passed  \n",
       "447  Passed  \n",
       "448  Passed  \n",
       "449  Passed  \n",
       "450  Passed  \n",
       "451  Passed  \n",
       "452  Passed  \n",
       "453  Passed  \n",
       "454  Passed  \n",
       "455  Passed  \n",
       "456  Passed  \n",
       "457  Passed  \n",
       "458  Passed  \n",
       "459  Passed  \n",
       "460  Passed  \n",
       "461  Passed  \n",
       "462  Passed  \n",
       "463  Passed  \n",
       "464  Passed  \n",
       "465  Passed  \n",
       "466  Passed  \n",
       "467  Passed  \n",
       "468  Passed  \n",
       "469  Passed  \n",
       "470  Passed  \n",
       "471  Passed  \n",
       "\n",
       "[472 rows x 12 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "scope = ['https://spreadsheets.google.com/feeds',\n",
    "        'https://www.googleapis.com/auth/drive']\n",
    "# Path to .json\n",
    "path = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/ProbeDesign-92ebdb6a17a7.json'\n",
    "creds = ServiceAccountCredentials.from_json_keyfile_name(path, scope)\n",
    "client = gspread.authorize(creds)\n",
    "# Start with a gene list \n",
    "expression_worksheet = client.open(\"MAPK_genes\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)\n",
    "expression_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "491"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "go_genes = []\n",
    "for i in go_gene_sheet[0]:\n",
    "    try:\n",
    "        go_genes.append(i.split(',')[0])\n",
    "    except:\n",
    "        continue\n",
    "go_genes = np.unique(go_genes)\n",
    "len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "294"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_sheet = pd.read_excel('/home/zach/PythonRepos/PyScripts/MAPK_MERFISH/MAPK_genes.xlsx',header=None)\n",
    "genes = np.unique([i.split(';')[0] for i in gene_sheet[1]])\n",
    "len(genes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "MAPK_genes = np.unique(list(go_genes)+list(genes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1NujfUqy1FliV-4YIj0HEoiNkM9X9YzAhE73QRiFZqoY',\n",
       " 'updatedRange': 'Sheet1!A2:A681',\n",
       " 'updatedRows': 680,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 680}"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene = expression_worksheet.range('A2:A'+str(len(MAPK_genes)+1))\n",
    "update = []\n",
    "for idx, g in enumerate(MAPK_genes):\n",
    "    gene[idx].value = str(g)\n",
    "expression_worksheet.update_cells(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Transcript type</th>\n",
       "      <th>Transcript length (including UTRs and CDS)</th>\n",
       "      <th>Transcript support level (TSL)</th>\n",
       "      <th>Gene description</th>\n",
       "      <th>Transcript count</th>\n",
       "      <th>RefSeq mRNA ID</th>\n",
       "      <th>NCBI gene ID</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>WikiGene description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUSG00000064336</td>\n",
       "      <td>ENSMUSG00000064336.1</td>\n",
       "      <td>ENSMUST00000082387</td>\n",
       "      <td>ENSMUST00000082387.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>68</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA phenylalanine [So...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tf</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUSG00000064337</td>\n",
       "      <td>ENSMUSG00000064337.1</td>\n",
       "      <td>ENSMUST00000082388</td>\n",
       "      <td>ENSMUST00000082388.1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>955</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded 12S rRNA [Source:MGI S...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Rnr1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUSG00000064338</td>\n",
       "      <td>ENSMUSG00000064338.1</td>\n",
       "      <td>ENSMUST00000082389</td>\n",
       "      <td>ENSMUST00000082389.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>69</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA valine [Source:MG...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tv</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUSG00000064339</td>\n",
       "      <td>ENSMUSG00000064339.1</td>\n",
       "      <td>ENSMUST00000082390</td>\n",
       "      <td>ENSMUST00000082390.1</td>\n",
       "      <td>Mt_rRNA</td>\n",
       "      <td>1582</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded 16S rRNA [Source:MGI S...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Rnr2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUSG00000064340</td>\n",
       "      <td>ENSMUSG00000064340.1</td>\n",
       "      <td>ENSMUST00000082391</td>\n",
       "      <td>ENSMUST00000082391.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>75</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA leucine 1 [Source...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tl1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ENSMUSG00000064341</td>\n",
       "      <td>ENSMUSG00000064341.1</td>\n",
       "      <td>ENSMUST00000082392</td>\n",
       "      <td>ENSMUST00000082392.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>957</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded NADH dehydrogenase 1 [...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17716.0</td>\n",
       "      <td>mt-Nd1</td>\n",
       "      <td>NADH dehydrogenase subunit 1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ENSMUSG00000064342</td>\n",
       "      <td>ENSMUSG00000064342.1</td>\n",
       "      <td>ENSMUST00000082393</td>\n",
       "      <td>ENSMUST00000082393.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>69</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA isoleucine [Sourc...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Ti</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ENSMUSG00000064343</td>\n",
       "      <td>ENSMUSG00000064343.1</td>\n",
       "      <td>ENSMUST00000082394</td>\n",
       "      <td>ENSMUST00000082394.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>71</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA glutamine [Source...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tq</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ENSMUSG00000064344</td>\n",
       "      <td>ENSMUSG00000064344.1</td>\n",
       "      <td>ENSMUST00000082395</td>\n",
       "      <td>ENSMUST00000082395.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>69</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA methionine [Sourc...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tm</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ENSMUSG00000064345</td>\n",
       "      <td>ENSMUSG00000064345.1</td>\n",
       "      <td>ENSMUST00000082396</td>\n",
       "      <td>ENSMUST00000082396.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1038</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded NADH dehydrogenase 2 [...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17717.0</td>\n",
       "      <td>mt-Nd2</td>\n",
       "      <td>NADH dehydrogenase subunit 2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>ENSMUSG00000064346</td>\n",
       "      <td>ENSMUSG00000064346.1</td>\n",
       "      <td>ENSMUST00000082397</td>\n",
       "      <td>ENSMUST00000082397.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>67</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA tryptophan [Sourc...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tw</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENSMUSG00000064347</td>\n",
       "      <td>ENSMUSG00000064347.1</td>\n",
       "      <td>ENSMUST00000082398</td>\n",
       "      <td>ENSMUST00000082398.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>69</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA alanine [Source:M...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Ta</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>ENSMUSG00000064348</td>\n",
       "      <td>ENSMUSG00000064348.1</td>\n",
       "      <td>ENSMUST00000082399</td>\n",
       "      <td>ENSMUST00000082399.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>71</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA asparagine [Sourc...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tn</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>ENSMUSG00000064349</td>\n",
       "      <td>ENSMUSG00000064349.1</td>\n",
       "      <td>ENSMUST00000082400</td>\n",
       "      <td>ENSMUST00000082400.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>66</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA cysteine [Source:...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tc</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>ENSMUSG00000064350</td>\n",
       "      <td>ENSMUSG00000064350.1</td>\n",
       "      <td>ENSMUST00000082401</td>\n",
       "      <td>ENSMUST00000082401.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>67</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA tyrosine [Source:...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Ty</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>ENSMUSG00000064351</td>\n",
       "      <td>ENSMUSG00000064351.1</td>\n",
       "      <td>ENSMUST00000082402</td>\n",
       "      <td>ENSMUST00000082402.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1545</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded cytochrome c oxidase I...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17708.0</td>\n",
       "      <td>mt-Co1</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>ENSMUSG00000064352</td>\n",
       "      <td>ENSMUSG00000064352.1</td>\n",
       "      <td>ENSMUST00000082403</td>\n",
       "      <td>ENSMUST00000082403.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>69</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA serine 1 [Source:...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Ts1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>ENSMUSG00000064353</td>\n",
       "      <td>ENSMUSG00000064353.1</td>\n",
       "      <td>ENSMUST00000082404</td>\n",
       "      <td>ENSMUST00000082404.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>70</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA aspartic acid [So...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Td</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>ENSMUSG00000064354</td>\n",
       "      <td>ENSMUSG00000064354.1</td>\n",
       "      <td>ENSMUST00000082405</td>\n",
       "      <td>ENSMUST00000082405.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>684</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded cytochrome c oxidase I...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17709.0</td>\n",
       "      <td>mt-Co2</td>\n",
       "      <td>cytochrome c oxidase subunit II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>ENSMUSG00000064355</td>\n",
       "      <td>ENSMUSG00000064355.1</td>\n",
       "      <td>ENSMUST00000082406</td>\n",
       "      <td>ENSMUST00000082406.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>65</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA lysine [Source:MG...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tk</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>ENSMUSG00000064356</td>\n",
       "      <td>ENSMUSG00000064356.3</td>\n",
       "      <td>ENSMUST00000082407</td>\n",
       "      <td>ENSMUST00000082407.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>204</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded ATP synthase 8 [Source...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17706.0</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>ATP synthase F0 subunit 8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>ENSMUSG00000064357</td>\n",
       "      <td>ENSMUSG00000064357.1</td>\n",
       "      <td>ENSMUST00000082408</td>\n",
       "      <td>ENSMUST00000082408.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>681</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded ATP synthase 6 [Source...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17705.0</td>\n",
       "      <td>mt-Atp6</td>\n",
       "      <td>ATP synthase F0 subunit 6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>ENSMUSG00000064358</td>\n",
       "      <td>ENSMUSG00000064358.1</td>\n",
       "      <td>ENSMUST00000082409</td>\n",
       "      <td>ENSMUST00000082409.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>784</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded cytochrome c oxidase I...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17710.0</td>\n",
       "      <td>mt-Co3</td>\n",
       "      <td>cytochrome c oxidase subunit III</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>ENSMUSG00000064359</td>\n",
       "      <td>ENSMUSG00000064359.1</td>\n",
       "      <td>ENSMUST00000082410</td>\n",
       "      <td>ENSMUST00000082410.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>68</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA glycine [Source:M...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tg</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>ENSMUSG00000064360</td>\n",
       "      <td>ENSMUSG00000064360.1</td>\n",
       "      <td>ENSMUST00000082411</td>\n",
       "      <td>ENSMUST00000082411.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>348</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded NADH dehydrogenase 3 [...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17718.0</td>\n",
       "      <td>mt-Nd3</td>\n",
       "      <td>NADH dehydrogenase subunit 3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>ENSMUSG00000064361</td>\n",
       "      <td>ENSMUSG00000064361.1</td>\n",
       "      <td>ENSMUST00000082412</td>\n",
       "      <td>ENSMUST00000082412.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>68</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA arginine [Source:...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Tr</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>ENSMUSG00000065947</td>\n",
       "      <td>ENSMUSG00000065947.3</td>\n",
       "      <td>ENSMUST00000084013</td>\n",
       "      <td>ENSMUST00000084013.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>297</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded NADH dehydrogenase 4L ...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17720.0</td>\n",
       "      <td>mt-Nd4l</td>\n",
       "      <td>NADH dehydrogenase subunit 4L</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>ENSMUSG00000064363</td>\n",
       "      <td>ENSMUSG00000064363.1</td>\n",
       "      <td>ENSMUST00000082414</td>\n",
       "      <td>ENSMUST00000082414.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>1378</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded NADH dehydrogenase 4 [...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17719.0</td>\n",
       "      <td>mt-Nd4</td>\n",
       "      <td>NADH dehydrogenase subunit 4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>ENSMUSG00000064364</td>\n",
       "      <td>ENSMUSG00000064364.1</td>\n",
       "      <td>ENSMUST00000082415</td>\n",
       "      <td>ENSMUST00000082415.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>67</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA histidine [Source...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Th</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>ENSMUSG00000064365</td>\n",
       "      <td>ENSMUSG00000064365.1</td>\n",
       "      <td>ENSMUST00000082416</td>\n",
       "      <td>ENSMUST00000082416.1</td>\n",
       "      <td>Mt_tRNA</td>\n",
       "      <td>59</td>\n",
       "      <td>tslNA</td>\n",
       "      <td>mitochondrially encoded tRNA serine 2 [Source:...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>mt-Ts2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152684</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000202301</td>\n",
       "      <td>ENSMUST00000202301.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NM_001369363</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152685</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000202301</td>\n",
       "      <td>ENSMUST00000202301.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6667</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NM_001205234</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152686</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000202948</td>\n",
       "      <td>ENSMUST00000202948.3</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>4456</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152687</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000201950</td>\n",
       "      <td>ENSMUST00000201950.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6059</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NM_001205235</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152688</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000236159</td>\n",
       "      <td>ENSMUST00000236159.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>5118</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NM_020253</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152689</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000202580</td>\n",
       "      <td>ENSMUST00000202580.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>5505</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152690</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000235467</td>\n",
       "      <td>ENSMUST00000235467.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>5001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152691</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000202463</td>\n",
       "      <td>ENSMUST00000202463.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>6314</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152692</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000201577</td>\n",
       "      <td>ENSMUST00000201577.2</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>604</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152693</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000201938</td>\n",
       "      <td>ENSMUST00000201938.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>648</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152694</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000202175</td>\n",
       "      <td>ENSMUST00000202175.3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>739</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152695</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000202055</td>\n",
       "      <td>ENSMUST00000202055.3</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>818</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152696</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000201830</td>\n",
       "      <td>ENSMUST00000201830.3</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>2748</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152697</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000201995</td>\n",
       "      <td>ENSMUST00000201995.1</td>\n",
       "      <td>retained_intron</td>\n",
       "      <td>1933</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152698</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000235689</td>\n",
       "      <td>ENSMUST00000235689.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>725</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152699</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000201543</td>\n",
       "      <td>ENSMUST00000201543.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>547</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152700</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000201188</td>\n",
       "      <td>ENSMUST00000201188.3</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3503</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152701</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000200719</td>\n",
       "      <td>ENSMUST00000200719.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3285</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152702</th>\n",
       "      <td>ENSMUSG00000107104</td>\n",
       "      <td>ENSMUSG00000107104.4</td>\n",
       "      <td>ENSMUST00000202198</td>\n",
       "      <td>ENSMUST00000202198.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>576</td>\n",
       "      <td>NaN</td>\n",
       "      <td>neurexin II [Source:MGI Symbol;Acc:MGI:1096362]</td>\n",
       "      <td>18</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18190.0</td>\n",
       "      <td>Nrxn2</td>\n",
       "      <td>neurexin II</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152703</th>\n",
       "      <td>ENSMUSG00000117954</td>\n",
       "      <td>ENSMUSG00000117954.1</td>\n",
       "      <td>ENSMUST00000236342</td>\n",
       "      <td>ENSMUST00000236342.1</td>\n",
       "      <td>TEC</td>\n",
       "      <td>2689</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted gene, 50151 [Source:MGI Symbol;Acc:M...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm50151</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152704</th>\n",
       "      <td>ENSMUSG00000107321</td>\n",
       "      <td>ENSMUSG00000107321.1</td>\n",
       "      <td>ENSMUST00000201735</td>\n",
       "      <td>ENSMUST00000201735.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>716</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted gene 14964 [Source:MGI Symbol;Acc:MG...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm14964</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152705</th>\n",
       "      <td>ENSMUSG00000107321</td>\n",
       "      <td>ENSMUSG00000107321.1</td>\n",
       "      <td>ENSMUST00000201326</td>\n",
       "      <td>ENSMUST00000201326.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>519</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted gene 14964 [Source:MGI Symbol;Acc:MG...</td>\n",
       "      <td>2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm14964</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152706</th>\n",
       "      <td>ENSMUSG00000107099</td>\n",
       "      <td>ENSMUSG00000107099.4</td>\n",
       "      <td>ENSMUST00000202867</td>\n",
       "      <td>ENSMUST00000202867.4</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>2226</td>\n",
       "      <td>NaN</td>\n",
       "      <td>solute carrier family 22 (organic anion/cation...</td>\n",
       "      <td>3</td>\n",
       "      <td>NM_009203</td>\n",
       "      <td>20521.0</td>\n",
       "      <td>Slc22a12</td>\n",
       "      <td>solute carrier family 22 (organic anion/cation...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152707</th>\n",
       "      <td>ENSMUSG00000107099</td>\n",
       "      <td>ENSMUSG00000107099.4</td>\n",
       "      <td>ENSMUST00000202974</td>\n",
       "      <td>ENSMUST00000202974.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>1192</td>\n",
       "      <td>NaN</td>\n",
       "      <td>solute carrier family 22 (organic anion/cation...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20521.0</td>\n",
       "      <td>Slc22a12</td>\n",
       "      <td>solute carrier family 22 (organic anion/cation...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152708</th>\n",
       "      <td>ENSMUSG00000107099</td>\n",
       "      <td>ENSMUSG00000107099.4</td>\n",
       "      <td>ENSMUST00000202211</td>\n",
       "      <td>ENSMUST00000202211.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>821</td>\n",
       "      <td>NaN</td>\n",
       "      <td>solute carrier family 22 (organic anion/cation...</td>\n",
       "      <td>3</td>\n",
       "      <td>NaN</td>\n",
       "      <td>20521.0</td>\n",
       "      <td>Slc22a12</td>\n",
       "      <td>solute carrier family 22 (organic anion/cation...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152709</th>\n",
       "      <td>ENSMUSG00000082803</td>\n",
       "      <td>ENSMUSG00000082803.2</td>\n",
       "      <td>ENSMUST00000121335</td>\n",
       "      <td>ENSMUST00000121335.2</td>\n",
       "      <td>miRNA</td>\n",
       "      <td>126</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted gene, 26460 [Source:MGI Symbol;Acc:M...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm26460</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152710</th>\n",
       "      <td>ENSMUSG00000118095</td>\n",
       "      <td>ENSMUSG00000118095.1</td>\n",
       "      <td>ENSMUST00000236254</td>\n",
       "      <td>ENSMUST00000236254.1</td>\n",
       "      <td>unprocessed_pseudogene</td>\n",
       "      <td>489</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted gene, 50435 [Source:MGI Symbol;Acc:M...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm50435</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152711</th>\n",
       "      <td>ENSMUSG00000106792</td>\n",
       "      <td>ENSMUSG00000106792.1</td>\n",
       "      <td>ENSMUST00000200952</td>\n",
       "      <td>ENSMUST00000200952.1</td>\n",
       "      <td>processed_pseudogene</td>\n",
       "      <td>635</td>\n",
       "      <td>NaN</td>\n",
       "      <td>high mobility group box 1, pseudogene 4 [Sourc...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Hmgb1-ps4</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152712</th>\n",
       "      <td>ENSMUSG00000107003</td>\n",
       "      <td>ENSMUSG00000107003.1</td>\n",
       "      <td>ENSMUST00000202738</td>\n",
       "      <td>ENSMUST00000202738.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>871</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RIKEN cDNA 2410152P15 gene [Source:MGI Symbol;...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2410152P15Rik</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>152713</th>\n",
       "      <td>ENSMUSG00000106796</td>\n",
       "      <td>ENSMUSG00000106796.1</td>\n",
       "      <td>ENSMUST00000201826</td>\n",
       "      <td>ENSMUST00000201826.1</td>\n",
       "      <td>lncRNA</td>\n",
       "      <td>769</td>\n",
       "      <td>NaN</td>\n",
       "      <td>predicted gene 14968 [Source:MGI Symbol;Acc:MG...</td>\n",
       "      <td>1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Gm14968</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>152714 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0       ENSMUSG00000064336   ENSMUSG00000064336.1   ENSMUST00000082387   \n",
       "1       ENSMUSG00000064337   ENSMUSG00000064337.1   ENSMUST00000082388   \n",
       "2       ENSMUSG00000064338   ENSMUSG00000064338.1   ENSMUST00000082389   \n",
       "3       ENSMUSG00000064339   ENSMUSG00000064339.1   ENSMUST00000082390   \n",
       "4       ENSMUSG00000064340   ENSMUSG00000064340.1   ENSMUST00000082391   \n",
       "5       ENSMUSG00000064341   ENSMUSG00000064341.1   ENSMUST00000082392   \n",
       "6       ENSMUSG00000064342   ENSMUSG00000064342.1   ENSMUST00000082393   \n",
       "7       ENSMUSG00000064343   ENSMUSG00000064343.1   ENSMUST00000082394   \n",
       "8       ENSMUSG00000064344   ENSMUSG00000064344.1   ENSMUST00000082395   \n",
       "9       ENSMUSG00000064345   ENSMUSG00000064345.1   ENSMUST00000082396   \n",
       "10      ENSMUSG00000064346   ENSMUSG00000064346.1   ENSMUST00000082397   \n",
       "11      ENSMUSG00000064347   ENSMUSG00000064347.1   ENSMUST00000082398   \n",
       "12      ENSMUSG00000064348   ENSMUSG00000064348.1   ENSMUST00000082399   \n",
       "13      ENSMUSG00000064349   ENSMUSG00000064349.1   ENSMUST00000082400   \n",
       "14      ENSMUSG00000064350   ENSMUSG00000064350.1   ENSMUST00000082401   \n",
       "15      ENSMUSG00000064351   ENSMUSG00000064351.1   ENSMUST00000082402   \n",
       "16      ENSMUSG00000064352   ENSMUSG00000064352.1   ENSMUST00000082403   \n",
       "17      ENSMUSG00000064353   ENSMUSG00000064353.1   ENSMUST00000082404   \n",
       "18      ENSMUSG00000064354   ENSMUSG00000064354.1   ENSMUST00000082405   \n",
       "19      ENSMUSG00000064355   ENSMUSG00000064355.1   ENSMUST00000082406   \n",
       "20      ENSMUSG00000064356   ENSMUSG00000064356.3   ENSMUST00000082407   \n",
       "21      ENSMUSG00000064357   ENSMUSG00000064357.1   ENSMUST00000082408   \n",
       "22      ENSMUSG00000064358   ENSMUSG00000064358.1   ENSMUST00000082409   \n",
       "23      ENSMUSG00000064359   ENSMUSG00000064359.1   ENSMUST00000082410   \n",
       "24      ENSMUSG00000064360   ENSMUSG00000064360.1   ENSMUST00000082411   \n",
       "25      ENSMUSG00000064361   ENSMUSG00000064361.1   ENSMUST00000082412   \n",
       "26      ENSMUSG00000065947   ENSMUSG00000065947.3   ENSMUST00000084013   \n",
       "27      ENSMUSG00000064363   ENSMUSG00000064363.1   ENSMUST00000082414   \n",
       "28      ENSMUSG00000064364   ENSMUSG00000064364.1   ENSMUST00000082415   \n",
       "29      ENSMUSG00000064365   ENSMUSG00000064365.1   ENSMUST00000082416   \n",
       "...                    ...                    ...                  ...   \n",
       "152684  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000202301   \n",
       "152685  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000202301   \n",
       "152686  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000202948   \n",
       "152687  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000201950   \n",
       "152688  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000236159   \n",
       "152689  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000202580   \n",
       "152690  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000235467   \n",
       "152691  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000202463   \n",
       "152692  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000201577   \n",
       "152693  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000201938   \n",
       "152694  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000202175   \n",
       "152695  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000202055   \n",
       "152696  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000201830   \n",
       "152697  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000201995   \n",
       "152698  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000235689   \n",
       "152699  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000201543   \n",
       "152700  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000201188   \n",
       "152701  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000200719   \n",
       "152702  ENSMUSG00000107104   ENSMUSG00000107104.4   ENSMUST00000202198   \n",
       "152703  ENSMUSG00000117954   ENSMUSG00000117954.1   ENSMUST00000236342   \n",
       "152704  ENSMUSG00000107321   ENSMUSG00000107321.1   ENSMUST00000201735   \n",
       "152705  ENSMUSG00000107321   ENSMUSG00000107321.1   ENSMUST00000201326   \n",
       "152706  ENSMUSG00000107099   ENSMUSG00000107099.4   ENSMUST00000202867   \n",
       "152707  ENSMUSG00000107099   ENSMUSG00000107099.4   ENSMUST00000202974   \n",
       "152708  ENSMUSG00000107099   ENSMUSG00000107099.4   ENSMUST00000202211   \n",
       "152709  ENSMUSG00000082803   ENSMUSG00000082803.2   ENSMUST00000121335   \n",
       "152710  ENSMUSG00000118095   ENSMUSG00000118095.1   ENSMUST00000236254   \n",
       "152711  ENSMUSG00000106792   ENSMUSG00000106792.1   ENSMUST00000200952   \n",
       "152712  ENSMUSG00000107003   ENSMUSG00000107003.1   ENSMUST00000202738   \n",
       "152713  ENSMUSG00000106796   ENSMUSG00000106796.1   ENSMUST00000201826   \n",
       "\n",
       "       Transcript stable ID version         Transcript type  \\\n",
       "0              ENSMUST00000082387.1                 Mt_tRNA   \n",
       "1              ENSMUST00000082388.1                 Mt_rRNA   \n",
       "2              ENSMUST00000082389.1                 Mt_tRNA   \n",
       "3              ENSMUST00000082390.1                 Mt_rRNA   \n",
       "4              ENSMUST00000082391.1                 Mt_tRNA   \n",
       "5              ENSMUST00000082392.1          protein_coding   \n",
       "6              ENSMUST00000082393.1                 Mt_tRNA   \n",
       "7              ENSMUST00000082394.1                 Mt_tRNA   \n",
       "8              ENSMUST00000082395.1                 Mt_tRNA   \n",
       "9              ENSMUST00000082396.1          protein_coding   \n",
       "10             ENSMUST00000082397.1                 Mt_tRNA   \n",
       "11             ENSMUST00000082398.1                 Mt_tRNA   \n",
       "12             ENSMUST00000082399.1                 Mt_tRNA   \n",
       "13             ENSMUST00000082400.1                 Mt_tRNA   \n",
       "14             ENSMUST00000082401.1                 Mt_tRNA   \n",
       "15             ENSMUST00000082402.1          protein_coding   \n",
       "16             ENSMUST00000082403.1                 Mt_tRNA   \n",
       "17             ENSMUST00000082404.1                 Mt_tRNA   \n",
       "18             ENSMUST00000082405.1          protein_coding   \n",
       "19             ENSMUST00000082406.1                 Mt_tRNA   \n",
       "20             ENSMUST00000082407.1          protein_coding   \n",
       "21             ENSMUST00000082408.1          protein_coding   \n",
       "22             ENSMUST00000082409.1          protein_coding   \n",
       "23             ENSMUST00000082410.1                 Mt_tRNA   \n",
       "24             ENSMUST00000082411.1          protein_coding   \n",
       "25             ENSMUST00000082412.1                 Mt_tRNA   \n",
       "26             ENSMUST00000084013.1          protein_coding   \n",
       "27             ENSMUST00000082414.1          protein_coding   \n",
       "28             ENSMUST00000082415.1                 Mt_tRNA   \n",
       "29             ENSMUST00000082416.1                 Mt_tRNA   \n",
       "...                             ...                     ...   \n",
       "152684         ENSMUST00000202301.3          protein_coding   \n",
       "152685         ENSMUST00000202301.3          protein_coding   \n",
       "152686         ENSMUST00000202948.3         retained_intron   \n",
       "152687         ENSMUST00000201950.3          protein_coding   \n",
       "152688         ENSMUST00000236159.1          protein_coding   \n",
       "152689         ENSMUST00000202580.3          protein_coding   \n",
       "152690         ENSMUST00000235467.1          protein_coding   \n",
       "152691         ENSMUST00000202463.3          protein_coding   \n",
       "152692         ENSMUST00000201577.2                  lncRNA   \n",
       "152693         ENSMUST00000201938.3          protein_coding   \n",
       "152694         ENSMUST00000202175.3                  lncRNA   \n",
       "152695         ENSMUST00000202055.3                  lncRNA   \n",
       "152696         ENSMUST00000201830.3         retained_intron   \n",
       "152697         ENSMUST00000201995.1         retained_intron   \n",
       "152698         ENSMUST00000235689.1                  lncRNA   \n",
       "152699         ENSMUST00000201543.1                  lncRNA   \n",
       "152700         ENSMUST00000201188.3          protein_coding   \n",
       "152701         ENSMUST00000200719.1          protein_coding   \n",
       "152702         ENSMUST00000202198.1                  lncRNA   \n",
       "152703         ENSMUST00000236342.1                     TEC   \n",
       "152704         ENSMUST00000201735.1                  lncRNA   \n",
       "152705         ENSMUST00000201326.1                  lncRNA   \n",
       "152706         ENSMUST00000202867.4          protein_coding   \n",
       "152707         ENSMUST00000202974.1                  lncRNA   \n",
       "152708         ENSMUST00000202211.1          protein_coding   \n",
       "152709         ENSMUST00000121335.2                   miRNA   \n",
       "152710         ENSMUST00000236254.1  unprocessed_pseudogene   \n",
       "152711         ENSMUST00000200952.1    processed_pseudogene   \n",
       "152712         ENSMUST00000202738.1                  lncRNA   \n",
       "152713         ENSMUST00000201826.1                  lncRNA   \n",
       "\n",
       "        Transcript length (including UTRs and CDS)  \\\n",
       "0                                               68   \n",
       "1                                              955   \n",
       "2                                               69   \n",
       "3                                             1582   \n",
       "4                                               75   \n",
       "5                                              957   \n",
       "6                                               69   \n",
       "7                                               71   \n",
       "8                                               69   \n",
       "9                                             1038   \n",
       "10                                              67   \n",
       "11                                              69   \n",
       "12                                              71   \n",
       "13                                              66   \n",
       "14                                              67   \n",
       "15                                            1545   \n",
       "16                                              69   \n",
       "17                                              70   \n",
       "18                                             684   \n",
       "19                                              65   \n",
       "20                                             204   \n",
       "21                                             681   \n",
       "22                                             784   \n",
       "23                                              68   \n",
       "24                                             348   \n",
       "25                                              68   \n",
       "26                                             297   \n",
       "27                                            1378   \n",
       "28                                              67   \n",
       "29                                              59   \n",
       "...                                            ...   \n",
       "152684                                        6667   \n",
       "152685                                        6667   \n",
       "152686                                        4456   \n",
       "152687                                        6059   \n",
       "152688                                        5118   \n",
       "152689                                        5505   \n",
       "152690                                        5001   \n",
       "152691                                        6314   \n",
       "152692                                         604   \n",
       "152693                                         648   \n",
       "152694                                         739   \n",
       "152695                                         818   \n",
       "152696                                        2748   \n",
       "152697                                        1933   \n",
       "152698                                         725   \n",
       "152699                                         547   \n",
       "152700                                        3503   \n",
       "152701                                        3285   \n",
       "152702                                         576   \n",
       "152703                                        2689   \n",
       "152704                                         716   \n",
       "152705                                         519   \n",
       "152706                                        2226   \n",
       "152707                                        1192   \n",
       "152708                                         821   \n",
       "152709                                         126   \n",
       "152710                                         489   \n",
       "152711                                         635   \n",
       "152712                                         871   \n",
       "152713                                         769   \n",
       "\n",
       "       Transcript support level (TSL)  \\\n",
       "0                               tslNA   \n",
       "1                               tslNA   \n",
       "2                               tslNA   \n",
       "3                               tslNA   \n",
       "4                               tslNA   \n",
       "5                               tslNA   \n",
       "6                               tslNA   \n",
       "7                               tslNA   \n",
       "8                               tslNA   \n",
       "9                               tslNA   \n",
       "10                              tslNA   \n",
       "11                              tslNA   \n",
       "12                              tslNA   \n",
       "13                              tslNA   \n",
       "14                              tslNA   \n",
       "15                              tslNA   \n",
       "16                              tslNA   \n",
       "17                              tslNA   \n",
       "18                              tslNA   \n",
       "19                              tslNA   \n",
       "20                              tslNA   \n",
       "21                              tslNA   \n",
       "22                              tslNA   \n",
       "23                              tslNA   \n",
       "24                              tslNA   \n",
       "25                              tslNA   \n",
       "26                              tslNA   \n",
       "27                              tslNA   \n",
       "28                              tslNA   \n",
       "29                              tslNA   \n",
       "...                               ...   \n",
       "152684                            NaN   \n",
       "152685                            NaN   \n",
       "152686                            NaN   \n",
       "152687                            NaN   \n",
       "152688                            NaN   \n",
       "152689                            NaN   \n",
       "152690                            NaN   \n",
       "152691                            NaN   \n",
       "152692                            NaN   \n",
       "152693                            NaN   \n",
       "152694                            NaN   \n",
       "152695                            NaN   \n",
       "152696                            NaN   \n",
       "152697                            NaN   \n",
       "152698                            NaN   \n",
       "152699                            NaN   \n",
       "152700                            NaN   \n",
       "152701                            NaN   \n",
       "152702                            NaN   \n",
       "152703                            NaN   \n",
       "152704                            NaN   \n",
       "152705                            NaN   \n",
       "152706                            NaN   \n",
       "152707                            NaN   \n",
       "152708                            NaN   \n",
       "152709                            NaN   \n",
       "152710                            NaN   \n",
       "152711                            NaN   \n",
       "152712                            NaN   \n",
       "152713                            NaN   \n",
       "\n",
       "                                         Gene description  Transcript count  \\\n",
       "0       mitochondrially encoded tRNA phenylalanine [So...                 1   \n",
       "1       mitochondrially encoded 12S rRNA [Source:MGI S...                 1   \n",
       "2       mitochondrially encoded tRNA valine [Source:MG...                 1   \n",
       "3       mitochondrially encoded 16S rRNA [Source:MGI S...                 1   \n",
       "4       mitochondrially encoded tRNA leucine 1 [Source...                 1   \n",
       "5       mitochondrially encoded NADH dehydrogenase 1 [...                 1   \n",
       "6       mitochondrially encoded tRNA isoleucine [Sourc...                 1   \n",
       "7       mitochondrially encoded tRNA glutamine [Source...                 1   \n",
       "8       mitochondrially encoded tRNA methionine [Sourc...                 1   \n",
       "9       mitochondrially encoded NADH dehydrogenase 2 [...                 1   \n",
       "10      mitochondrially encoded tRNA tryptophan [Sourc...                 1   \n",
       "11      mitochondrially encoded tRNA alanine [Source:M...                 1   \n",
       "12      mitochondrially encoded tRNA asparagine [Sourc...                 1   \n",
       "13      mitochondrially encoded tRNA cysteine [Source:...                 1   \n",
       "14      mitochondrially encoded tRNA tyrosine [Source:...                 1   \n",
       "15      mitochondrially encoded cytochrome c oxidase I...                 1   \n",
       "16      mitochondrially encoded tRNA serine 1 [Source:...                 1   \n",
       "17      mitochondrially encoded tRNA aspartic acid [So...                 1   \n",
       "18      mitochondrially encoded cytochrome c oxidase I...                 1   \n",
       "19      mitochondrially encoded tRNA lysine [Source:MG...                 1   \n",
       "20      mitochondrially encoded ATP synthase 8 [Source...                 1   \n",
       "21      mitochondrially encoded ATP synthase 6 [Source...                 1   \n",
       "22      mitochondrially encoded cytochrome c oxidase I...                 1   \n",
       "23      mitochondrially encoded tRNA glycine [Source:M...                 1   \n",
       "24      mitochondrially encoded NADH dehydrogenase 3 [...                 1   \n",
       "25      mitochondrially encoded tRNA arginine [Source:...                 1   \n",
       "26      mitochondrially encoded NADH dehydrogenase 4L ...                 1   \n",
       "27      mitochondrially encoded NADH dehydrogenase 4 [...                 1   \n",
       "28      mitochondrially encoded tRNA histidine [Source...                 1   \n",
       "29      mitochondrially encoded tRNA serine 2 [Source:...                 1   \n",
       "...                                                   ...               ...   \n",
       "152684    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152685    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152686    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152687    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152688    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152689    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152690    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152691    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152692    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152693    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152694    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152695    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152696    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152697    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152698    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152699    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152700    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152701    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152702    neurexin II [Source:MGI Symbol;Acc:MGI:1096362]                18   \n",
       "152703  predicted gene, 50151 [Source:MGI Symbol;Acc:M...                 1   \n",
       "152704  predicted gene 14964 [Source:MGI Symbol;Acc:MG...                 2   \n",
       "152705  predicted gene 14964 [Source:MGI Symbol;Acc:MG...                 2   \n",
       "152706  solute carrier family 22 (organic anion/cation...                 3   \n",
       "152707  solute carrier family 22 (organic anion/cation...                 3   \n",
       "152708  solute carrier family 22 (organic anion/cation...                 3   \n",
       "152709  predicted gene, 26460 [Source:MGI Symbol;Acc:M...                 1   \n",
       "152710  predicted gene, 50435 [Source:MGI Symbol;Acc:M...                 1   \n",
       "152711  high mobility group box 1, pseudogene 4 [Sourc...                 1   \n",
       "152712  RIKEN cDNA 2410152P15 gene [Source:MGI Symbol;...                 1   \n",
       "152713  predicted gene 14968 [Source:MGI Symbol;Acc:MG...                 1   \n",
       "\n",
       "       RefSeq mRNA ID  NCBI gene ID      Gene name  \\\n",
       "0                 NaN           NaN          mt-Tf   \n",
       "1                 NaN           NaN        mt-Rnr1   \n",
       "2                 NaN           NaN          mt-Tv   \n",
       "3                 NaN           NaN        mt-Rnr2   \n",
       "4                 NaN           NaN         mt-Tl1   \n",
       "5                 NaN       17716.0         mt-Nd1   \n",
       "6                 NaN           NaN          mt-Ti   \n",
       "7                 NaN           NaN          mt-Tq   \n",
       "8                 NaN           NaN          mt-Tm   \n",
       "9                 NaN       17717.0         mt-Nd2   \n",
       "10                NaN           NaN          mt-Tw   \n",
       "11                NaN           NaN          mt-Ta   \n",
       "12                NaN           NaN          mt-Tn   \n",
       "13                NaN           NaN          mt-Tc   \n",
       "14                NaN           NaN          mt-Ty   \n",
       "15                NaN       17708.0         mt-Co1   \n",
       "16                NaN           NaN         mt-Ts1   \n",
       "17                NaN           NaN          mt-Td   \n",
       "18                NaN       17709.0         mt-Co2   \n",
       "19                NaN           NaN          mt-Tk   \n",
       "20                NaN       17706.0        mt-Atp8   \n",
       "21                NaN       17705.0        mt-Atp6   \n",
       "22                NaN       17710.0         mt-Co3   \n",
       "23                NaN           NaN          mt-Tg   \n",
       "24                NaN       17718.0         mt-Nd3   \n",
       "25                NaN           NaN          mt-Tr   \n",
       "26                NaN       17720.0        mt-Nd4l   \n",
       "27                NaN       17719.0         mt-Nd4   \n",
       "28                NaN           NaN          mt-Th   \n",
       "29                NaN           NaN         mt-Ts2   \n",
       "...               ...           ...            ...   \n",
       "152684   NM_001369363       18190.0          Nrxn2   \n",
       "152685   NM_001205234       18190.0          Nrxn2   \n",
       "152686            NaN       18190.0          Nrxn2   \n",
       "152687   NM_001205235       18190.0          Nrxn2   \n",
       "152688      NM_020253       18190.0          Nrxn2   \n",
       "152689            NaN       18190.0          Nrxn2   \n",
       "152690            NaN       18190.0          Nrxn2   \n",
       "152691            NaN       18190.0          Nrxn2   \n",
       "152692            NaN       18190.0          Nrxn2   \n",
       "152693            NaN       18190.0          Nrxn2   \n",
       "152694            NaN       18190.0          Nrxn2   \n",
       "152695            NaN       18190.0          Nrxn2   \n",
       "152696            NaN       18190.0          Nrxn2   \n",
       "152697            NaN       18190.0          Nrxn2   \n",
       "152698            NaN       18190.0          Nrxn2   \n",
       "152699            NaN       18190.0          Nrxn2   \n",
       "152700            NaN       18190.0          Nrxn2   \n",
       "152701            NaN       18190.0          Nrxn2   \n",
       "152702            NaN       18190.0          Nrxn2   \n",
       "152703            NaN           NaN        Gm50151   \n",
       "152704            NaN           NaN        Gm14964   \n",
       "152705            NaN           NaN        Gm14964   \n",
       "152706      NM_009203       20521.0       Slc22a12   \n",
       "152707            NaN       20521.0       Slc22a12   \n",
       "152708            NaN       20521.0       Slc22a12   \n",
       "152709            NaN           NaN        Gm26460   \n",
       "152710            NaN           NaN        Gm50435   \n",
       "152711            NaN           NaN      Hmgb1-ps4   \n",
       "152712            NaN           NaN  2410152P15Rik   \n",
       "152713            NaN           NaN        Gm14968   \n",
       "\n",
       "                                     WikiGene description  \n",
       "0                                                     NaN  \n",
       "1                                                     NaN  \n",
       "2                                                     NaN  \n",
       "3                                                     NaN  \n",
       "4                                                     NaN  \n",
       "5                            NADH dehydrogenase subunit 1  \n",
       "6                                                     NaN  \n",
       "7                                                     NaN  \n",
       "8                                                     NaN  \n",
       "9                            NADH dehydrogenase subunit 2  \n",
       "10                                                    NaN  \n",
       "11                                                    NaN  \n",
       "12                                                    NaN  \n",
       "13                                                    NaN  \n",
       "14                                                    NaN  \n",
       "15                         cytochrome c oxidase subunit I  \n",
       "16                                                    NaN  \n",
       "17                                                    NaN  \n",
       "18                        cytochrome c oxidase subunit II  \n",
       "19                                                    NaN  \n",
       "20                              ATP synthase F0 subunit 8  \n",
       "21                              ATP synthase F0 subunit 6  \n",
       "22                       cytochrome c oxidase subunit III  \n",
       "23                                                    NaN  \n",
       "24                           NADH dehydrogenase subunit 3  \n",
       "25                                                    NaN  \n",
       "26                          NADH dehydrogenase subunit 4L  \n",
       "27                           NADH dehydrogenase subunit 4  \n",
       "28                                                    NaN  \n",
       "29                                                    NaN  \n",
       "...                                                   ...  \n",
       "152684                                        neurexin II  \n",
       "152685                                        neurexin II  \n",
       "152686                                        neurexin II  \n",
       "152687                                        neurexin II  \n",
       "152688                                        neurexin II  \n",
       "152689                                        neurexin II  \n",
       "152690                                        neurexin II  \n",
       "152691                                        neurexin II  \n",
       "152692                                        neurexin II  \n",
       "152693                                        neurexin II  \n",
       "152694                                        neurexin II  \n",
       "152695                                        neurexin II  \n",
       "152696                                        neurexin II  \n",
       "152697                                        neurexin II  \n",
       "152698                                        neurexin II  \n",
       "152699                                        neurexin II  \n",
       "152700                                        neurexin II  \n",
       "152701                                        neurexin II  \n",
       "152702                                        neurexin II  \n",
       "152703                                                NaN  \n",
       "152704                                                NaN  \n",
       "152705                                                NaN  \n",
       "152706  solute carrier family 22 (organic anion/cation...  \n",
       "152707  solute carrier family 22 (organic anion/cation...  \n",
       "152708  solute carrier family 22 (organic anion/cation...  \n",
       "152709                                                NaN  \n",
       "152710                                                NaN  \n",
       "152711                                                NaN  \n",
       "152712                                                NaN  \n",
       "152713                                                NaN  \n",
       "\n",
       "[152714 rows x 13 columns]"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "annotation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>tracking_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>nearest_ref_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_short_name</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>locus</th>\n",
       "      <th>length</th>\n",
       "      <th>coverage</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>FPKM_conf_lo</th>\n",
       "      <th>FPKM_conf_hi</th>\n",
       "      <th>FPKM_status</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>tracking_id</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000193812</th>\n",
       "      <td>ENSMUST00000193812</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3073252-3074322</td>\n",
       "      <td>1070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000082908</th>\n",
       "      <td>ENSMUST00000082908</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3102015-3102125</td>\n",
       "      <td>110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000193244</th>\n",
       "      <td>ENSMUST00000193244</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000102348</td>\n",
       "      <td>Gm10568</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3680154-3681788</td>\n",
       "      <td>1634</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000194454</th>\n",
       "      <td>ENSMUST00000194454</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000102592</td>\n",
       "      <td>Gm38385</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3752009-3754360</td>\n",
       "      <td>2351</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000157708</th>\n",
       "      <td>ENSMUST00000157708</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000088333</td>\n",
       "      <td>Gm27396</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3783875-3783933</td>\n",
       "      <td>58</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           tracking_id class_code nearest_ref_id  \\\n",
       "tracking_id                                                        \n",
       "ENSMUST00000193812  ENSMUST00000193812          -              -   \n",
       "ENSMUST00000082908  ENSMUST00000082908          -              -   \n",
       "ENSMUST00000193244  ENSMUST00000193244          -              -   \n",
       "ENSMUST00000194454  ENSMUST00000194454          -              -   \n",
       "ENSMUST00000157708  ENSMUST00000157708          -              -   \n",
       "\n",
       "                               gene_id gene_short_name tss_id  \\\n",
       "tracking_id                                                     \n",
       "ENSMUST00000193812  ENSMUSG00000102693   4933401J01Rik      -   \n",
       "ENSMUST00000082908  ENSMUSG00000064842         Gm26206      -   \n",
       "ENSMUST00000193244  ENSMUSG00000102348         Gm10568      -   \n",
       "ENSMUST00000194454  ENSMUSG00000102592         Gm38385      -   \n",
       "ENSMUST00000157708  ENSMUSG00000088333         Gm27396      -   \n",
       "\n",
       "                                locus  length  coverage  FPKM  FPKM_conf_lo  \\\n",
       "tracking_id                                                                   \n",
       "ENSMUST00000193812  1:3073252-3074322    1070       0.0   0.0           0.0   \n",
       "ENSMUST00000082908  1:3102015-3102125     110       0.0   0.0           0.0   \n",
       "ENSMUST00000193244  1:3680154-3681788    1634       0.0   0.0           0.0   \n",
       "ENSMUST00000194454  1:3752009-3754360    2351       0.0   0.0           0.0   \n",
       "ENSMUST00000157708  1:3783875-3783933      58       0.0   0.0           0.0   \n",
       "\n",
       "                    FPKM_conf_hi FPKM_status  \n",
       "tracking_id                                   \n",
       "ENSMUST00000193812           0.0          OK  \n",
       "ENSMUST00000082908           0.0          OK  \n",
       "ENSMUST00000193244           0.0          OK  \n",
       "ENSMUST00000194454           0.0          OK  \n",
       "ENSMUST00000157708           0.0          OK  "
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/bigstore/binfo/mouse/3T3/GSM2609311_cufflinks/isoforms.fpkm_tracking'\n",
    "ReadsPerGene = pd.read_csv(f,sep='\\t')\n",
    "ReadsPerGene.index=ReadsPerGene.tracking_id\n",
    "fpkm2 = ReadsPerGene\n",
    "ReadsPerGene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>nearest_ref_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_short_name</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>locus</th>\n",
       "      <th>length</th>\n",
       "      <th>coverage</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>FPKM_conf_lo</th>\n",
       "      <th>FPKM_conf_hi</th>\n",
       "      <th>FPKM_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSMUST00000193812</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000102693</td>\n",
       "      <td>4933401J01Rik</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3073252-3074322</td>\n",
       "      <td>1070</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSMUST00000082908</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064842</td>\n",
       "      <td>Gm26206</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3102015-3102125</td>\n",
       "      <td>110</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSMUST00000193244</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000102348</td>\n",
       "      <td>Gm10568</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3680154-3681788</td>\n",
       "      <td>1634</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSMUST00000194454</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000102592</td>\n",
       "      <td>Gm38385</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3752009-3754360</td>\n",
       "      <td>2351</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSMUST00000157708</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000088333</td>\n",
       "      <td>Gm27396</td>\n",
       "      <td>-</td>\n",
       "      <td>1:3783875-3783933</td>\n",
       "      <td>58</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        transcript_id class_code nearest_ref_id             gene_id  \\\n",
       "0  ENSMUST00000193812          -              -  ENSMUSG00000102693   \n",
       "1  ENSMUST00000082908          -              -  ENSMUSG00000064842   \n",
       "2  ENSMUST00000193244          -              -  ENSMUSG00000102348   \n",
       "3  ENSMUST00000194454          -              -  ENSMUSG00000102592   \n",
       "4  ENSMUST00000157708          -              -  ENSMUSG00000088333   \n",
       "\n",
       "  gene_short_name tss_id              locus  length  coverage  FPKM  \\\n",
       "0   4933401J01Rik      -  1:3073252-3074322    1070       0.0   0.0   \n",
       "1         Gm26206      -  1:3102015-3102125     110       0.0   0.0   \n",
       "2         Gm10568      -  1:3680154-3681788    1634       0.0   0.0   \n",
       "3         Gm38385      -  1:3752009-3754360    2351       0.0   0.0   \n",
       "4         Gm27396      -  1:3783875-3783933      58       0.0   0.0   \n",
       "\n",
       "   FPKM_conf_lo  FPKM_conf_hi FPKM_status  \n",
       "0           0.0           0.0          OK  \n",
       "1           0.0           0.0          OK  \n",
       "2           0.0           0.0          OK  \n",
       "3           0.0           0.0          OK  \n",
       "4           0.0           0.0          OK  "
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/bigstore/binfo/mouse/3T3/GSM2609311_cufflinks/isoforms.fpkm_tracking'\n",
    "ReadsPerGene = pd.read_csv(f,sep='\\t')\n",
    "ReadsPerGene.columns = ['transcript_id', 'class_code', 'nearest_ref_id', 'gene_id',\n",
    "       'gene_short_name', 'tss_id', 'locus', 'length', 'coverage', 'FPKM',\n",
    "       'FPKM_conf_lo', 'FPKM_conf_hi', 'FPKM_status']\n",
    "ReadsPerGene.to_csv('/bigstore/binfo/mouse/3T3/GSM2609311_cufflinks/isoforms.csv')\n",
    "ReadsPerGene.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zach/miniconda3/envs/pyspots/lib/python3.6/site-packages/ipykernel_launcher.py:80: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/zach/miniconda3/envs/pyspots/lib/python3.6/site-packages/ipykernel_launcher.py:81: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n",
      "/home/zach/miniconda3/envs/pyspots/lib/python3.6/site-packages/ipykernel_launcher.py:76: RuntimeWarning: invalid value encountered in double_scalars\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "gid not found for  Ccn2 None\n",
      "gid not found for  Ccn3 None\n",
      "gid not found for  Mir1186 None\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1NujfUqy1FliV-4YIj0HEoiNkM9X9YzAhE73QRiFZqoY',\n",
       " 'updatedRange': 'Sheet1!B2:B289',\n",
       " 'updatedRows': 288,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 288}"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_worksheet = client.open(\"MAPK_genes\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)\n",
    "\n",
    "Transcript_IDs = expression_worksheet.range('G2:G'+str(len(expression_df)+1))\n",
    "Gene_IDs = expression_worksheet.range('F2:F'+str(len(expression_df)+1))\n",
    "FPKMs = expression_worksheet.range('C2:C'+str(len(expression_df)+1))\n",
    "lengths = expression_worksheet.range('D2:D'+str(len(expression_df)+1))\n",
    "Descriptions = expression_worksheet.range('B2:B'+str(len(expression_df)+1))\n",
    "Genes = expression_worksheet.range('A2:A'+str(len(expression_df)+1))\n",
    "Isoforms = expression_worksheet.range('I2:I'+str(len(expression_df)+1))\n",
    "Type = expression_worksheet.range('J2:J'+str(len(expression_df)+1))\n",
    "\n",
    "f = '/bigstore/binfo/mouse/3T3/GSM2609311_cufflinks/isoforms.fpkm_tracking'\n",
    "ReadsPerGene = pd.read_csv(f,sep='\\t')\n",
    "ReadsPerGene.index=ReadsPerGene.tracking_id\n",
    "fpkm = ReadsPerGene\n",
    "\n",
    "annotation = pd.read_csv('/bigstore/binfo/mouse/mart_export_25Nov2019.txt',sep='\\t')\n",
    "\n",
    "flag = 'gene'\n",
    "FPKM_gene_name_col = 'gene_short_name'#'Gene name'\n",
    "FPKM_gid_col = 'gene_id'#'Gene stable ID'\n",
    "FPKM_tid_col = 'tracking_id'#'Transcript stable ID'\n",
    "tid_col = 'Transcript stable ID'\n",
    "gid_col = 'Gene stable ID'\n",
    "ttype_col = 'Transcript type'\n",
    "gene_name_col = 'Gene name'\n",
    "length_col = 'Transcript length (including UTRs and CDS)'\n",
    "gene_description_col = 'Gene description'\n",
    "FPKM_FPKM_col = 'FPKM'#'IFNAR-0_tot.bam'\n",
    "completed = []\n",
    "\n",
    "for i,cell in enumerate(Genes):\n",
    "    gene = cell.value\n",
    "    if gene == '':\n",
    "        continue\n",
    "    if gene in completed:\n",
    "        Type[i].value= 'duplicate'\n",
    "        print(gene,' duplicated')\n",
    "    if flag=='gene':\n",
    "        try:\n",
    "            gid = fpkm[fpkm[FPKM_gene_name_col]==gene][FPKM_gid_col].unique()[0]\n",
    "        except:\n",
    "            gid = 'None'\n",
    "    elif flag == 'gid':\n",
    "        gid = gene\n",
    "    elif flag == 'tid':\n",
    "        try:\n",
    "            gid = fpkm[fpkm[FPKM_tid_col]==gene][FPKM_gid_col].unique()[0]\n",
    "            master_tid = gene.split('.')[0]\n",
    "        except:\n",
    "            gid = 'None'\n",
    "    else:\n",
    "        print('flag is wrong')\n",
    "    if gid=='None':\n",
    "        Type[i].value= 'Failed'\n",
    "        Transcript_IDs[i].value = ''\n",
    "        Gene_IDs[i].value = ''\n",
    "        tid_fpkm = ''\n",
    "        FPKMs[i].value = ''\n",
    "        lengths[i].value = ''\n",
    "        Descriptions[i].value = ''\n",
    "#         Genes[i].value = ''\n",
    "        print('gid not found for ',gene,gid)\n",
    "        continue\n",
    "    FPKM = []\n",
    "    I = []\n",
    "    g_annotation = annotation[annotation[gid_col]==gid]\n",
    "    g_annotation.drop_duplicates(tid_col)\n",
    "    for t in g_annotation[tid_col]:\n",
    "        try:\n",
    "            f = fpkm[fpkm[FPKM_tid_col]==t][FPKM_FPKM_col].iloc[0]\n",
    "            g = np.sum(fpkm[fpkm[FPKM_gid_col]==gid][FPKM_FPKM_col])\n",
    "            FPKM.append(f)\n",
    "            I.append(f/g)\n",
    "        except:\n",
    "            FPKM.append(np.nan)\n",
    "            I.append(np.nan)\n",
    "    g_annotation['FPKM'] = FPKM\n",
    "    g_annotation['I'] = I\n",
    "    gene_FPKM_df = g_annotation\n",
    "    if 'protein_coding' in list(gene_FPKM_df['Transcript type']):\n",
    "        gene_FPKM_df = gene_FPKM_df[gene_FPKM_df['Transcript type']=='protein_coding']\n",
    "    mask = []\n",
    "    for t in gene_FPKM_df['Transcript support level (TSL)']:\n",
    "        try:\n",
    "            if 'tsl1' in t:\n",
    "                mask.append(True)\n",
    "            else: \n",
    "                mask.append(False)\n",
    "        except:\n",
    "            mask.append(False)\n",
    "    if len(gene_FPKM_df[mask])>0:\n",
    "        gene_FPKM_df = gene_FPKM_df[mask]\n",
    "    \n",
    "    gene_FPKM_df = gene_FPKM_df.sort_values(length_col, ascending=False)\n",
    "    gene_FPKM_df = gene_FPKM_df.sort_values('FPKM', ascending=False)\n",
    "    \n",
    "    if len(gene_FPKM_df)==0:\n",
    "        print('Failed finding: ', gid)\n",
    "        Type[i].value= 'Failed'\n",
    "        Transcript_IDs[i].value = ''\n",
    "        Gene_IDs[i].value = ''\n",
    "        tid_fpkm = ''\n",
    "        FPKMs[i].value = ''\n",
    "        lengths[i].value = ''\n",
    "        Descriptions[i].value = ''\n",
    "#         Genes[i].value = ''\n",
    "        continue\n",
    "    tid = str(gene_FPKM_df[tid_col])\n",
    "    tid_FPKM_df = gene_FPKM_df.iloc[0]\n",
    "    Transcript_IDs[i].value = str(tid_FPKM_df[tid_col])\n",
    "    Gene_IDs[i].value = str(tid_FPKM_df[gid_col])\n",
    "    tid_fpkm = tid_FPKM_df['FPKM']\n",
    "    FPKMs[i].value = tid_FPKM_df['FPKM']\n",
    "    lengths[i].value = int(tid_FPKM_df[length_col])\n",
    "    Descriptions[i].value = str(tid_FPKM_df['WikiGene description'])\n",
    "    Genes[i].value = str(tid_FPKM_df[gene_name_col])\n",
    "    Type[i].value= 'Passed'\n",
    "    if tid_fpkm==0:\n",
    "        I = 0\n",
    "    else:\n",
    "        I = tid_FPKM_df['FPKM']/np.sum(g_annotation['FPKM'])\n",
    "    Isoforms[i].value = I\n",
    "    completed.append(gene)\n",
    "expression_worksheet.update_cells(Transcript_IDs)\n",
    "expression_worksheet.update_cells(FPKMs)\n",
    "expression_worksheet.update_cells(lengths)\n",
    "#expression_worksheet.update_cells(Genes)\n",
    "expression_worksheet.update_cells(Gene_IDs)\n",
    "expression_worksheet.update_cells(Isoforms)\n",
    "expression_worksheet.update_cells(Type)\n",
    "expression_worksheet.update_cells(Descriptions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1NujfUqy1FliV-4YIj0HEoiNkM9X9YzAhE73QRiFZqoY',\n",
       " 'updatedRange': 'Sheet1!J2:J473',\n",
       " 'updatedRows': 472,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 472}"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_worksheet = client.open(\"MAPK_genes\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)\n",
    "\n",
    "Transcript_IDs = expression_worksheet.range('G2:G'+str(len(expression_df)+1))\n",
    "Gene_IDs = expression_worksheet.range('F2:F'+str(len(expression_df)+1))\n",
    "FPKMs = expression_worksheet.range('C2:C'+str(len(expression_df)+1))\n",
    "lengths = expression_worksheet.range('D2:D'+str(len(expression_df)+1))\n",
    "Descriptions = expression_worksheet.range('B2:B'+str(len(expression_df)+1))\n",
    "Genes = expression_worksheet.range('A2:A'+str(len(expression_df)+1))\n",
    "Isoforms = expression_worksheet.range('I2:I'+str(len(expression_df)+1))\n",
    "Type = expression_worksheet.range('J2:J'+str(len(expression_df)+1))\n",
    "\n",
    "f = '/bigstore/binfo/mouse/3T3/GSM2609311_cufflinks/isoforms.fpkm_tracking'\n",
    "ReadsPerGene = pd.read_csv(f,sep='\\t')\n",
    "ReadsPerGene.index=ReadsPerGene.tracking_id\n",
    "fpkm = ReadsPerGene\n",
    "\n",
    "annotation = pd.read_csv('/bigstore/binfo/mouse/mart_export_25Nov2019.txt',sep='\\t')\n",
    "\n",
    "flag = 'tid'\n",
    "FPKM_gene_name_col = 'gene_short_name'#'Gene name'\n",
    "FPKM_gid_col = 'gene_id'#'Gene stable ID'\n",
    "FPKM_tid_col = 'tracking_id'#'Transcript stable ID'\n",
    "tid_col = 'Transcript stable ID'\n",
    "gid_col = 'Gene stable ID'\n",
    "ttype_col = 'Transcript type'\n",
    "gene_name_col = 'Gene name'\n",
    "length_col = 'Transcript length (including UTRs and CDS)'\n",
    "gene_description_col = 'Gene description'\n",
    "FPKM_FPKM_col = 'FPKM'#'IFNAR-0_tot.bam'\n",
    "completed = []\n",
    "\n",
    "for i,cell in enumerate(Transcript_IDs):\n",
    "    gene = cell.value\n",
    "    if gene == '':\n",
    "        continue\n",
    "    tid_fpkm_df = fpkm[fpkm[FPKM_tid_col]==gene].iloc[0]\n",
    "#     Transcript_IDs[i].value = str(tid_FPKM_df[tid_col])\n",
    "#     Gene_IDs[i].value = str(tid_FPKM_df[gid_col])\n",
    "    tid_fpkm = tid_FPKM_df['FPKM']\n",
    "    FPKMs[i].value = tid_FPKM_df['FPKM']\n",
    "#     lengths[i].value = int(tid_FPKM_df[length_col])\n",
    "#     Descriptions[i].value = str(tid_FPKM_df['WikiGene description'])\n",
    "    Genes[i].value = str(tid_FPKM_df[gene_name_col])\n",
    "    Type[i].value= 'Passed'\n",
    "    if tid_fpkm==0:\n",
    "        I = 0\n",
    "    else:\n",
    "        I = tid_FPKM_df['FPKM']/np.sum(g_annotation['FPKM'])\n",
    "    Isoforms[i].value = I\n",
    "    completed.append(gene)\n",
    "# expression_worksheet.update_cells(Transcript_IDs)\n",
    "expression_worksheet.update_cells(FPKMs)\n",
    "# expression_worksheet.update_cells(lengths)\n",
    "#expression_worksheet.update_cells(Genes)\n",
    "# expression_worksheet.update_cells(Gene_IDs)\n",
    "expression_worksheet.update_cells(Isoforms)\n",
    "expression_worksheet.update_cells(Type)\n",
    "# expression_worksheet.update_cells(Descriptions)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1NujfUqy1FliV-4YIj0HEoiNkM9X9YzAhE73QRiFZqoY',\n",
       " 'updatedRange': 'Sheet1!K2:K472',\n",
       " 'updatedRows': 471,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 471}"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_worksheet = client.open(\"MAPK_genes\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)\n",
    "Status = expression_worksheet.range('K2:K'+str(len(expression_df)+1))\n",
    "Genes = expression_worksheet.range('A2:A'+str(len(expression_df)+1))\n",
    "completed = []\n",
    "for i,cell in enumerate(Genes):\n",
    "    gene = cell.value\n",
    "    if gene in completed:\n",
    "        Status[i].value = 'Duplicate'\n",
    "    else:\n",
    "        Status[i].value = 'Passed'\n",
    "        completed.append(gene)\n",
    "expression_worksheet.update_cells(Status)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['RS0095_cy5',\n",
       " 'RS0109_cy5',\n",
       " 'RS0175_cy5',\n",
       " 'RS0237_cy5',\n",
       " 'RS0307_cy5',\n",
       " 'RS0332_cy5',\n",
       " 'RS0384_atto565',\n",
       " 'RS0406_atto565',\n",
       " 'RS0451_atto565',\n",
       " 'RS0468_atto565',\n",
       " 'RS0548_atto565',\n",
       " 'RS64.0_atto565',\n",
       " 'RS156.0_alexa488',\n",
       " 'RS278.0_alexa488',\n",
       " 'RS313.0_alexa488',\n",
       " 'RS643.0_alexa488',\n",
       " 'RS740.0_alexa488',\n",
       " 'RS810.0_alexa488',\n",
       " 'RSN9927.0_cy5',\n",
       " 'RSN2336.0_cy5',\n",
       " 'RSN1807.0_cy5',\n",
       " 'RSN4287.0_atto565',\n",
       " 'RSN1252.0_atto565',\n",
       " 'RSN9535.0_atto565']"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f = '/bigstore/binfo/mouse/24bit_disulfide_readouts.fasta'\n",
    "from Bio import SeqIO\n",
    "readout_names = []\n",
    "with open(f) as fasta_file:  # Will close handle cleanly\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        readout_names.append(seq_record.name)\n",
    "readout_names\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1393.92 220.79\n",
      "1141.86 185.47\n",
      "975.81 135.75\n",
      "960.19 160.38\n",
      "948.94 104.25\n",
      "920.6 129.2\n",
      "916.73 129.12\n",
      "914.57 99.2\n",
      "907.98 121.59\n",
      "904.87 115.62\n",
      "902.52 133.0\n",
      "893.06 128.33\n",
      "892.62 124.73\n",
      "884.57 96.3\n",
      "871.33 95.55\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-20-7132935a0819>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     25\u001b[0m     \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfpkm_barcodes\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     26\u001b[0m         \u001b[0mfpkm_barcodes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mfpkm_barcodes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mtemp_fpkm\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 27\u001b[0;31m     \u001b[0mtemp_std\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmax\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msum\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mfpkm_barcodes\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     28\u001b[0m     \u001b[0;32mif\u001b[0m \u001b[0mtemp_std\u001b[0m\u001b[0;34m<\u001b[0m\u001b[0mstd\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     29\u001b[0m         \u001b[0mstd\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtemp_std\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/numpy/core/fromnumeric.py\u001b[0m in \u001b[0;36msum\u001b[0;34m(a, axis, dtype, out, keepdims, initial)\u001b[0m\n\u001b[1;32m   2074\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2075\u001b[0m     return _wrapreduction(a, np.add, 'sum', axis, dtype, out, keepdims=keepdims,\n\u001b[0;32m-> 2076\u001b[0;31m                           initial=initial)\n\u001b[0m\u001b[1;32m   2077\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2078\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/numpy/core/fromnumeric.py\u001b[0m in \u001b[0;36m_wrapreduction\u001b[0;34m(obj, ufunc, method, axis, dtype, out, **kwargs)\u001b[0m\n\u001b[1;32m     82\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mreduction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mpasskwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     83\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 84\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mreduction\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mpasskwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     85\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     86\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mufunc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreduce\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mobj\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mout\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m**\u001b[0m\u001b[0mpasskwargs\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mstat_func\u001b[0;34m(self, axis, skipna, level, numeric_only, min_count, **kwargs)\u001b[0m\n\u001b[1;32m   9591\u001b[0m                                       skipna=skipna, min_count=min_count)\n\u001b[1;32m   9592\u001b[0m         return self._reduce(f, name, axis=axis, skipna=skipna,\n\u001b[0;32m-> 9593\u001b[0;31m                             numeric_only=numeric_only, min_count=min_count)\n\u001b[0m\u001b[1;32m   9594\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   9595\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mset_function_name\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mstat_func\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mcls\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_reduce\u001b[0;34m(self, op, name, axis, skipna, numeric_only, filter_type, **kwds)\u001b[0m\n\u001b[1;32m   6863\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mnumeric_only\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6864\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 6865\u001b[0;31m                 \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   6866\u001b[0m                 \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   6867\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mvalues\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   4630\u001b[0m         \u001b[0mpandas\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mDataFrame\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m \u001b[0;34m:\u001b[0m \u001b[0mRetrieving\u001b[0m \u001b[0mthe\u001b[0m \u001b[0mcolumn\u001b[0m \u001b[0mnames\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4631\u001b[0m         \"\"\"\n\u001b[0;32m-> 4632\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_consolidate_inplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4633\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_array\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtranspose\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_AXIS_REVERSED\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4634\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_consolidate_inplace\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   4437\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconsolidate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4438\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4439\u001b[0;31m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_protect_consolidate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4440\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4441\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_consolidate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mFalse\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_protect_consolidate\u001b[0;34m(self, f)\u001b[0m\n\u001b[1;32m   4426\u001b[0m         \"\"\"\n\u001b[1;32m   4427\u001b[0m         \u001b[0mblocks_before\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4428\u001b[0;31m         \u001b[0mresult\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4429\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m!=\u001b[0m \u001b[0mblocks_before\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4430\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_clear_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36mf\u001b[0;34m()\u001b[0m\n\u001b[1;32m   4435\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4436\u001b[0m         \u001b[0;32mdef\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4437\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mconsolidate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4438\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4439\u001b[0m         \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_protect_consolidate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mf\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mconsolidate\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   4096\u001b[0m         \u001b[0mbm\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m__class__\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0maxes\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4097\u001b[0m         \u001b[0mbm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_consolidated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mFalse\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4098\u001b[0;31m         \u001b[0mbm\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_consolidate_inplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4099\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mbm\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4100\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36m_consolidate_inplace\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m   4101\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_consolidate_inplace\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4102\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_consolidated\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4103\u001b[0;31m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtuple\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0m_consolidate\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4104\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_is_consolidated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4105\u001b[0m             \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_known_consolidated\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36m_consolidate\u001b[0;34m(blocks)\u001b[0m\n\u001b[1;32m   5062\u001b[0m     \u001b[0;31m# sort by _can_consolidate, dtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5063\u001b[0m     \u001b[0mgkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_consolidate_key\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5064\u001b[0;31m     \u001b[0mgrouper\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitertools\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5065\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5066\u001b[0m     \u001b[0mnew_blocks\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36m<lambda>\u001b[0;34m(x)\u001b[0m\n\u001b[1;32m   5061\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5062\u001b[0m     \u001b[0;31m# sort by _can_consolidate, dtype\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 5063\u001b[0;31m     \u001b[0mgkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mlambda\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m:\u001b[0m \u001b[0mx\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_consolidate_key\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   5064\u001b[0m     \u001b[0mgrouper\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mitertools\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msorted\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mblocks\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mgkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mgkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   5065\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36m_consolidate_key\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    166\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    167\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_consolidate_key\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 168\u001b[0;31m         \u001b[0;32mreturn\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_can_consolidate\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    169\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    170\u001b[0m     \u001b[0;34m@\u001b[0m\u001b[0mproperty\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/numpy/core/_dtype.py\u001b[0m in \u001b[0;36m_name_get\u001b[0;34m(dtype)\u001b[0m\n\u001b[1;32m    332\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    333\u001b[0m     \u001b[0;31m# append bit counts to str, unicode, and void\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 334\u001b[0;31m     \u001b[0;32mif\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0missubdtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mflexible\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0m_isunsized\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    335\u001b[0m         \u001b[0mname\u001b[0m \u001b[0;34m+=\u001b[0m \u001b[0;34m\"{}\"\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mformat\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mdtype\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitemsize\u001b[0m \u001b[0;34m*\u001b[0m \u001b[0;36m8\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    336\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/numpy/core/numerictypes.py\u001b[0m in \u001b[0;36missubdtype\u001b[0;34m(arg1, arg2)\u001b[0m\n\u001b[1;32m    363\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    364\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 365\u001b[0;31m \u001b[0;34m@\u001b[0m\u001b[0mset_module\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'numpy'\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    366\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0missubdtype\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0marg1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0marg2\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    367\u001b[0m     \"\"\"\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "expression_worksheet = client.open(\"MAPK_genes\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)\n",
    "import random\n",
    "barcodes = pd.read_csv('/home/zach/PythonRepos/PySpots/hybescope_config/MHD4_24bit_472cwords.csv',header=None)\n",
    "# del fpkm\n",
    "fpkm = expression_df.FPKM.copy()\n",
    "fpkm.index = expression_df.Gene\n",
    "i=0\n",
    "while len(fpkm)<len(barcodes):\n",
    "    key = 'blank_'+str(i)\n",
    "    fpkm[key] = 0*fpkm.iloc[0]\n",
    "    i+=1\n",
    "fpkm = pd.DataFrame(fpkm)\n",
    "fpkm = fpkm.sort_index()\n",
    "genes = list(fpkm.index)\n",
    "s = 0\n",
    "std = 10**6\n",
    "while s==0:\n",
    "    temp_barcodes = barcodes.copy()\n",
    "    temp_barcodes.index = random.sample(genes,len(genes))\n",
    "    temp_barcodes = temp_barcodes.sort_index()\n",
    "    temp_fpkm = list(fpkm['FPKM'])\n",
    "    fpkm_barcodes = temp_barcodes.copy()\n",
    "    for i in fpkm_barcodes.columns:\n",
    "        fpkm_barcodes[i] = fpkm_barcodes[i]*temp_fpkm\n",
    "    temp_std = np.max(np.sum(fpkm_barcodes,axis=0))\n",
    "    if temp_std<std:\n",
    "        std = temp_std\n",
    "        print(round(std,2),round(np.std(np.sum(fpkm_barcodes,axis=0)),2))\n",
    "        master_barcodes = temp_barcodes.copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0     665.07\n",
       "1     815.56\n",
       "2     863.80\n",
       "3     745.39\n",
       "4     803.67\n",
       "5     497.17\n",
       "6     599.20\n",
       "7     814.77\n",
       "8     754.67\n",
       "9     719.83\n",
       "10    695.19\n",
       "11    794.16\n",
       "12    730.01\n",
       "13    800.67\n",
       "14    679.30\n",
       "15    871.33\n",
       "16    864.82\n",
       "17    810.46\n",
       "18    859.70\n",
       "19    759.18\n",
       "20    784.76\n",
       "21    688.12\n",
       "22    567.69\n",
       "23    672.96\n",
       "dtype: float64"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "temp_barcodes = master_barcodes.copy()\n",
    "temp_fpkm = list(fpkm['FPKM'])\n",
    "fpkm_barcodes = temp_barcodes.copy()\n",
    "for i in fpkm_barcodes.columns:\n",
    "    fpkm_barcodes[i] = fpkm_barcodes[i]*temp_fpkm\n",
    "np.sum(fpkm_barcodes,axis=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1NujfUqy1FliV-4YIj0HEoiNkM9X9YzAhE73QRiFZqoY',\n",
       " 'updatedRange': 'Sheet1!H2:H473',\n",
       " 'updatedRows': 472,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 472}"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_worksheet = client.open(\"MAPK_genes\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)\n",
    "Barcodes = expression_worksheet.range('H2:H'+str(len(master_barcodes)+1))\n",
    "Genes = expression_worksheet.range('A2:A'+str(len(master_barcodes)+1))\n",
    "b = 0\n",
    "genes = []\n",
    "for i,cell in enumerate(Genes):\n",
    "    genes.append(cell.value)\n",
    "for i,cell in enumerate(Genes):\n",
    "    gene = cell.value\n",
    "    if gene=='':\n",
    "        b=0\n",
    "        gene = 'blank_'+str(b)\n",
    "        while gene in genes:\n",
    "            b+=1\n",
    "            gene = 'blank_'+str(b)\n",
    "    Barcodes[i].value = ''.join(str(i) for i in master_barcodes.loc[gene])\n",
    "    Genes[i].value = gene\n",
    "expression_worksheet.update_cells(Genes)\n",
    "expression_worksheet.update_cells(Barcodes)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['RS0095_cy5', 'RS0109_cy5', 'RS0175_cy5', 'RS0237_cy5', 'RS0307_cy5', 'RS0332_cy5', 'RS0384_atto565', 'RS0406_atto565', 'RS0451_atto565', 'RS0468_atto565', 'RS0548_atto565', 'RS64.0_atto565', 'RS156.0_alexa488', 'RS278.0_alexa488', 'RS313.0_alexa488', 'RS643.0_alexa488', 'RS740.0_alexa488', 'RS810.0_alexa488', 'RSN9927.0_cy5', 'RSN2336.0_cy5', 'RSN1807.0_cy5', 'RSN4287.0_atto565', 'RSN1252.0_atto565', 'RSN9535.0_atto565']\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "[('Acta2', 'ENSMUST00000039631', '001000000000000001011000'),\n",
       " ('Anxa2', 'ENSMUST00000034756', '000000000000010110100000'),\n",
       " ('Atf4', 'ENSMUST00000109605', '000100010001100000000000'),\n",
       " ('Hspa9', 'ENSMUST00000025217', '010000001100001000000000'),\n",
       " ('Thbs1', 'ENSMUST00000039559', '100000110100000000000000'),\n",
       " ('Cdc42', 'ENSMUST00000051477', '000100010000010000000100'),\n",
       " ('Gnai2', 'ENSMUST00000055704', '000100000000000010001001'),\n",
       " ('Akt1', 'ENSMUST00000001780', '100000001100000100000000'),\n",
       " ('Ctnnb1', 'ENSMUST00000007130', '110001000001000000000000'),\n",
       " ('Rac1', 'ENSMUST00000100489', '010000001010000100000000'),\n",
       " ('Eif3a', 'ENSMUST00000025955', '000010000001100000100000'),\n",
       " ('Pdgfra', 'ENSMUST00000000476', '000000100000101000100000'),\n",
       " ('Sdha', 'ENSMUST00000022062', '001100010000000000100000'),\n",
       " ('Grb10', 'ENSMUST00000093321', '010000010010000000000100'),\n",
       " ('Sdc4', 'ENSMUST00000017153', '000010000010001000100000'),\n",
       " ('Reep5', 'ENSMUST00000006027', '000010000001000001000100'),\n",
       " ('Rcc2', 'ENSMUST00000038893', '100100000000100000000001'),\n",
       " ('Trp53', 'ENSMUST00000108658', '000000000000010000011001'),\n",
       " ('Lbh', 'ENSMUST00000024857', '000000010000000100000011'),\n",
       " ('Stom', 'ENSMUST00000028241', '100000000000001000100010'),\n",
       " ('Csf1', 'ENSMUST00000014743', '000000100100000010000100'),\n",
       " ('Jak1', 'ENSMUST00000102781', '001000001010000000000100'),\n",
       " ('Actr3', 'ENSMUST00000178474', '001000010000011000000000'),\n",
       " ('F2r', 'ENSMUST00000059193', '010000000010000010001000'),\n",
       " ('Tnfrsf1a', 'ENSMUST00000032491', '000000010000100000000110'),\n",
       " ('Lrp1', 'ENSMUST00000049149', '000010000000000110000001'),\n",
       " ('Atp6ap1', 'ENSMUST00000019231', '010100000100000001000000'),\n",
       " ('Map2k3', 'ENSMUST00000019076', '100000000001101000000000'),\n",
       " ('Mat2a', 'ENSMUST00000059472', '000000100110000000001000'),\n",
       " ('Mta2', 'ENSMUST00000096240', '000001001101000000000000'),\n",
       " ('Kars', 'ENSMUST00000034426', '100100000010001000000000'),\n",
       " ('Pmp22', 'ENSMUST00000018361', '010010001000100000000000'),\n",
       " ('Aqp1', 'ENSMUST00000004774', '010010000000010000001000'),\n",
       " ('Prkaca', 'ENSMUST00000005606', '010000000000001100100000'),\n",
       " ('Mapkapk2', 'ENSMUST00000016672', '000010000000000010110000'),\n",
       " ('Icam1', 'ENSMUST00000086399', '000010000011000000010000'),\n",
       " ('Pdgfrb', 'ENSMUST00000025522', '010001000000000100000100'),\n",
       " ('Ppp3r1', 'ENSMUST00000102880', '000100001000000010100000'),\n",
       " ('Ltbr', 'ENSMUST00000032489', '000000010100100000000001'),\n",
       " ('Ilk', 'ENSMUST00000163389', '001000000100000000010010'),\n",
       " ('Pla2g4a', 'ENSMUST00000070200', '010000100000000001000100'),\n",
       " ('Jun', 'ENSMUST00000107094', '001000000001000000000011'),\n",
       " ('Iqgap1', 'ENSMUST00000167377', '000000100001001000010000'),\n",
       " ('Errfi1', 'ENSMUST00000030811', '001010010000000000000100'),\n",
       " ('Ruvbl1', 'ENSMUST00000032165', '100010000000000010000010'),\n",
       " ('Gpc1', 'ENSMUST00000045970', '010001000000000011000000'),\n",
       " ('Emilin1', 'ENSMUST00000031055', '000110000000010010000000'),\n",
       " ('Crk', 'ENSMUST00000017920', '000010100000010000000010'),\n",
       " ('Cdkn1a', 'ENSMUST00000023829', '000001100000000001010000'),\n",
       " ('Klf6', 'ENSMUST00000000080', '010010000001000010000000'),\n",
       " ('Rps6ka4', 'ENSMUST00000170516', '001010000000000001000010'),\n",
       " ('Fosl2', 'ENSMUST00000031017', '000001001010000000100000'),\n",
       " ('Btg1', 'ENSMUST00000038377', '100010001000000001000000'),\n",
       " ('Tgfbr2', 'ENSMUST00000035014', '000000001000011000100000'),\n",
       " ('Casp3', 'ENSMUST00000211115', '000010001001001000000000'),\n",
       " ('Ipo7', 'ENSMUST00000084731', '010000100000000000010001'),\n",
       " ('Rab14', 'ENSMUST00000028238', '000000001011000000000010'),\n",
       " ('Flnb', 'ENSMUST00000052678', '100001000010000000000100'),\n",
       " ('Tmem109', 'ENSMUST00000038128', '010000000101000000100000'),\n",
       " ('Flnc', 'ENSMUST00000065090', '010000100010000000100000'),\n",
       " ('Nr4a1', 'ENSMUST00000023779', '000000001100100000001000'),\n",
       " ('Glg1', 'ENSMUST00000169020', '000001000110000000000001'),\n",
       " ('Lemd2', 'ENSMUST00000055117', '010000001001000000000100'),\n",
       " ('Ccna2', 'ENSMUST00000029270', '100000100000100010000000'),\n",
       " ('Aldh9a1', 'ENSMUST00000028004', '000000000000110001000010'),\n",
       " ('Nfkb2', 'ENSMUST00000073116', '000000000101000100001000'),\n",
       " ('Lmnb1', 'ENSMUST00000025486', '000000001000000001000011'),\n",
       " ('Api5', 'ENSMUST00000028617', '100000001001000000000001'),\n",
       " ('Srf', 'ENSMUST00000015749', '011000100000000100000000'),\n",
       " ('Atp6ap2', 'ENSMUST00000033313', '000000000110110000000000'),\n",
       " ('Cdc42se1', 'ENSMUST00000053872', '000011000000000000000101'),\n",
       " ('Dag1', 'ENSMUST00000191899', '000010000100010000100000'),\n",
       " ('Fer', 'ENSMUST00000000129', '000011000100000100000000'),\n",
       " ('Ptgs2', 'ENSMUST00000035065', '000000000001001000100001'),\n",
       " ('Dvl2', 'ENSMUST00000019362', '000000011010000000010000'),\n",
       " ('Gna12', 'ENSMUST00000000153', '000000000000001100010001'),\n",
       " ('Rela', 'ENSMUST00000025867', '000001000000100000001100'),\n",
       " ('Mapk14', 'ENSMUST00000062694', '010000010000100100000000'),\n",
       " ('Pik3r5', 'ENSMUST00000021283', '000100000000000001100001'),\n",
       " ('Hmgcr', 'ENSMUST00000022176', '000100001000100100000000'),\n",
       " ('Mapk1', 'ENSMUST00000069107', '010000000001011000000000'),\n",
       " ('Coro1c', 'ENSMUST00000004646', '101000100000000000100000'),\n",
       " ('Arhgap19', 'ENSMUST00000026150', '000100001000000000000110'),\n",
       " ('Il1r1', 'ENSMUST00000114795', '000010000010000000001010'),\n",
       " ('Pik3r1', 'ENSMUST00000055518', '000011010000100000000000'),\n",
       " ('Tcp11l2', 'ENSMUST00000020223', '001000010000000110000000'),\n",
       " ('Sfrp1', 'ENSMUST00000033952', '000000000000010011010000'),\n",
       " ('Tgfb3', 'ENSMUST00000003687', '000000000000100100010100'),\n",
       " ('Ski', 'ENSMUST00000030917', '011000000011000000000000'),\n",
       " ('Grb2', 'ENSMUST00000021090', '000001000000000100010010'),\n",
       " ('Tgfbr1', 'ENSMUST00000007757', '001001000000010000000100'),\n",
       " ('Hdac3', 'ENSMUST00000043498', '000011000001010000000000'),\n",
       " ('Vegfd', 'ENSMUST00000033751', '000100000010000000000101'),\n",
       " ('Map2k1', 'ENSMUST00000005066', '000000000000010100000101'),\n",
       " ('Axin1', 'ENSMUST00000074370', '001000000000100001000100'),\n",
       " ('Prkacb', 'ENSMUST00000102515', '001001010000000000000010'),\n",
       " ('Inppl1', 'ENSMUST00000165052', '100110000000000000001000'),\n",
       " ('Serpine1', 'ENSMUST00000041388', '000000000000100010010001'),\n",
       " ('Tgfbr3', 'ENSMUST00000031224', '001000100000100000000010'),\n",
       " ('Naa40', 'ENSMUST00000025675', '001010000000100100000000'),\n",
       " ('Il1rap', 'ENSMUST00000023156', '000010011000010000000000'),\n",
       " ('C3', 'ENSMUST00000024988', '100000000001000010100000'),\n",
       " ('Raf1', 'ENSMUST00000000451', '000011000000000001001000'),\n",
       " ('Zmiz1', 'ENSMUST00000162645', '000000011001000100000000'),\n",
       " ('Itgav', 'ENSMUST00000028499', '010000000001100000000001'),\n",
       " ('Emc7', 'ENSMUST00000069747', '010010000000000001100000'),\n",
       " ('Zmpste24', 'ENSMUST00000058754', '000011100000000010000000'),\n",
       " ('Mknk2', 'ENSMUST00000003433', '001000001000000100000001'),\n",
       " ('Gas6', 'ENSMUST00000033828', '000110000000100000010000'),\n",
       " ('Rasa1', 'ENSMUST00000109552', '000000101010000010000000'),\n",
       " ('Pak2', 'ENSMUST00000023467', '101010000000000000010000'),\n",
       " ('Nras', 'ENSMUST00000029445', '000000110000000010010000'),\n",
       " ('Glipr2', 'ENSMUST00000030202', '000100000111000000000000'),\n",
       " ('Cx3cl1', 'ENSMUST00000034230', '000000010000010001001000'),\n",
       " ('Dusp1', 'ENSMUST00000025025', '000001000001000001000001'),\n",
       " ('Grk2', 'ENSMUST00000088737', '000100000000000101000100'),\n",
       " ('Map3k3', 'ENSMUST00000002044', '011000000000001000000100'),\n",
       " ('Ikbke', 'ENSMUST00000062108', '010110000000000100000000'),\n",
       " ('P2rx7', 'ENSMUST00000100737', '000100000100000100000001'),\n",
       " ('Spry2', 'ENSMUST00000022709', '001000000000000010100001'),\n",
       " ('Foxm1', 'ENSMUST00000073316', '001000000000000010001010'),\n",
       " ('Hacd3', 'ENSMUST00000036615', '010000001000000000011000'),\n",
       " ('Dusp6', 'ENSMUST00000020118', '000100000010000000100010'),\n",
       " ('Stk3', 'ENSMUST00000018476', '110000000000100000000010'),\n",
       " ('Gba', 'ENSMUST00000077367', '000000100010001001000000'),\n",
       " ('Dusp7', 'ENSMUST00000172306', '000000010000000010100010'),\n",
       " ('E2f1', 'ENSMUST00000103145', '100010000010000000000001'),\n",
       " ('Kitl', 'ENSMUST00000105283', '001100100010000000000000'),\n",
       " ('Map3k12', 'ENSMUST00000023812', '000001000001100000000010'),\n",
       " ('Trim32', 'ENSMUST00000050850', '000100100001000010000000'),\n",
       " ('Map2k4', 'ENSMUST00000046963', '000001010000010010000000'),\n",
       " ('Efna5', 'ENSMUST00000076840', '000010001100000000000001'),\n",
       " ('Ndst1', 'ENSMUST00000169273', '100000010010100000000000'),\n",
       " ('Nf2', 'ENSMUST00000109910', '000100011100000000000000'),\n",
       " ('Cacng7', 'ENSMUST00000092891', '000000110000001000000010'),\n",
       " ('Eif2ak2', 'ENSMUST00000024884', '001101000000000001000000'),\n",
       " ('Zhx1', 'ENSMUST00000070143', '001000001000001000000010'),\n",
       " ('Hipk3', 'ENSMUST00000028600', '001000000000101000000001'),\n",
       " ('Ajuba', 'ENSMUST00000054487', '000000010000010000010010'),\n",
       " ('Myd88', 'ENSMUST00000035092', '001000000100000100000100'),\n",
       " ('Pik3r2', 'ENSMUST00000034296', '000000101000001000000100'),\n",
       " ('Smad4', 'ENSMUST00000025393', '000000000100001001010000'),\n",
       " ('Brap', 'ENSMUST00000031414', '000100100000010001000000'),\n",
       " ('Ackr3', 'ENSMUST00000065587', '000000011000000000000101'),\n",
       " ('Flcn', 'ENSMUST00000091246', '101000000000010010000000'),\n",
       " ('Nbr1', 'ENSMUST00000107213', '000000000000001010011000'),\n",
       " ('Fgfrl1', 'ENSMUST00000013633', '000000100010100000000100'),\n",
       " ('Fgf10', 'ENSMUST00000022246', '010001000000010000010000'),\n",
       " ('Ezr', 'ENSMUST00000064234', '000000000100010010001000'),\n",
       " ('Igf1r', 'ENSMUST00000005671', '000000000000000100100110'),\n",
       " ('Epha2', 'ENSMUST00000006614', '001010100100000000000000'),\n",
       " ('Crkl', 'ENSMUST00000006293', '001000000000010101000000'),\n",
       " ('Adam17', 'ENSMUST00000064536', '001000000000001100001000'),\n",
       " ('Fndc3b', 'ENSMUST00000046157', '000000100010010000010000'),\n",
       " ('Klf4', 'ENSMUST00000107619', '010000000000000001010010'),\n",
       " ('Map3k10', 'ENSMUST00000036453', '000000010101001000000000'),\n",
       " ('Ino80c', 'ENSMUST00000153360', '011000001000000001000000'),\n",
       " ('Dusp5', 'ENSMUST00000038287', '000010010000000100001000'),\n",
       " ('Shcbp1', 'ENSMUST00000022945', '001000000010000000001001'),\n",
       " ('Frs2', 'ENSMUST00000020381', '000100000000011000000001'),\n",
       " ('Cdc25b', 'ENSMUST00000079857', '100100000001000000000100'),\n",
       " ('Cdk5rap3', 'ENSMUST00000103152', '000000100000000010000011'),\n",
       " ('Nfatc3', 'ENSMUST00000109308', '000010001000000000100100'),\n",
       " ('Map4k3', 'ENSMUST00000025089', '000000000010000010010010'),\n",
       " ('Cdc6', 'ENSMUST00000093937', '010100000000001010000000'),\n",
       " ('Igfbp4', 'ENSMUST00000017637', '110000010000000001000000'),\n",
       " ('Araf', 'ENSMUST00000001155', '000000010001000011000000'),\n",
       " ('Mapkapk3', 'ENSMUST00000035194', '001100000100000010000000'),\n",
       " ('Leo1', 'ENSMUST00000048937', '000000010010000001000010'),\n",
       " ('Dusp4', 'ENSMUST00000033930', '000101000000001100000000'),\n",
       " ('Bdnf', 'ENSMUST00000111045', '100010000101000000000000'),\n",
       " ('Fam83d', 'ENSMUST00000029183', '000000001000001101000000'),\n",
       " ('Chuk', 'ENSMUST00000026217', '000001000000001000010100'),\n",
       " ('Arid5b', 'ENSMUST00000219238', '000000000001010000100010'),\n",
       " ('Schip1', 'ENSMUST00000182719', '000100000010000100001000'),\n",
       " ('Nr1d1', 'ENSMUST00000064941', '010000100001000000000010'),\n",
       " ('Map2k7', 'ENSMUST00000062686', '000000000011000010000100'),\n",
       " ('Mapk8ip1', 'ENSMUST00000050312', '000000111000000001000000'),\n",
       " ('Nbn', 'ENSMUST00000029879', '100000001000001010000000'),\n",
       " ('Wdr18', 'ENSMUST00000045247', '000000000000110100001000'),\n",
       " ('Gclc', 'ENSMUST00000034905', '000000100000000100011000'),\n",
       " ('Iqgap3', 'ENSMUST00000071812', '001000010110000000000000'),\n",
       " ('Smad1', 'ENSMUST00000066091', '000000100001000001100000'),\n",
       " ('Fbrs', 'ENSMUST00000205432', '000000000011000100000001'),\n",
       " ('Erbb2', 'ENSMUST00000058295', '000001100001000000001000'),\n",
       " ('Pik3ca', 'ENSMUST00000108243', '000010000000010001000001'),\n",
       " ('Tab1', 'ENSMUST00000023050', '000001010100000001000000'),\n",
       " ('Dhx33', 'ENSMUST00000108527', '000001100000100000000001'),\n",
       " ('Med1', 'ENSMUST00000107545', '000010000000001011000000'),\n",
       " ('Kras', 'ENSMUST00000032399', '000001001000000001000100'),\n",
       " ('Elk1', 'ENSMUST00000009550', '001100000000010000000010'),\n",
       " ('Klhdc10', 'ENSMUST00000068259', '000000000000100000100101'),\n",
       " ('Cflar', 'ENSMUST00000114313', '000010000000100000001001'),\n",
       " ('Runx2', 'ENSMUST00000113572', '000000000000010000001110'),\n",
       " ('Fam20c', 'ENSMUST00000026972', '100000000000010001100000'),\n",
       " ('Mfhas1', 'ENSMUST00000037666', '100000000010010000000010'),\n",
       " ('Gsk3b', 'ENSMUST00000023507', '000000001100000000010100'),\n",
       " ('Abl2', 'ENSMUST00000166172', '000100000000101000000010'),\n",
       " ('Fntb', 'ENSMUST00000041008', '100000010001010000000000'),\n",
       " ('Mknk1', 'ENSMUST00000019677', '010010010000001000000000'),\n",
       " ('Dgat1', 'ENSMUST00000023214', '000110000001000000000001'),\n",
       " ('B4galt7', 'ENSMUST00000064701', '000100000100100000100000'),\n",
       " ('Pak3', 'ENSMUST00000172330', '000000000110000100010000'),\n",
       " ('Itch', 'ENSMUST00000029126', '000100000010000001010000'),\n",
       " ('Akap12', 'ENSMUST00000045730', '000010000000101000000100'),\n",
       " ('Map3k7', 'ENSMUST00000080933', '000001000010100000010000'),\n",
       " ('Relb', 'ENSMUST00000094762', '100000100000000101000000'),\n",
       " ('Fem1a', 'ENSMUST00000060253', '000000000100101100000000'),\n",
       " ('Bmper', 'ENSMUST00000071982', '000001000100001000000010'),\n",
       " ('Appl2', 'ENSMUST00000020500', '100000010000000100000100'),\n",
       " ('Cnksr3', 'ENSMUST00000015346', '001000100000010000000001'),\n",
       " ('Rps6ka3', 'ENSMUST00000033671', '000000101100000000000010'),\n",
       " ('Pml', 'ENSMUST00000085673', '100100000000000110000000'),\n",
       " ('Apln', 'ENSMUST00000039026', '000000001001000000101000'),\n",
       " ('Map2k5', 'ENSMUST00000034920', '000110010000000001000000'),\n",
       " ('Mmp9', 'ENSMUST00000017881', '100100001000010000000000'),\n",
       " ('Taok2', 'ENSMUST00000117394', '100011000000001000000000'),\n",
       " ('Ikbkb', 'ENSMUST00000033939', '110000000000010000000001'),\n",
       " ('Map2k6', 'ENSMUST00000020949', '100000100000011000000000'),\n",
       " ('Setx', 'ENSMUST00000061578', '000100000001000000010010'),\n",
       " ('Mapk9', 'ENSMUST00000020634', '000100010010000010000000'),\n",
       " ('Jak2', 'ENSMUST00000065796', '100000010000000010001000'),\n",
       " ('Itpkb', 'ENSMUST00000070181', '100000000000000011000100'),\n",
       " ('Wapl', 'ENSMUST00000048263', '000000000100010000000011'),\n",
       " ('Igf2', 'ENSMUST00000000033', '000100000000001000100100'),\n",
       " ('Sos1', 'ENSMUST00000068714', '000000000000001001000110'),\n",
       " ('Dusp10', 'ENSMUST00000048655', '101000000000000000000101'),\n",
       " ('Rps6ka2', 'ENSMUST00000024575', '000010000000000101010000'),\n",
       " ('Cspg4', 'ENSMUST00000035661', '000000101001100000000000'),\n",
       " ('Dab2ip', 'ENSMUST00000112983', '010000000000010100000010'),\n",
       " ('Map3k11', 'ENSMUST00000004156', '000100010000001000001000'),\n",
       " ('Lrrk2', 'ENSMUST00000060642', '000000010001000000001001'),\n",
       " ('Ctc1', 'ENSMUST00000116359', '001000000001001010000000'),\n",
       " ('Taok1', 'ENSMUST00000058496', '010000000000000100001001'),\n",
       " ('Taok3', 'ENSMUST00000111978', '000000000011000001001000'),\n",
       " ('Ksr1', 'ENSMUST00000018478', '001010001000000010000000'),\n",
       " ('Morc3', 'ENSMUST00000202261', '100000001000100000000100'),\n",
       " ('Dstyk', 'ENSMUST00000045110', '000101010000000000000001'),\n",
       " ('Inpp5k', 'ENSMUST00000006286', '000000100000000000110010'),\n",
       " ('Creb3l1', 'ENSMUST00000028663', '000000010100000000011000'),\n",
       " ('Map3k14', 'ENSMUST00000021324', '010000010000000010000001'),\n",
       " ('Stk4', 'ENSMUST00000018353', '000100001000000001001000'),\n",
       " ('Map3k20', 'ENSMUST00000135469', '000001000000000000101001'),\n",
       " ('Ndrg4', 'ENSMUST00000080666', '000000001000010010000100'),\n",
       " ('Snrpd3', 'ENSMUST00000020397', '000000010010001100000000'),\n",
       " ('Ash1l', 'ENSMUST00000090933', '011000000000000000100010'),\n",
       " ('Ereg', 'ENSMUST00000031324', '000100000000100010000100'),\n",
       " ('Appl1', 'ENSMUST00000036570', '000001000000000001100010'),\n",
       " ('Mef2a', 'ENSMUST00000156690', '000000000010010000101000'),\n",
       " ('Mapk11', 'ENSMUST00000088823', '100001000000100100000000'),\n",
       " ('Hipk2', 'ENSMUST00000161779', '000000000010100000000011'),\n",
       " ('Sp2', 'ENSMUST00000107623', '000000010000000001010100'),\n",
       " ('Cacna1g', 'ENSMUST00000107790', '000000000000111000010000'),\n",
       " ('Tmem201', 'ENSMUST00000103208', '001000000010001000010000'),\n",
       " ('Mul1', 'ENSMUST00000044058', '001001000100000000001000'),\n",
       " ('Flrt2', 'ENSMUST00000057324', '000110000100000000000100'),\n",
       " ('Smoc2', 'ENSMUST00000024660', '000000010100000000100100'),\n",
       " ('Smad5', 'ENSMUST00000069557', '000000001000000100001100'),\n",
       " ('Parp10', 'ENSMUST00000075689', '000010010100000010000000'),\n",
       " ('Angpt1', 'ENSMUST00000022921', '000000000000001000001011'),\n",
       " ('Bag4', 'ENSMUST00000038498', '100000000001000001000010'),\n",
       " ('Bcar3', 'ENSMUST00000029766', '001000100000000011000000'),\n",
       " ('Dact1', 'ENSMUST00000061273', '000010000010000100000100'),\n",
       " ('Stk17b', 'ENSMUST00000027263', '000000000001100010001000'),\n",
       " ('Elk4', 'ENSMUST00000086556', '000000000000100110000010'),\n",
       " ('Cd14', 'ENSMUST00000061829', '100010000000010000000100'),\n",
       " ('Map3k15', 'ENSMUST00000033665', '000010100001000100000000'),\n",
       " ('Insr', 'ENSMUST00000091291', '101000000100001000000000'),\n",
       " ('Sod2', 'ENSMUST00000007012', '000010010000000000010001'),\n",
       " ('Fzd5', 'ENSMUST00000063982', '110000000000001000010000'),\n",
       " ('Braf', 'ENSMUST00000002487', '000000101000010100000000'),\n",
       " ('Nlk', 'ENSMUST00000142739', '000000000100001010000001'),\n",
       " ('Dusp3', 'ENSMUST00000003612', '000010000100100000000010'),\n",
       " ('Cacna1c', 'ENSMUST00000186889', '001001001000100000000000'),\n",
       " ('Il6ra', 'ENSMUST00000029559', '000001000001000110000000'),\n",
       " ('Ofd1', 'ENSMUST00000049501', '000100100100001000000000'),\n",
       " ('Spry4', 'ENSMUST00000025295', '000000100000001110000000'),\n",
       " ('Fgf2', 'ENSMUST00000200585', '000000000001010010000001'),\n",
       " ('Ddias', 'ENSMUST00000032877', '100000100011000000000000'),\n",
       " ('Bcl2l11', 'ENSMUST00000110341', '100000000100000010010000'),\n",
       " ('Garem1', 'ENSMUST00000049260', '001000000001000000100100'),\n",
       " ('Rasa2', 'ENSMUST00000034984', '000001010010000000001000'),\n",
       " ('Epha4', 'ENSMUST00000027451', '001100001001000000000000'),\n",
       " ('Xrcc2', 'ENSMUST00000030773', '000000000001001100000010'),\n",
       " ('Traf6', 'ENSMUST00000004949', '000001100010000000000010'),\n",
       " ('Mapkbp1', 'ENSMUST00000066058', '000000001001000010010000'),\n",
       " ('Cacna2d1', 'ENSMUST00000078272', '000010001000000100000010'),\n",
       " ('Cdk6', 'ENSMUST00000165117', '000001001000000010000001'),\n",
       " ('Gab1', 'ENSMUST00000034150', '010000000000110000000100'),\n",
       " ('Map3k2', 'ENSMUST00000096575', '010010000100000000010000'),\n",
       " ('Prkce', 'ENSMUST00000097275', '101000000000100000001000'),\n",
       " ('Fbxw7', 'ENSMUST00000107679', '000000110000110000000000'),\n",
       " ('Dusp9', 'ENSMUST00000019701', '000010000000000010001100'),\n",
       " ('Irak4', 'ENSMUST00000109248', '010001010000000000100000'),\n",
       " ('Dusp16', 'ENSMUST00000100857', '100001000000000000000011'),\n",
       " ('Fat4', 'ENSMUST00000061260', '000010000100001000001000'),\n",
       " ('Ednra', 'ENSMUST00000034029', '000100000100010000010000'),\n",
       " ('Trp53inp1', 'ENSMUST00000029865', '010001000000000000001010'),\n",
       " ('Gata4', 'ENSMUST00000118022', '011010000000000000000001'),\n",
       " ('Ern1', 'ENSMUST00000001059', '000000000000100101000001'),\n",
       " ('Arhgef5', 'ENSMUST00000031750', '001000000010100000100000'),\n",
       " ('Itgb3', 'ENSMUST00000021028', '000100100000000100000010'),\n",
       " ('Hspa1l', 'ENSMUST00000007248', '010000000000001001000001'),\n",
       " ('Nfatc1', 'ENSMUST00000078049', '000000000000000111001000'),\n",
       " ('Sos2', 'ENSMUST00000035773', '101000000010000001000000'),\n",
       " ('Sulf2', 'ENSMUST00000088086', '000100100000100000001000'),\n",
       " ('Lpar2', 'ENSMUST00000034325', '001000001000010000001000'),\n",
       " ('Fos', 'ENSMUST00000021674', '000000000010000000110001'),\n",
       " ('C1qtnf1', 'ENSMUST00000017590', '010000000010001000000010'),\n",
       " ('Dnajc27', 'ENSMUST00000020986', '000000110000000000001100'),\n",
       " ('Ankrd26', 'ENSMUST00000112830', '010000000000000010100100'),\n",
       " ('Fzd8', 'ENSMUST00000041080', '110000000000000000101000'),\n",
       " ('Fzd4', 'ENSMUST00000058755', '000000010010010000000001'),\n",
       " ('Prr5l', 'ENSMUST00000043845', '010100000000010000100000'),\n",
       " ('Gcnt2', 'ENSMUST00000067778', '000101000001000000100000'),\n",
       " ('Wdpcp', 'ENSMUST00000020568', '010000000001000101000000'),\n",
       " ('Prdm11', 'ENSMUST00000111272', '000000000010000011000001'),\n",
       " ('Dok4', 'ENSMUST00000046461', '000010000110000001000000'),\n",
       " ('Arrb1', 'ENSMUST00000098266', '000000100100100000010000'),\n",
       " ('Sirt1', 'ENSMUST00000020257', '100100010000000000000010'),\n",
       " ('Vegfc', 'ENSMUST00000033919', '000000011000000000001010'),\n",
       " ('Ccl9', 'ENSMUST00000019266', '101000000001000100000000'),\n",
       " ('Cdc73', 'ENSMUST00000018337', '000010101000000000001000'),\n",
       " ('Csf1r', 'ENSMUST00000115268', '000000001010100001000000'),\n",
       " ('Kif16b', 'ENSMUST00000043589', '000001000100000000110000'),\n",
       " ('Nlrc3', 'ENSMUST00000180200', '010000000100000110000000'),\n",
       " ('Eda2r', 'ENSMUST00000037353', '111100000000000000000000'),\n",
       " ('Lif', 'ENSMUST00000040750', '000001110000000100000000'),\n",
       " ('Ercc6', 'ENSMUST00000066807', '000000100001000000000101'),\n",
       " ('Map3k6', 'ENSMUST00000030677', '010001000000101000000000'),\n",
       " ('Foxo1', 'ENSMUST00000053764', '001100000000000000010001'),\n",
       " ('Pgf', 'ENSMUST00000004913', '000000000000100000011010'),\n",
       " ('Tgfb2', 'ENSMUST00000045288', '000001000011001000000000'),\n",
       " ('Prkdc', 'ENSMUST00000023352', '000101000000000010010000'),\n",
       " ('Map3k1', 'ENSMUST00000109267', '000000000000100011100000'),\n",
       " ('Smad3', 'ENSMUST00000034973', '001000110001000000000000'),\n",
       " ('Gpnmb', 'ENSMUST00000031840', '000000110000000000100001'),\n",
       " ('Map3k8', 'ENSMUST00000025078', '100001010000000000010000'),\n",
       " ('Agt', 'ENSMUST00000063278', '010011000010000000000000'),\n",
       " ('Ccr7', 'ENSMUST00000103134', '000000001000000001110000'),\n",
       " ('Map3k5', 'ENSMUST00000095806', '000001000000001010100000'),\n",
       " ('Fktn', 'ENSMUST00000128667', '101000011000000000000000'),\n",
       " ('Map3k4', 'ENSMUST00000089058', '000010100000100001000000'),\n",
       " ('Adrb2', 'ENSMUST00000053640', '010100110000000000000000'),\n",
       " ('Flt1', 'ENSMUST00000031653', '000010100000000000010100'),\n",
       " ('Prkcg', 'ENSMUST00000100301', '011000000000100010000000'),\n",
       " ('Anxa9', 'ENSMUST00000015846', '000000000110000010100000'),\n",
       " ('Edn1', 'ENSMUST00000021796', '011000010000000000001000'),\n",
       " ('Chil1', 'ENSMUST00000082060', '000100000000000000111000'),\n",
       " ('Itga1', 'ENSMUST00000061673', '000000000000000001001101'),\n",
       " ('Adam8', 'ENSMUST00000106069', '000001010001000000000100'),\n",
       " ('Il17ra', 'ENSMUST00000002976', '010000000000100001001000'),\n",
       " ('Agtr2', 'ENSMUST00000089188', '010001100100000000000000'),\n",
       " ('Lyn', 'ENSMUST00000041377', '000000010000000101100000'),\n",
       " ('Htr2a', 'ENSMUST00000036653', '001010000001000000001000'),\n",
       " ('Map4k1', 'ENSMUST00000085835', '100000100000000000000110'),\n",
       " ('Angpt2', 'ENSMUST00000033846', '010000010100000000000010'),\n",
       " ('Rasgrp4', 'ENSMUST00000032811', '000010110010000000000000'),\n",
       " ('Irak3', 'ENSMUST00000020448', '000001000100100010000000'),\n",
       " ('Rasgrp3', 'ENSMUST00000095204', '100000000100000000100001'),\n",
       " ('Fgf14', 'ENSMUST00000095529', '000010010001000000000010'),\n",
       " ('Angpt4', 'ENSMUST00000028955', '010000001000000000100001'),\n",
       " ('Gper1', 'ENSMUST00000066211', '000000001010001000000001'),\n",
       " ('Hyal1', 'ENSMUST00000112387', '100000001000000000010010'),\n",
       " ('Hras', 'ENSMUST00000026572', '000101100000000000000100'),\n",
       " ('Tsc2', 'ENSMUST00000097373', '000000100101010000000000'),\n",
       " ('Kiss1r', 'ENSMUST00000045529', '000000000010101000001000'),\n",
       " ('Mylk2', 'ENSMUST00000028970', '100000000100000000001100'),\n",
       " ('Dusp8', 'ENSMUST00000039926', '110000000110000000000000'),\n",
       " ('Havcr2', 'ENSMUST00000020668', '100000000001000000011000'),\n",
       " ('Flrt3', 'ENSMUST00000110057', '000000000101000000010001'),\n",
       " ('Egfr', 'ENSMUST00000020329', '000101000010010000000000'),\n",
       " ('Gata3', 'ENSMUST00000102976', '100101000100000000000000'),\n",
       " ('Btc', 'ENSMUST00000121044', '001000010000100000010000'),\n",
       " ('Fgf11', 'ENSMUST00000102585', '010100000010100000000000'),\n",
       " ('Cygb', 'ENSMUST00000021166', '010000001000000010000010'),\n",
       " ('Has1', 'ENSMUST00000003762', '010100000001000000001000'),\n",
       " ('Wnt2', 'ENSMUST00000010941', '000000010000101001000000'),\n",
       " ('Pparg', 'ENSMUST00000000450', '100001000000010000001000'),\n",
       " ('Trim71', 'ENSMUST00000111816', '000000010000001000110000'),\n",
       " ('Flrt1', 'ENSMUST00000113383', '001001100000001000000000'),\n",
       " ('Fgfr4', 'ENSMUST00000005452', '000010000000011100000000'),\n",
       " ('Fbln1', 'ENSMUST00000109432', '000000000010000000011100'),\n",
       " ('Pdgfb', 'ENSMUST00000000500', '100000010000001000000001'),\n",
       " ('Klb', 'ENSMUST00000031096', '000110100000000000100000'),\n",
       " ('Tdgf1', 'ENSMUST00000035075', '001000000000001001100000'),\n",
       " ('Wnt1', 'ENSMUST00000023734', '100000000100100001000000'),\n",
       " ('Hhip', 'ENSMUST00000079038', '000001011000001000000000'),\n",
       " ('Fgf1', 'ENSMUST00000040647', '000010000000001000010010'),\n",
       " ('Dll4', 'ENSMUST00000102517', '100000000000001001001000'),\n",
       " ('Shisa2', 'ENSMUST00000053949', '100010010000000000100000'),\n",
       " ('Fgf23', 'ENSMUST00000000186', '000000000100010001000100'),\n",
       " ('Rasgrf1', 'ENSMUST00000034912', '001000000000000010010100'),\n",
       " ('Wnt4', 'ENSMUST00000045747', '000000100000000001001010'),\n",
       " ('Prdm14', 'ENSMUST00000047577', '100000000000010100010000'),\n",
       " ('Fgf4', 'ENSMUST00000060336', '100000100000000000001001'),\n",
       " ('Fgfr3', 'ENSMUST00000202138', '000100001000001000010000')]"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# readout_names = ['1009169.0','458122.0','218057.0','275708.0','756916.0',\n",
    "#                  '638557.0','285692.0','719418.0','516032.0','432678.0',\n",
    "#                  '135628.0','184948.0','471571.0','942696.0','1074935.0',\n",
    "#                  '760360.0','1095367.0','428400.0','655176.0','202915.0']\n",
    "# readout_names = ['RS0095_cy5','RS0109_cy5','RS0175_cy5','RS0237_cy5','RS0307_cy5',\n",
    "#                  'RS0332_cy5','RS0384_atto565','RS0406_atto565','RS0451_atto565',\n",
    "#                  'RS0468_atto565','RS0548_atto565','RS64.0_atto565','RSN9927.0_cy5',\n",
    "#                  'RSN2336.0_cy5','RSN1807.0_cy5','RSN4287.0_atto565','RSN1252.0_atto565',\n",
    "#                  'RSN9535.0_atto565']\n",
    "# readout_names = ['RS0095_cy5', 'RS0109_cy5', 'RS0175_cy5', 'RS0237_cy5', 'RS0307_cy5',\n",
    "#                  'RS0332_cy5', 'RS0384_atto565', 'RS0406_atto565', 'RS0451_atto565',\n",
    "#                  'RS0468_atto565', 'RS0548_atto565', 'RS64.0_atto565', 'RS156.0_alexa488',\n",
    "#                  'RS278.0_alexa488', 'RS313.0_alexa488', 'RS643.0_alexa488', 'RS740.0_alexa488',\n",
    "#                  'RS810.0_alexa488', 'RSN9927.0_cy5', 'RSN2336.0_cy5', 'RSN1807.0_cy5',\n",
    "#                  'RSN4287.0_atto565', 'RSN1252.0_atto565', 'RSN9535.0_atto565']\n",
    "f = '/bigstore/binfo/mouse/24bit_disulfide_readouts.fasta'\n",
    "from Bio import SeqIO\n",
    "readout_names = []\n",
    "with open(f) as fasta_file:  # Will close handle cleanly\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        readout_names.append(seq_record.name)\n",
    "print(readout_names)\n",
    "\n",
    "def write_codebook(rows, fname, readout_names, codebook_style = '148MHD4'):\n",
    "    with open(fname, 'w') as f:\n",
    "        f.write('version'+','+str(1)+'\\n')\n",
    "        f.write('codebook_name'+','+codebook_style+'\\n')\n",
    "        f.write('bit_names,'+','.join(readout_names)+'\\n')\n",
    "        f.write('name, id, barcode\\n')\n",
    "        for row in rows:\n",
    "            f.write(','.join([str(row[0]), str(row[1]), str(row[2])+'\\n']))\n",
    "# Generate Codebook\n",
    "row_tuples = []\n",
    "expression_worksheet = client.open(\"MAPK_genes\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)\n",
    "for idx, row in expression_df.drop_duplicates('Transcript_ID').iterrows():\n",
    "    if len(row['Transcript_ID'].split('.')[0])>0:\n",
    "        row_tuples.append((row['Gene'], row['Transcript_ID'].split('.')[0], str(row['Barcode']).zfill(24)))\n",
    "#         row_tuples.append((row['Gene'], row['Transcript_ID'].split('.')[0], str('111000000000000001').zfill(24)))\n",
    "base = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/results/MAPK'\n",
    "if not os.path.exists(base):\n",
    "    os.mkdir(base)\n",
    "write_codebook(row_tuples, os.path.join(base,'Codebook_Final.txt'), readout_names)\n",
    "row_tuples"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1NujfUqy1FliV-4YIj0HEoiNkM9X9YzAhE73QRiFZqoY',\n",
       " 'updatedRange': 'Sheet1!E2:E473',\n",
       " 'updatedRows': 472,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 472}"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse Fasta\n",
    "base = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/results/MAPK'\n",
    "f = os.path.join(base,'Final_Pass/MAPK_oligos.fasta')\n",
    "from Bio import SeqIO\n",
    "with open(f) as fasta_file:  # Will close handle cleanly\n",
    "    gene = []\n",
    "    gid = []\n",
    "    seqs = []\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        name = seq_record.description.split(' ')\n",
    "        seqs.append(str(seq_record.seq))\n",
    "        for i in name:\n",
    "            if 'ENS' in i:\n",
    "                gene.append(i.split('_')[0])\n",
    "                gid.append(i.split('_')[2])\n",
    "fasta = pd.DataFrame(columns = ['gene','gene_ID','sequence'])\n",
    "fasta['gene'] = gene\n",
    "fasta['gene_ID'] = gid\n",
    "fasta['sequence'] = seqs\n",
    "expression_worksheet = client.open(\"MAPK_genes\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)\n",
    "gene_name = expression_worksheet.range('A2:A'+str(len(expression_df)+1))\n",
    "oligos = expression_worksheet.range('E2:E'+str(len(expression_df)+1))\n",
    "genes = [i.value for i in gene_name if len(i.value)>0]\n",
    "for i,gene in enumerate(genes):\n",
    "    oligos[i].value = len(fasta[fasta['gene']==gene])\n",
    "expression_worksheet.update_cells(oligos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "108123\n"
     ]
    }
   ],
   "source": [
    "# Parse Fasta\n",
    "base = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/results/OligoPool_19Dec2019/'\n",
    "f = os.path.join(base,'combined_19Dec2019.fasta')\n",
    "from Bio import SeqIO\n",
    "with open(f) as fasta_file:  # Will close handle cleanly\n",
    "    headers = []\n",
    "    genes = []\n",
    "    sets = []\n",
    "    seqs = []\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        h = seq_record.description\n",
    "        headers.append(h)\n",
    "        sets.append(h.split(' ')[0])\n",
    "        for f in h.split(' '):\n",
    "            if '__' in f:\n",
    "                genes.append(f.split('_')[0])\n",
    "        seqs.append(str(seq_record.seq))\n",
    "fasta = pd.DataFrame()\n",
    "fasta['genes'] = genes\n",
    "fasta['headers'] = headers\n",
    "fasta['sequence'] = seqs\n",
    "fasta['sets'] = sets\n",
    "print(len(fasta))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('GCGGCGGGTGAAACAGCTATTGTGATGGAAGTTAGAGGGTATATGGACCCATCT...ACT', SingleLetterAlphabet()), id='bmdm_nfkb_v2', name='bmdm_nfkb_v2', description='bmdm_nfkb_v2 ef8d0a5f-004 RS0109_cy5 Sdha__ENSMUST00000022062__775__30__0.43333__68.3806__1 1 RS0175_cy5 RS0384_atto565 ef8d0a5f-001', dbxrefs=[])"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['zebra_finch', 'Wound_Response', 'MAPK', 'bmdm_nfkb_v2'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_fasta.sets.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "99327\n",
      "53.40833333333333\n"
     ]
    }
   ],
   "source": [
    "new_fasta = fasta.copy()\n",
    "import random\n",
    "limit = 92918\n",
    "length = 100\n",
    "for s in new_fasta.sets.unique():\n",
    "    if 'zeb' in s:\n",
    "        continue\n",
    "    if 'nfkb' in s:\n",
    "        continue\n",
    "    for gene,cc in Counter(new_fasta[new_fasta.sets==s].genes).items():\n",
    "        if len(new_fasta)>limit:\n",
    "            if cc>length:\n",
    "                new_fasta = new_fasta.drop(index=random.sample(list(new_fasta[(new_fasta.sets==s)&(new_fasta.genes==gene)].index),cc-length))\n",
    "print(len(new_fasta))\n",
    "print((len(new_fasta)-limit)/120)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "64.09\n"
     ]
    }
   ],
   "source": [
    "print((len(new_fasta)-limit)/100)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "tossers = ['Fgfr3','Fgf4','Prdm14','Wnt4','Rasgrf1','Fgf23','Shisa2','Dll4','Fgf1','Hhip','Wnt1','Tdgf1','Klb','Pdgfb','Fbln1','Fgfr4','Flrt1','Trim71','Pparg','Wnt2','Has1','Cygb','Fgf11','Btc','Gata3','Egfr','Flrt3','Havcr2','Dusp8','Mylk2','Kiss1r','Tsc2','Hras','Hyal1','Gper1','Angpt4','Fgf14','Rasgrp3','Irak3','Rasgrp4','Angpt2','Map4k1','Htr2a','Lyn','Agtr2','Il17ra','Adam8','Itga1','Chil1','Edn1','Anxa9','Prkcg','Flt1','Adrb2','Map3k4','Fktn','Map3k5','Ccr7','Agt','Map3k8','Gpnmb','Smad3','Map3k1','Prkdc','Tgfb2','Pgf']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "92855\n",
      "-0.525\n"
     ]
    }
   ],
   "source": [
    "# new_fasta = fasta.copy()\n",
    "import random\n",
    "limit = 92918\n",
    "length = 100\n",
    "tossed = []\n",
    "for s in new_fasta.sets.unique():\n",
    "    if 'zeb' in s:\n",
    "        continue\n",
    "    if 'nfkb' in s:\n",
    "        continue\n",
    "    for gene,cc in Counter(new_fasta[new_fasta.sets==s].genes).items():\n",
    "        if len(new_fasta)>limit:\n",
    "            if gene in tossers:\n",
    "                tossed.append(gene)\n",
    "                new_fasta = new_fasta.drop(index=list(new_fasta[(new_fasta.sets==s)&(new_fasta.genes==gene)].index))\n",
    "print(len(new_fasta))\n",
    "print((len(new_fasta)-limit)/120)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genes</th>\n",
       "      <th>headers</th>\n",
       "      <th>sequence</th>\n",
       "      <th>sets</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>FOXP4</td>\n",
       "      <td>zebra_finch c9cfd45a-004 RS810.0 RS278.0 FOXP4...</td>\n",
       "      <td>GACGCCGCTAGATAGTGCCTGAGTGTGTGTTAAGGTAGGTAGAGAA...</td>\n",
       "      <td>zebra_finch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>FOXP4</td>\n",
       "      <td>zebra_finch c9cfd45a-004 RS0468 FOXP4__XM_0302...</td>\n",
       "      <td>GACGCCGCTAGATAGTGCCTAGGAGGAGGGTAATGATAGAAGTTCT...</td>\n",
       "      <td>zebra_finch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>FOXP4</td>\n",
       "      <td>zebra_finch c9cfd45a-004 RS0109 RS278.0 FOXP4_...</td>\n",
       "      <td>GACGCCGCTAGATAGTGCCTTGTGATGGAAGTTAGAGGGTAGAGAA...</td>\n",
       "      <td>zebra_finch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>FOXP4</td>\n",
       "      <td>zebra_finch c9cfd45a-004 RS0109 FOXP4__XM_0302...</td>\n",
       "      <td>GACGCCGCTAGATAGTGCCTTGTGATGGAAGTTAGAGGGTATAACT...</td>\n",
       "      <td>zebra_finch</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>FOXP4</td>\n",
       "      <td>zebra_finch c9cfd45a-004 RS278.0 RS0109 FOXP4_...</td>\n",
       "      <td>GACGCCGCTAGATAGTGCCTGAGAAGTGGTTGTAGAGTGTATGTGA...</td>\n",
       "      <td>zebra_finch</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   genes                                            headers  \\\n",
       "0  FOXP4  zebra_finch c9cfd45a-004 RS810.0 RS278.0 FOXP4...   \n",
       "1  FOXP4  zebra_finch c9cfd45a-004 RS0468 FOXP4__XM_0302...   \n",
       "2  FOXP4  zebra_finch c9cfd45a-004 RS0109 RS278.0 FOXP4_...   \n",
       "3  FOXP4  zebra_finch c9cfd45a-004 RS0109 FOXP4__XM_0302...   \n",
       "4  FOXP4  zebra_finch c9cfd45a-004 RS278.0 RS0109 FOXP4_...   \n",
       "\n",
       "                                            sequence         sets  \n",
       "0  GACGCCGCTAGATAGTGCCTGAGTGTGTGTTAAGGTAGGTAGAGAA...  zebra_finch  \n",
       "1  GACGCCGCTAGATAGTGCCTAGGAGGAGGGTAATGATAGAAGTTCT...  zebra_finch  \n",
       "2  GACGCCGCTAGATAGTGCCTTGTGATGGAAGTTAGAGGGTAGAGAA...  zebra_finch  \n",
       "3  GACGCCGCTAGATAGTGCCTTGTGATGGAAGTTAGAGGGTATAACT...  zebra_finch  \n",
       "4  GACGCCGCTAGATAGTGCCTGAGAAGTGGTTGTAGAGTGTATGTGA...  zebra_finch  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_fasta.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('GCGGCGGGTGAAACAGCTATTGTGATGGAAGTTAGAGGGTATATGGACCCATCT...ACT', SingleLetterAlphabet()), id='bmdm_nfkb_v2', name='bmdm_nfkb_v2', description='bmdm_nfkb_v2 ef8d0a5f-004 RS0109_cy5 Sdha__ENSMUST00000022062__775__30__0.43333__68.3806__1 1 RS0175_cy5 RS0384_atto565 ef8d0a5f-001', dbxrefs=[])"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "seq_record"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "10000\n",
      "20000\n",
      "30000\n",
      "40000\n",
      "50000\n",
      "60000\n",
      "90000\n",
      "100000\n"
     ]
    }
   ],
   "source": [
    "from Bio import SeqIO,Seq\n",
    "import re\n",
    "results = []\n",
    "for i,row in new_fasta.iterrows():\n",
    "    if i%10000==0:\n",
    "        print(i)\n",
    "    result = SeqIO.SeqRecord(Seq.Seq(row['sequence']))\n",
    "    result.id = str(row['sets'])\n",
    "    result.name = str(row['sets'])\n",
    "    result.description = str(row['headers'])\n",
    "    results.append(result)\n",
    "fname = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/results/OligoPool_19Dec2019/Final_combined_19Dec2019.fasta'\n",
    "with open(fname, 'w') as f:\n",
    "    SeqIO.write(results,f,\"fasta\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      ">zebra_finch c9cfd45a-004 RS810.0 RS278.0 FOXP4__XM_030291731__629__30__0.53333__72.1189__1__0.72709 RS0109 c9cfd45a-001\r\n",
      "GACGCCGCTAGATAGTGCCTGAGTGTGTGTTAAGGTAGGTAGAGAAGTGGTTGTAGAGTG\r\n",
      "TAAGGTGAAGCTGCTCCTGCTGCTTCTTGTAGATGTGATGGAAGTTAGAGGGTGATCGCA\r\n",
      "CAGTCAGGACGCC\r\n",
      ">zebra_finch c9cfd45a-004 RS0468 FOXP4__XM_030291731__1740__30__0.6__75.4354__1__0.72709 RS278.0 RS810.0 c9cfd45a-001\r\n",
      "GACGCCGCTAGATAGTGCCTAGGAGGAGGGTAATGATAGAAGTTCTCCACGCGCACGAAG\r\n",
      "CACTTGTGCAGAGAGAAGTGGTTGTAGAGTGTAGAGTGTGTGTTAAGGTAGGTGATCGCA\r\n",
      "CAGTCAGGACGCC\r\n",
      ">zebra_finch c9cfd45a-004 RS0109 RS278.0 FOXP4__XM_030291731__1304__30__0.56667__74.4808__1__0.72709 RS0468 c9cfd45a-001\r\n",
      "GACGCCGCTAGATAGTGCCTTGTGATGGAAGTTAGAGGGTAGAGAAGTGGTTGTAGAGTG\r\n"
     ]
    }
   ],
   "source": [
    "!head {fname}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "133"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len('GACGCCGCTAGATAGTGCCTGAGTGTGTGTTAAGGTAGGTAGAGAAGTGGTTGTAGAGTGTAAGGTGAAGCTGCTCCTGCTGCTTCTTGTAGATGTGATGGAAGTTAGAGGGTGATCGCACAGTCAGGACGCC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "92855\n"
     ]
    }
   ],
   "source": [
    "# Parse Fasta\n",
    "base = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/results/OligoPool_19Dec2019/'\n",
    "f = os.path.join(base,'Final_combined_19Dec2019.fasta')\n",
    "from Bio import SeqIO\n",
    "with open(f) as fasta_file:  # Will close handle cleanly\n",
    "    headers = []\n",
    "    genes = []\n",
    "    sets = []\n",
    "    seqs = []\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        h = seq_record.description\n",
    "        headers.append(h)\n",
    "        sets.append(h.split(' ')[0])\n",
    "        for f in h.split(' '):\n",
    "            if '__' in f:\n",
    "                genes.append(f.split('_')[0])\n",
    "        seqs.append(str(seq_record.seq))\n",
    "fasta = pd.DataFrame()\n",
    "fasta['genes'] = genes\n",
    "fasta['headers'] = headers\n",
    "fasta['sequence'] = seqs\n",
    "fasta['sets'] = sets\n",
    "print(len(fasta))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs = pd.DataFrame(fasta['sequence'])\n",
    "seqs.to_excel('/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/results/OligoPool_19Dec2019/Final_combined_19Dec2019.xlsx',index=False)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyspots [conda env:pyspots]",
   "language": "python",
   "name": "conda-env-pyspots-pyspots"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
