{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import SeqIO\n",
    "from fish_helpers import*"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['CTCTGCAATCACTGGTTCTTCCTTGACATGGATAG',\n",
       " 'GGACTGTGATTTGCTGTTGTTACCAAGGACATTGG',\n",
       " 'AGGCTCTTCCTCAATCTCTCTATCATCTTCCAGCT',\n",
       " 'CTTAGCAATGGTAAACTGCTCTCTGCAAGTGCAGC',\n",
       " 'TAAGCCACTGGATGCGTTATTTATCAGTCCTGGGT',\n",
       " 'CCAGCGAACCATTGAGATCTTCATGGACTGCCTGC',\n",
       " 'CAGCCTGGACTGCTATTTCCATTACTGTCAATGTG',\n",
       " 'AGCTGCTCCATAGCCTAAGCTGGTGGGTATGTTTT',\n",
       " 'AGTGTTAACTGCCTGTCACTCGATTCCATGATAGC',\n",
       " 'CCTGAAGTAAGCAAATGTCCGTGTAAACCAGCTGT',\n",
       " 'CCAACATGTTCTTAGACATGGTGACACTAGACACC',\n",
       " 'GTTGTGGGAGTTTGAGGTAAGCTCTGTGGGGATGT',\n",
       " 'AGGTCCCTGGGTAATTGGGGTGACTGGGGCCGTTG',\n",
       " 'CTCCCACATTGGGTACACTGGCTGGGGTGATTACA',\n",
       " 'GAGATGGCTTGGGCTCCGAGGGTCGCATGTGCAAG',\n",
       " 'TCTAACTGCTGAACCACTTGCATTTGCACCCGACA',\n",
       " 'GAGTATGGGAGGCCCCAGTCTCCTCATGTGACGAG',\n",
       " 'TCACAGCCAGGCCATTTGCAAACTCCATGACCATA',\n",
       " 'CCATAACTGCTGAATCTCAGCAGGACTTAAGCCAG',\n",
       " 'GTGATGCTTTGGAAGTGGTGGAGGAGGTAGTAGAG',\n",
       " 'TGACGCTGAAGGTTCAGCAGATGTTGCTGCTGCTG',\n",
       " 'AAGAGCAGATTGGCCAGGTGGGATGGAAATGAGTC',\n",
       " 'ACCTGGGGAGTCATCATGGCCACTGACACAGGCAC',\n",
       " 'GGAGAAGTGCTTGAAGCTGCTGCGGAGACAACACT']"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Cai_Probes = ['CTCTGCAATCACTGGTTCTTCCTTGACATGGATAG','GGACTGTGATTTGCTGTTGTTACCAAGGACATTGG','AGGCTCTTCCTCAATCTCTCTATCATCTTCCAGCT','CTTAGCAATGGTAAACTGCTCTCTGCAAGTGCAGC','TAAGCCACTGGATGCGTTATTTATCAGTCCTGGGT','CCAGCGAACCATTGAGATCTTCATGGACTGCCTGC','CAGCCTGGACTGCTATTTCCATTACTGTCAATGTG','AGCTGCTCCATAGCCTAAGCTGGTGGGTATGTTTT','AGTGTTAACTGCCTGTCACTCGATTCCATGATAGC','CCTGAAGTAAGCAAATGTCCGTGTAAACCAGCTGT','CCAACATGTTCTTAGACATGGTGACACTAGACACC','GTTGTGGGAGTTTGAGGTAAGCTCTGTGGGGATGT','AGGTCCCTGGGTAATTGGGGTGACTGGGGCCGTTG','CTCCCACATTGGGTACACTGGCTGGGGTGATTACA','GAGATGGCTTGGGCTCCGAGGGTCGCATGTGCAAG','TCTAACTGCTGAACCACTTGCATTTGCACCCGACA','GAGTATGGGAGGCCCCAGTCTCCTCATGTGACGAG','TCACAGCCAGGCCATTTGCAAACTCCATGACCATA','CCATAACTGCTGAATCTCAGCAGGACTTAAGCCAG','GTGATGCTTTGGAAGTGGTGGAGGAGGTAGTAGAG','TGACGCTGAAGGTTCAGCAGATGTTGCTGCTGCTG','AAGAGCAGATTGGCCAGGTGGGATGGAAATGAGTC','ACCTGGGGAGTCATCATGGCCACTGACACAGGCAC','GGAGAAGTGCTTGAAGCTGCTGCGGAGACAACACT']\n",
    "Cai_Probes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>100131</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTAGAGTGTGTGTTAAGGTAGGTAGAAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100132</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTAACCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100133</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAGCAAAGCAGATGCAGAAAAGCAGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100134</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTAAGCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100135</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAAGAGTGAGTAGTAGTGGAGTAGTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100136</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAGAGTGTGTGTTAAGGTAGGTAATGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100137</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGAAAGGAATGGGTTGTGGTAGCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100138</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTAAATT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100139</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTATTAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100140</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTACTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100141</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTACACATTGGGTACACTGGCTGGGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100142</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTACCTTTAGGGCTGTTCTTCTTTATCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100143</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTAATTGTCTTCCATACTGTGAACTCCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100144</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTAGACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100145</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAAGAGTGAGTAGTAGTGGAGTATTTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100146</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATCTCTGCAAGTGCAGCCTGCAAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100147</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTAATATGTTAAGCTCTAGCAGCCACCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100148</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGAAAGGAATGGGTTGTGGTACTGT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100149</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTAAGAGTGAGTAGTAGTGGAGTATGGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100150</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTAGCATAAGTAAATGGAGGTCTGACAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100151</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTATAAGAGCTGCTCCATAGCCTAAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100152</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTAAGAGTGAGTAGTAGTGGAGTATTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100153</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAGAGTGTGTGTTAAGGTAGGTAAACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100154</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100155</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTAGGTTTCTCAAGTCCCTTTTCATTCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100156</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTAGAGTGTGTGTTAAGGTAGGTACCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100157</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTAGTATGTTTTTCACTAACGTTGGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100158</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTACCCTCCATGTTTAATGCCATTGTCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100159</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTACAACTGTGCCATGAATTCAACCACA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100160</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAGCAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100221</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAAGGACATTTGTAAATCAGACTACGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100222</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTAGGAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100223</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTATGTGATGGAAGTTAGAGGGTACTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100224</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAAATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100225</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTAAGAGTGAGTAGTAGTGGAGTACCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100226</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGCTTTGGAAGTGGTGGAGGAGGTA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100227</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTATGAAAGGAATGGGTTGTGGTAGCAG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100228</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTACTTATGTCATTTACAGTTGGTGGCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100229</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTACCTGCATCATGACTTAATACCTGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100230</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTAGAGTGTGTGTTAAGGTAGGTAGATA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100231</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTAGAGTGTGTGTTAAGGTAGGTAGGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100232</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTATTATCGGGAGGAGACACCGTGTAGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100233</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTATGTGATGGAAGTTAGAGGGTAGACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100234</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTAGAAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100235</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTATGTGATGGAAGTTAGAGGGTAACTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100236</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTAAGAGTGAGTAGTAGTGGAGTATTGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100237</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAGGCCGTTGGTGTTGTGGGAGTTTGA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100238</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTAGCCG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100239</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTAAGAGTGAGTAGTAGTGGAGTAGCGG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100240</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTAAGTTATTGGTGGTGATGCTTTGGAA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100241</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAGTCA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100242</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAACCC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100243</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTATGAAAGGAATGGGTTGTGGTAGGTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100244</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTATCTC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100245</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAAACATGCACTCGGAACACCCATCAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100246</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>AGAGTGAGTAGTAGTGGAGTACATGGTGACACTAGACACCAAATTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100247</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTATGAT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100248</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTAAGAGTGAGTAGTAGTGGAGTACTGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100249</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGAAAGGAATGGGTTGTGGTAGAGTGTGTGTTAAGGTAGGTACCTT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100250</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTAGCTG...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>120 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         gene transcript_ID                                           sequence\n",
       "100131  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTAGAGTGTGTGTTAAGGTAGGTAGAAG...\n",
       "100132  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTAACCA...\n",
       "100133  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAGCAAAGCAGATGCAGAAAAGCAGAT...\n",
       "100134  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTAAGCC...\n",
       "100135  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAAGAGTGAGTAGTAGTGGAGTAGTCC...\n",
       "100136  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAGAGTGTGTGTTAAGGTAGGTAATGT...\n",
       "100137  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGAAAGGAATGGGTTGTGGTAGCAA...\n",
       "100138  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTAAATT...\n",
       "100139  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTATTAG...\n",
       "100140  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTACTTT...\n",
       "100141  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTACACATTGGGTACACTGGCTGGGGTG...\n",
       "100142  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTACCTTTAGGGCTGTTCTTCTTTATCC...\n",
       "100143  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTAATTGTCTTCCATACTGTGAACTCCA...\n",
       "100144  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTAGACA...\n",
       "100145  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAAGAGTGAGTAGTAGTGGAGTATTTT...\n",
       "100146  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATCTCTGCAAGTGCAGCCTGCAAGCT...\n",
       "100147  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTAATATGTTAAGCTCTAGCAGCCACCT...\n",
       "100148  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGAAAGGAATGGGTTGTGGTACTGT...\n",
       "100149  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTAAGAGTGAGTAGTAGTGGAGTATGGC...\n",
       "100150  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTAGCATAAGTAAATGGAGGTCTGACAT...\n",
       "100151  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTATAAGAGCTGCTCCATAGCCTAAGCT...\n",
       "100152  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTAAGAGTGAGTAGTAGTGGAGTATTCA...\n",
       "100153  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAGAGTGTGTGTTAAGGTAGGTAAACT...\n",
       "100154  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAGAAA...\n",
       "100155  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTAGGTTTCTCAAGTCCCTTTTCATTCC...\n",
       "100156  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTAGAGTGTGTGTTAAGGTAGGTACCTT...\n",
       "100157  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTAGTATGTTTTTCACTAACGTTGGGCT...\n",
       "100158  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTACCCTCCATGTTTAATGCCATTGTCT...\n",
       "100159  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTACAACTGTGCCATGAATTCAACCACA...\n",
       "100160  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAGCAA...\n",
       "...       ...           ...                                                ...\n",
       "100221  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAAGGACATTTGTAAATCAGACTACGG...\n",
       "100222  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTAGGAG...\n",
       "100223  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTATGTGATGGAAGTTAGAGGGTACTGG...\n",
       "100224  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAAATA...\n",
       "100225  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTAAGAGTGAGTAGTAGTGGAGTACCTC...\n",
       "100226  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGCTTTGGAAGTGGTGGAGGAGGTA...\n",
       "100227  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTATGAAAGGAATGGGTTGTGGTAGCAG...\n",
       "100228  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTACTTATGTCATTTACAGTTGGTGGCA...\n",
       "100229  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTACCTGCATCATGACTTAATACCTGCT...\n",
       "100230  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTAGAGTGTGTGTTAAGGTAGGTAGATA...\n",
       "100231  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTAGAGTGTGTGTTAAGGTAGGTAGGCT...\n",
       "100232  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTATTATCGGGAGGAGACACCGTGTAGG...\n",
       "100233  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTATGTGATGGAAGTTAGAGGGTAGACT...\n",
       "100234  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTAGAAA...\n",
       "100235  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTATGTGATGGAAGTTAGAGGGTAACTT...\n",
       "100236  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTAAGAGTGAGTAGTAGTGGAGTATTGG...\n",
       "100237  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAGGCCGTTGGTGTTGTGGGAGTTTGA...\n",
       "100238  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTAGCCG...\n",
       "100239  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTAAGAGTGAGTAGTAGTGGAGTAGCGG...\n",
       "100240  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTAAGTTATTGGTGGTGATGCTTTGGAA...\n",
       "100241  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAGTCA...\n",
       "100242  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGTGATGGAAGTTAGAGGGTAACCC...\n",
       "100243  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTATGAAAGGAATGGGTTGTGGTAGGTT...\n",
       "100244  FOXP2  XM_030289405  GAGTGTGTGTTAAGGTAGGTATGTGATGGAAGTTAGAGGGTATCTC...\n",
       "100245  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAAACATGCACTCGGAACACCCATCAT...\n",
       "100246  FOXP2  XM_030289405  AGAGTGAGTAGTAGTGGAGTACATGGTGACACTAGACACCAAATTG...\n",
       "100247  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTATGAT...\n",
       "100248  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTAAGAGTGAGTAGTAGTGGAGTACTGC...\n",
       "100249  FOXP2  XM_030289405  TGAAAGGAATGGGTTGTGGTAGAGTGTGTGTTAAGGTAGGTACCTT...\n",
       "100250  FOXP2  XM_030289405  TGTGATGGAAGTTAGAGGGTATGAAAGGAATGGGTTGTGGTAGCTG...\n",
       "\n",
       "[120 rows x 3 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse Fasta\n",
    "f = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/results/Caitlin/NCBI_Fourth_Pass/zebra_finch_possible_oligos.fasta'\n",
    "with open(f) as fasta_file:  # Will close handle cleanly\n",
    "    gene = []\n",
    "    tid = []\n",
    "    seqs = []\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        name = seq_record.description.split(' ')\n",
    "        seqs.append(str(seq_record.seq))\n",
    "        des = ''\n",
    "        for i in seq_record.description.split(' '):\n",
    "            if len(i)>len(des):\n",
    "                des=i\n",
    "        gene.append(des.split('__')[0])\n",
    "        tid.append(des.split('__')[1])\n",
    "fasta = pd.DataFrame(columns = ['gene','transcript_ID','sequence'])\n",
    "fasta['gene'] = gene\n",
    "fasta['transcript_ID'] = tid\n",
    "fasta['sequence'] = seqs\n",
    "probes = fasta[fasta.gene=='FOXP2']\n",
    "probes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>transcript_ID</th>\n",
       "      <th>sequence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>NM_001048263</td>\n",
       "      <td>GCCCTTTCTCTCTCGGTGTCAGCTCGCCGGGACGCCCTACACGGTG...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>41381</th>\n",
       "      <td>FOXP2</td>\n",
       "      <td>XM_030289405</td>\n",
       "      <td>CCCGCGCGGCGGCTCCGCCCGCAGGGACTCTGTATGTGCGCGCGGC...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene transcript_ID                                           sequence\n",
       "10     FOXP2  NM_001048263  GCCCTTTCTCTCTCGGTGTCAGCTCGCCGGGACGCCCTACACGGTG...\n",
       "41381  FOXP2  XM_030289405  CCCGCGCGGCGGCTCCGCCCGCAGGGACTCTGTATGTGCGCGCGGC..."
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Parse Fasta\n",
    "f = '/bigstore/binfo/zebra_finch/NCBI_Taeniopygia_guttata.transcripts.fa'\n",
    "FPKM_gid_col = 'gene_id'\n",
    "FPKM_tid_col = 'tracking_id'\n",
    "from Bio import SeqIO\n",
    "with open(f) as fasta_file:  # Will close handle cleanly\n",
    "    gene = []\n",
    "    tid = []\n",
    "    seqs = []\n",
    "    i=0\n",
    "    for seq_record in SeqIO.parse(fasta_file, 'fasta'):  # (generator)\n",
    "        seqs.append(str(seq_record.seq))\n",
    "        ID = seq_record.id\n",
    "        tid.append(ID)\n",
    "        try:\n",
    "            gene.append(str(seq_record.description.split(' gene=')[1]))\n",
    "        except:\n",
    "            gene.append('NaN')\n",
    "fasta = pd.DataFrame(columns = ['gene','transcript_ID','sequence'])\n",
    "fasta['gene'] = gene\n",
    "fasta['transcript_ID'] = tid\n",
    "fasta['sequence'] = seqs\n",
    "transcript_df = fasta[fasta.gene=='FOXP2']\n",
    "transcript_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AGAGTGAGTAGTAGTGGAGTAGAGTGTGTGTTAAGGTAGGTAGAAGACAAGCTGCTGGGCTGCCAACTGCTGATGAAAGGAATGGGTTGTGGT\n",
      "ACCACAACCCATTCCTTTCATCAGCAGTTGGCAGCCCAGCAGCTTGTCTTCTACCTACCTTAACACACACTCTACTCCACTACTACTCACTCT\n"
     ]
    }
   ],
   "source": [
    "seq = probes.iloc[0]['sequence']\n",
    "def reverse_complement(seq):\n",
    "    complement = {'A': 'T', 'C': 'G', 'G': 'C', 'T': 'A'} \n",
    "    rev = seq[::-1]\n",
    "    new_seq = []\n",
    "    for s in rev:\n",
    "        new_seq.append(complement[s.upper()])\n",
    "    return ''.join(new_seq)\n",
    "print(seq)\n",
    "print(reverse_complement(seq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "NM_001048263\n",
      "bad: GGACTGTGATTTGCTGTTGTTACCAAGGACATTGG\n",
      "GGACTGTGATTTGCTGTTGT   ACAACAGCAAATCACAGTCC\n",
      "bad: GTGATGCTTTGGAAGTGGTGGAGGAGGTAGTAGAG\n",
      "GTGATGCTTTGGAAGTGGTGGAGGAGGTAGT   ACTACCTCCTCCACCACTTCCAAAGCATCAC\n",
      "XM_030289405\n",
      "bad: GGACTGTGATTTGCTGTTGTTACCAAGGACATTGG\n",
      "GGACTGTGATTTGCTGTTGT   ACAACAGCAAATCACAGTCC\n"
     ]
    }
   ],
   "source": [
    "for i in transcript_df.index:\n",
    "    seq = transcript_df.loc[i]['sequence']\n",
    "    print(transcript_df.loc[i]['transcript_ID'])\n",
    "    for probe in Cai_Probes:\n",
    "        if not reverse_complement(probe) in seq:\n",
    "            print('bad:',probe)\n",
    "            largest_overlap = ''\n",
    "            for i in range(len(probe)):\n",
    "                for l in range(0,len(probe)):\n",
    "                    temp_probe = probe[i:i+l]\n",
    "                    if len(temp_probe)<10:\n",
    "                        continue\n",
    "                    if len(temp_probe)<=len(largest_overlap):\n",
    "                        continue\n",
    "                    elif reverse_complement(probe[i:i+l]) in seq:\n",
    "                        largest_overlap = probe[i:i+l]\n",
    "            print(largest_overlap,' ',reverse_complement(largest_overlap))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "100138 16 GGCCCCAGTC   GACTGGGGCC\n",
      "100166 9 CTGAAGTAAGCA   TGCTTACTTCAG\n",
      "100172 5 TTCATGGACTG   CAGTCCATGAA\n",
      "100194 20 TTCAGCAGAT   ATCTGCTGAA\n",
      "100195 6 CAGCCTGGACT   AGTCCAGGCTG\n",
      "100201 20 GGTTCAGCAG   CTGCTGAACC\n",
      "100210 9 TAAACCAGCT   AGCTGGTTTA\n",
      "100213 15 CTGCTGAACC   GGTTCAGCAG\n",
      "100227 9 CTGAAGTAAGCA   TGCTTACTTCAG\n"
     ]
    }
   ],
   "source": [
    "for row in probes.iterrows():\n",
    "    seq = row[1]['sequence']\n",
    "    for p,probe in enumerate(Cai_Probes):\n",
    "        largest_overlap = ''\n",
    "        for i in range(len(probe)):\n",
    "            for l in range(10,len(probe)):\n",
    "                temp_probe = probe[i:i+l]\n",
    "                if len(temp_probe)<10:\n",
    "                    continue\n",
    "                if len(temp_probe)<=len(largest_overlap):\n",
    "                    continue\n",
    "                elif reverse_complement(probe[i:i+l]) in seq:\n",
    "                    largest_overlap = probe[i:i+l]\n",
    "        if len(largest_overlap)!=0:\n",
    "            print(row[0],p,largest_overlap,' ',reverse_complement(largest_overlap))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "readout_probes = ['ACTCCACTACTACTCACTCT','ACTACACATCAACCTACTCC','TATTCACCTTACAAACCCTC','TATCCTTCAATCCCTCCACA','CCACCTTCCTACATAATACC','TCTATCATTACCCTCCTCCT','ACCCTCTAACTTCCATCACA','ACACTCTACAACCACTTCTC','TCCTAACAACCAACTACTCC','ACATTACACCTCATTCTCCC','AACACCACAACCTACTAACC','TTCTCCCTCTATCAACTCTA','ACCACAACCCATTCCTTTCA','ACCTACCTTAACACACACTC','ACCCTTACTACTACATCATC','CACATTCTCACCACTCACAT','CACCACCAATCACCTTATAC','CCTCACAAATTCTAACCTCC','TTTCTACCACTAATCAACCC','TCACTCAATCACCTCACTTC','CAACCACTAACCTCTAACCA','CCTCCTAACATAACACCTAC','ACCATCCTTAATCAACCACC','CCAATACCTAATCCTCTCTC']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('CCTGAGACAATGACGCGGGA', 'TGCTGGTATTGTGACGGCCA', 'GATGATGTAGTAGTAAGGGT', 'GTAGGTGTTATGTTAGGAGG', 'GAGAGAGGATTAGGTATTGG', 'TCCTGAGAGGAGCTGTAGGCTTGTCAGGGT')\n",
      "('', '', 'TGTGATGGAAGTTAGAGGGT', 'TGAAAGGAATGGGTTGTGGT', 'GAGTGTGTGTTAAGGTAGGT', 'AATATGGAAATCGGGGGTACTGGGCGGCCG')\n"
     ]
    }
   ],
   "source": [
    "def breakdown_probe(probe,readout_probes):\n",
    "    if len(probe)==133:\n",
    "        F = probe[0:20]\n",
    "        R = probe[-20:]\n",
    "        i=20\n",
    "    else:\n",
    "        F =''\n",
    "        R = ''\n",
    "        i=0\n",
    "    Readout=[]\n",
    "    for p in readout_probes:\n",
    "        rp = reverse_complement(p)\n",
    "        if rp in probe:\n",
    "            Readout.append(probe.find(rp))\n",
    "    R1 = probe[Readout[0]:Readout[0]+20]\n",
    "    R2 = probe[Readout[1]:Readout[1]+20]\n",
    "    R3 = probe[Readout[2]:Readout[2]+20]\n",
    "    q = probe[i+21:i+41]\n",
    "    if reverse_complement(q) in readout_probes:\n",
    "        T = probe[i+42:i+72]\n",
    "    else:\n",
    "        T = probe[i+21:i+51]\n",
    "    bad=True\n",
    "    if not reverse_complement(R1) in readout_probes:\n",
    "        print('Bad R1',R1)\n",
    "    elif not reverse_complement(R2) in readout_probes:\n",
    "        print('Bad R2',R2)\n",
    "    elif not reverse_complement(R3) in readout_probes:\n",
    "        print('Bad R3',R3)\n",
    "    return(F,R,R1,R2,R3,T)\n",
    "probe = 'CCTGAGACAATGACGCGGGAGTAGGTGTTATGTTAGGAGGAGATGATGTAGTAGTAAGGGTATCCTGAGAGGAGCTGTAGGCTTGTCAGGGTAGAGAGAGGATTAGGTATTGGTGCTGGTATTGTGACGGCCA'\n",
    "print(breakdown_probe(probe,readout_probes))\n",
    "probe = 'TGAAAGGAATGGGTTGTGGTAAATATGGAAATCGGGGGTACTGGGCGGCCGATGTGATGGAAGTTAGAGGGTAGAGTGTGTGTTAAGGTAGGT'\n",
    "print(breakdown_probe(probe,readout_probes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = transcript_df[transcript_df['transcript_ID']==probes['transcript_ID'].unique()[0]]['sequence'].iloc[0]\n",
    "for p,row in enumerate(probes.iterrows()):\n",
    "    probe = row[1]['sequence']\n",
    "    probe = breakdown_probe(probe,readout_probes)[5]\n",
    "    if not reverse_complement(probe) in seq:\n",
    "        print('bad:',p,probe,reverse_complement(probe))\n",
    "        largest_overlap = ''\n",
    "        for i in range(len(probe)):\n",
    "            for l in range(0,len(probe)):\n",
    "                temp_probe = probe[i:i+l]\n",
    "                if len(temp_probe)<10:\n",
    "                    continue\n",
    "                if len(temp_probe)<=len(largest_overlap):\n",
    "                    continue\n",
    "                elif reverse_complement(probe[i:i+l]) in seq:\n",
    "                    largest_overlap = probe[i:i+l]\n",
    "        if len(largest_overlap)!=0:\n",
    "            print(largest_overlap,' ',reverse_complement(largest_overlap))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyspots [conda env:pyspots]",
   "language": "python",
   "name": "conda-env-pyspots-pyspots"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
