{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from Bio import SeqUtils\n",
    "from fish_helpers import *\n",
    "from scipy.signal import lfilter\n",
    "import tqdm#.notebook as tqdm\n",
    "from Bio import SeqIO\n",
    "from scipy import sparse\n",
    "from multiprocessing import Pool\n",
    "from functools import partial\n",
    "import sys"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>filtered_genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0610009B22Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0610010K14Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0610011F06Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0610012G03Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0610037L13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>1110001J03Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>1110004E09Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>1110004F10Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>1110008F13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>1110008P14Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>1110018N20Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>1110032A03Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>1110032F04Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1110046J04Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>1110058L19Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>1110059E24Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>1110059G10Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1110065P20Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>1190002N15Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>1500009C09Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>1500009L16Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>1500011B03Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>1500011K16Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>1500012F01Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>1700001L19Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>1700001O22Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>1700003E16Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>1700007K13Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>1700008O03Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td>1700016K19Rik</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5424</th>\n",
       "      <td>Zfp703</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5425</th>\n",
       "      <td>Zfp704</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5426</th>\n",
       "      <td>Zfp706</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5427</th>\n",
       "      <td>Zfp710</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5428</th>\n",
       "      <td>Zfp804a</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5429</th>\n",
       "      <td>Zfp827</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5430</th>\n",
       "      <td>Zfp831</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5431</th>\n",
       "      <td>Zfp871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5432</th>\n",
       "      <td>Zfp941</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5433</th>\n",
       "      <td>Zfpm1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5434</th>\n",
       "      <td>Zfpm2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5435</th>\n",
       "      <td>Zfr</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5436</th>\n",
       "      <td>Zfyve9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5437</th>\n",
       "      <td>Zhx1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5438</th>\n",
       "      <td>Zkscan16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5439</th>\n",
       "      <td>Zmat2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5440</th>\n",
       "      <td>Zmat5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5441</th>\n",
       "      <td>Zmiz1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5442</th>\n",
       "      <td>Zmiz2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5443</th>\n",
       "      <td>Zmynd11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5444</th>\n",
       "      <td>Zmynd8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5445</th>\n",
       "      <td>Znhit1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5446</th>\n",
       "      <td>Znhit3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5447</th>\n",
       "      <td>Znrd1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5448</th>\n",
       "      <td>Znrf1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5449</th>\n",
       "      <td>Zranb2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5450</th>\n",
       "      <td>Zrsr1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5451</th>\n",
       "      <td>Zswim7</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5452</th>\n",
       "      <td>Zyg11b</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5453</th>\n",
       "      <td>l7Rn6</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5453 rows Ã— 1 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     filtered_genes\n",
       "1     0610009B22Rik\n",
       "2     0610010K14Rik\n",
       "3     0610011F06Rik\n",
       "4     0610012G03Rik\n",
       "5     0610037L13Rik\n",
       "6     1110001J03Rik\n",
       "7     1110004E09Rik\n",
       "8     1110004F10Rik\n",
       "9     1110008F13Rik\n",
       "10    1110008P14Rik\n",
       "11    1110018N20Rik\n",
       "12    1110032A03Rik\n",
       "13    1110032F04Rik\n",
       "14    1110046J04Rik\n",
       "15    1110058L19Rik\n",
       "16    1110059E24Rik\n",
       "17    1110059G10Rik\n",
       "18    1110065P20Rik\n",
       "19    1190002N15Rik\n",
       "20    1500009C09Rik\n",
       "21    1500009L16Rik\n",
       "22    1500011B03Rik\n",
       "23    1500011K16Rik\n",
       "24    1500012F01Rik\n",
       "25    1700001L19Rik\n",
       "26    1700001O22Rik\n",
       "27    1700003E16Rik\n",
       "28    1700007K13Rik\n",
       "29    1700008O03Rik\n",
       "30    1700016K19Rik\n",
       "...             ...\n",
       "5424         Zfp703\n",
       "5425         Zfp704\n",
       "5426         Zfp706\n",
       "5427         Zfp710\n",
       "5428        Zfp804a\n",
       "5429         Zfp827\n",
       "5430         Zfp831\n",
       "5431         Zfp871\n",
       "5432         Zfp941\n",
       "5433          Zfpm1\n",
       "5434          Zfpm2\n",
       "5435            Zfr\n",
       "5436         Zfyve9\n",
       "5437           Zhx1\n",
       "5438       Zkscan16\n",
       "5439          Zmat2\n",
       "5440          Zmat5\n",
       "5441          Zmiz1\n",
       "5442          Zmiz2\n",
       "5443        Zmynd11\n",
       "5444         Zmynd8\n",
       "5445         Znhit1\n",
       "5446         Znhit3\n",
       "5447          Znrd1\n",
       "5448          Znrf1\n",
       "5449         Zranb2\n",
       "5450          Zrsr1\n",
       "5451         Zswim7\n",
       "5452         Zyg11b\n",
       "5453          l7Rn6\n",
       "\n",
       "[5453 rows x 1 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv('/bigstore/binfo/mouse/Hippocampus/Allen/R_filtered_genes.csv',index_col=0)\n",
    "genes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Transcriptome: 125570it [00:05, 22651.16it/s]\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Parse Transcriptome\n",
    "transcript identifier (tid) (index)\n",
    "gene identifier (gene)\n",
    "sequence (seq)\n",
    "Convert seq to int (intseq)\n",
    "for each seed length region:\n",
    "find each identity(hash)\n",
    "find if it contains N bases(isvalid)\n",
    "\"\"\"\n",
    "resourcePath = '/bigstore/binfo/mouse/'\n",
    "rawTranscriptomeFasta = os.path.join(resourcePath,'mer_transcripts.fa')\n",
    "\n",
    "tids = []\n",
    "gids = []\n",
    "seqs = []\n",
    "intSeqs = []\n",
    "seed_hashs = []\n",
    "probe_len = 30\n",
    "seedLength = 17\n",
    "nt2int = {'A':0,'C':1,'G':2,'T':3,'N':np.nan}\n",
    "seedhashBase = [4**i for i in np.linspace(seedLength-1,0,seedLength)]\n",
    "with open(rawTranscriptomeFasta) as fasta_file:  # Will close handle cleanly\n",
    "    for seq_record in tqdm.tqdm(SeqIO.parse(fasta_file, 'fasta'),desc='Loading Transcriptome'):  # (generator)\n",
    "        tid,gid = seq_record.description.split(' ')\n",
    "        gid = gid.split('=')[-1]\n",
    "        seq = str(seq_record.seq)\n",
    "        if len(seq)>probe_len:\n",
    "#             intSeq = np.array([nt2int[i] for i in seq])\n",
    "#             seed_hash = list(lfilter(seedhashBase,1,intSeq)[seedLength-1:-1].astype(int))\n",
    "            tids.append(tid)\n",
    "            gids.append(gid)\n",
    "            seqs.append(seq)\n",
    "#             intSeqs.append(intSeq)\n",
    "#             seed_hashs.append(seed_hash)\n",
    "transcriptome = pd.DataFrame(index=tids)\n",
    "transcriptome['gene'] = gids\n",
    "transcriptome['seq'] = seqs\n",
    "# transcriptome['intseq'] = intSeqs\n",
    "# transcriptome['hash'] = seed_hashs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Load Annotation Data\n",
    "\"\"\"\n",
    "annotation = pd.read_csv('/bigstore/binfo/mouse/mart_export_25Nov2019.txt',sep='\\t')\n",
    "annotation.index = list(annotation['Transcript stable ID'])\n",
    "\"\"\"\n",
    "Filter Transcriptome to tsl<5\n",
    "\"\"\"\n",
    "tids = list(transcriptome.index)\n",
    "TSL = []\n",
    "for tsl in annotation['Transcript support level (TSL)']:\n",
    "    try:\n",
    "        TSL.append(tsl.split(' (')[0])\n",
    "    except:\n",
    "        TSL.append('tslNA')\n",
    "keepers = np.unique(TSL)\n",
    "keepers = [i for i in keepers if not i=='tslNA']\n",
    "keepers = [i for i in keepers if not i=='tsl5']\n",
    "mask = []\n",
    "for tsl in TSL:\n",
    "    if tsl in keepers:\n",
    "        mask.append(True)\n",
    "    else:\n",
    "        mask.append(False)\n",
    "TSL_tids = list(annotation[mask].index)\n",
    "filtered_tids = list(set(tids).intersection(TSL_tids))\n",
    "transcriptome = transcriptome.loc[filtered_tids]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading Expression Data: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 74535/74535 [01:06<00:00, 1114.61it/s]\n"
     ]
    },
    {
     "ename": "KeyError",
     "evalue": "'hash'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3077\u001b[0m             \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3078\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3079\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'hash'",
      "\nDuring handling of the above exception, another exception occurred:\n",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-5-e5b11ba648e6>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m     18\u001b[0m \u001b[0mmedian_expression\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcell_type_expression\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mmedian\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     19\u001b[0m \u001b[0mtranscriptome\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'expression'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mmedian_expression\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 20\u001b[0;31m \u001b[0mtranscriptome\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'expression_vector'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mones\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mh\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mfor\u001b[0m \u001b[0mh\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mtranscriptome\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'hash'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     21\u001b[0m \u001b[0mtranscriptome\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'expression_vector'\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mtranscriptome\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'expression_vector'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m*\u001b[0m\u001b[0mtranscriptome\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'expression'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0mBrain\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mhead\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2686\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_multilevel\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2687\u001b[0m         \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2688\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2689\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2690\u001b[0m     \u001b[0;32mdef\u001b[0m \u001b[0m_getitem_column\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m_getitem_column\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2693\u001b[0m         \u001b[0;31m# get column\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2694\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mis_unique\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2695\u001b[0;31m             \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_item_cache\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2696\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2697\u001b[0m         \u001b[0;31m# duplicate columns & possible reduce dimensionality\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/generic.py\u001b[0m in \u001b[0;36m_get_item_cache\u001b[0;34m(self, item)\u001b[0m\n\u001b[1;32m   2487\u001b[0m         \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcache\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2488\u001b[0m         \u001b[0;32mif\u001b[0m \u001b[0mres\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2489\u001b[0;31m             \u001b[0mvalues\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_data\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2490\u001b[0m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_box_item_values\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mvalues\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2491\u001b[0m             \u001b[0mcache\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mres\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/internals.py\u001b[0m in \u001b[0;36mget\u001b[0;34m(self, item, fastpath)\u001b[0m\n\u001b[1;32m   4113\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4114\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 4115\u001b[0;31m                 \u001b[0mloc\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mitem\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   4116\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   4117\u001b[0m                 \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0marange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0misna\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mitems\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/envs/pyspots/lib/python3.6/site-packages/pandas/core/indexes/base.py\u001b[0m in \u001b[0;36mget_loc\u001b[0;34m(self, key, method, tolerance)\u001b[0m\n\u001b[1;32m   3078\u001b[0m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3079\u001b[0m             \u001b[0;32mexcept\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 3080\u001b[0;31m                 \u001b[0;32mreturn\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_engine\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_loc\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_maybe_cast_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   3081\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   3082\u001b[0m         \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mget_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mmethod\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mmethod\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mtolerance\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mtolerance\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/index.pyx\u001b[0m in \u001b[0;36mpandas._libs.index.IndexEngine.get_loc\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;32mpandas/_libs/hashtable_class_helper.pxi\u001b[0m in \u001b[0;36mpandas._libs.hashtable.PyObjectHashTable.get_item\u001b[0;34m()\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'hash'"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Load Expression Data\n",
    "\"\"\"\n",
    "Brain = pd.read_csv('/bigstore/binfo/mouse/Hippocampus/medians.csv',index_col=0) # not the right way to do this\n",
    "\n",
    "expression = []\n",
    "empty_genes = []\n",
    "empty = Brain.iloc[0]\n",
    "for i,gene in tqdm.tqdm(enumerate(transcriptome.gene),total=len(transcriptome),desc='Loading Expression Data'):\n",
    "    try:\n",
    "        expression.append(pd.DataFrame(Brain.loc[gene]).T)\n",
    "    except:\n",
    "        empty.name = gene\n",
    "        empty_genes.append(gene)\n",
    "        expression.append(pd.DataFrame(empty).T)\n",
    "        continue\n",
    "cell_type_expression = pd.concat(expression)\n",
    "median_expression = cell_type_expression.median(axis=1)\n",
    "transcriptome['expression'] = list(median_expression)\n",
    "transcriptome['expression_vector'] = [np.ones(len(h)) for h in transcriptome['hash']]\n",
    "transcriptome['expression_vector'] = transcriptome['expression_vector']*transcriptome['expression']\n",
    "Brain.head()\n",
    "del Brain\n",
    "del cell_type_expression\n",
    "del median_expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Maybe do below with more than just 17 bp\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Generating Lookup: 100%|â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ| 69807459/69807459 [25:42<00:00, 45261.65it/s]   \n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Build Off Target Vector\n",
    "Filter by expression?\n",
    "\"\"\"\n",
    "count_cutoff = 0\n",
    "background_transcriptome = transcriptome[transcriptome['expression']>count_cutoff]\n",
    "background_vector = np.concatenate(background_transcriptome['hash'])\n",
    "background_expression_vector = np.concatenate(background_transcriptome['expression_vector'])\n",
    "background_array = np.concatenate((background_vector[:,None],background_expression_vector[:,None]),axis=1)\n",
    "background_lookup = {}\n",
    "for h in tqdm.tqdm(np.unique(background_vector),desc='Initializing Lookup',leave=False):\n",
    "    background_lookup[h] = 0\n",
    "for h,c in tqdm.tqdm(background_array,desc='Generating Lookup'):\n",
    "     background_lookup[h] = background_lookup[h]+c"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"\n",
    "Build Target Vector and mask\n",
    "Limit target transcriptome to just protien coding\n",
    "Filter by expression?\n",
    "Filter by genes without enough probes\n",
    "\"\"\"\n",
    "count_cutoff = 0\n",
    "tids = list(transcriptome.index)\n",
    "mRNA_tids = list(annotation[annotation['Transcript type']=='protein_coding'].index)\n",
    "filtered_tids = list(set(tids).intersection(mRNA_tids))\n",
    "target_transcriptome = transcriptome.loc[filtered_tids]\n",
    "target_transcriptome = target_transcriptome[target_transcriptome['expression']>count_cutoff]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "def generateHashLookup(partial_transcriptome):\n",
    "    vector = np.concatenate(list(partial_transcriptome['hash']))\n",
    "    expression_vector = np.concatenate(list(partial_transcriptome['expression_vector']))\n",
    "    array = np.concatenate((vector[:,None],expression_vector[:,None]),axis=1)\n",
    "    lookup = {}\n",
    "    for h in np.unique(vector):\n",
    "        lookup[h] = 0\n",
    "    for h,c in array:\n",
    "         lookup[h] =  lookup[h]+c\n",
    "    return lookup\n",
    "\n",
    "def generateProbeScores(gene_transcriptome,background_lookup,\n",
    "                        probe_len=30,\n",
    "                        monovalentSalt=0.3,\n",
    "                        seedLength=17,\n",
    "                        probeConc=5e-9,\n",
    "                        nt2int = {'A':0,'C':1,'G':2,'T':3,'N':np.nan},\n",
    "                        HT = [0,-7.6,-8.4,-7.8,-7.2,\n",
    "                              -8.5,-8.0,-10.6,-7.8,\n",
    "                              -8.2,-9.8,-8.0,-8.4,\n",
    "                              -7.2,-8.2,-8.5,-7.6],\n",
    "                        ST = [0,-21.3,-22.4,-21.0,-20.4,\n",
    "                              -22.7,-19.9,-27.2,-21.0,\n",
    "                              -22.2,-24.4,-19.9,-22.4,\n",
    "                              -21.3,-22.2,-22.7,-21.3],\n",
    "                        tm_max = 76,\n",
    "                        tm_min = 66,\n",
    "                        GC_max = 63/100,\n",
    "                        GC_min = 43/100):\n",
    "    \n",
    "    out = []\n",
    "    gene_lookup = generateHashLookup(gene_transcriptome)\n",
    "    hashBase = [4**i for i in np.linspace(probe_len-1,0,probe_len)]\n",
    "    for idx,row in gene_transcriptome.iterrows():\n",
    "        seq = row['seq']\n",
    "        gene = row['gene']\n",
    "        seed_hash = row['hash']\n",
    "        intSeq = row['intseq']\n",
    "        background_score  = [background_lookup[sh]-gene_lookup[sh] for sh in seed_hash]\n",
    "        gene_score  = [gene_lookup[sh] for sh in seed_hash]\n",
    "        h = lfilter(hashBase,1,intSeq)[probe_len-1:-1].astype(int)\n",
    "        nnID = (4*intSeq[:-1] + intSeq[1:])+1 # convert seq to pairs\n",
    "        nnID = np.nan_to_num(nnID).astype(int) # N to 0\n",
    "        # Calculate Free energy\n",
    "        dG = np.zeros([2,len(seq)-1])\n",
    "        dG[0,:] = np.array([HT[i] for i in nnID])\n",
    "        dG[1,:] = np.array([ST[i] for i in nnID])\n",
    "        # Calculate Entropy and Enthalpy\n",
    "        H = lfilter(np.ones([probe_len]),1,dG[0,:])[probe_len-1:]\n",
    "        S = lfilter(np.ones([probe_len]),1,dG[1,:])[probe_len-1:]\n",
    "        fivePrimeAT = (1*(intSeq==0)+1*(intSeq==3))[:-(probe_len)]\n",
    "        threePrimeAT = (1*(intSeq==0)+1*(intSeq==3))[probe_len:]\n",
    "        H = H+0.2+(2.2*fivePrimeAT)+(2.2*threePrimeAT)\n",
    "        S = S-5.7+(6.9*fivePrimeAT)+(6.9*threePrimeAT)\n",
    "        S = S + 0.368*(probe_len-1)*np.log(monovalentSalt)\n",
    "        # Calcuate Melting Temp for 30 bp probes\n",
    "        tm = (H*1000)/(S+1.9872*np.log(probeConc))-273.15\n",
    "        # Calculate GC content for 30 bp probes\n",
    "        gc = 1*((intSeq==1)|(intSeq==2))\n",
    "        gc = lfilter(np.ones([probe_len])/probe_len,1,gc)[probe_len-1:-1]\n",
    "        p_mask = 1*(((1*(tm>tm_max))+(1*(tm<tm_min))+(1*(gc>GC_max))+(1*(gc<GC_min)))==0)\n",
    "        values = [seq,gene,seed_hash,intSeq,h,tm,gc,p_mask,background_score,gene_score]\n",
    "        idx_df = pd.DataFrame(values,columns=[idx],index=['seq','gene','seed_hash','intSeq','hash','tm','gc','p_mask','background_score','gene_score']).T\n",
    "        out.append(idx_df)\n",
    "    return pd.concat(out)\n",
    "\n",
    "gene = target_transcriptome.gene.unique()[100]\n",
    "probe_len = 30\n",
    "gene_transcriptome = target_transcriptome[target_transcriptome.gene==gene]\n",
    "gene_transcriptome_processed = generateProbeScores(gene_transcriptome,background_lookup)\n",
    "\n",
    "# Input = []\n",
    "# for gene in tqdm.tqdm(target_transcriptome.gene.unique(),desc='Generating Input',leave=False):\n",
    "#     gene_transcriptome = target_transcriptome[target_transcriptome.gene==gene].copy()\n",
    "#     Input.append(gene_transcriptome)\n",
    "\n",
    "# sys.stdout.flush()\n",
    "# pfunc = partial(generateProbeScores,background_lookup=background_lookup)\n",
    "# gene_transcriptome_dict = {}\n",
    "# with Pool(30) as p:\n",
    "#     for gene_transcriptome_processed in tqdm.tqdm(p.imap(pfunc,Input),total=len(Input),desc='Outer'):\n",
    "#         gene = gene_transcriptome_processed.gene.iloc[0]\n",
    "#         gene_transcriptome_dict[gene] = gene_transcriptome_processed\n",
    "# sys.stdout.flush()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "p_mask = gene_transcriptome_processed['p_mask'].iloc[0]\n",
    "background_score = gene_transcriptome_processed['background_score'].iloc[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "81\n",
      "85\n"
     ]
    }
   ],
   "source": [
    "\"\"\"\n",
    "Find All passing probes that are a certain overlap apart\n",
    "\"\"\"\n",
    "p_mask = gene_transcriptome_processed['p_mask'].iloc[0]\n",
    "background_score = gene_transcriptome_processed['background_score'].iloc[0]\n",
    "p_mask = np.array(p_mask)\n",
    "background_score = np.array(background_score)\n",
    "p_score = np.array([np.sum(background_score[i:i+probe_len]) for i in range(len(p_mask))])\n",
    "\n",
    "thresh = 0\n",
    "overlap = 15\n",
    "probes = []\n",
    "for i,p in enumerate(p_mask):\n",
    "    if p==1:\n",
    "        if p_score[i]<=thresh:\n",
    "            if len(set(list(range(i-30+overlap,i+30-overlap))).intersection(probes))==0:\n",
    "                probes.append(i)\n",
    "print(len(probes))\n",
    "thresh = thresh+10\n",
    "for i,p in enumerate(p_mask):\n",
    "    if p==1:\n",
    "        if p_score[i]<=thresh:\n",
    "            if len(set(list(range(i-30+overlap,i+30-overlap))).intersection(probes))==0:\n",
    "                probes.append(i)\n",
    "print(len(probes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,\n",
       "       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.])"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "np.array(background_score)[np.array(probes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([4.53013149, 4.53013149, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 9.06026298, 9.06026298, 9.06026298, 9.06026298,\n",
       "       9.06026298, 4.53013149, 9.06026298, 9.06026298, 9.06026298])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_score = gene_transcriptome_processed['gene_score'].iloc[0]\n",
    "np.array(gene_score)[np.array(probes)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seq</th>\n",
       "      <th>gene</th>\n",
       "      <th>seed_hash</th>\n",
       "      <th>intSeq</th>\n",
       "      <th>hash</th>\n",
       "      <th>tm</th>\n",
       "      <th>gc</th>\n",
       "      <th>p_mask</th>\n",
       "      <th>background_score</th>\n",
       "      <th>gene_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000109280</th>\n",
       "      <td>ATGCTATTGAATCAAACAGAAGAAGAGGCCAGTGATTGTGAGAATC...</td>\n",
       "      <td>Pwwp2a</td>\n",
       "      <td>[4324520812, 1081130203, 8860217142, 221505428...</td>\n",
       "      <td>[0, 3, 2, 1, 3, 0, 3, 3, 2, 0, 0, 3, 1, 0, 0, ...</td>\n",
       "      <td>[405895854273262464, 101473963568315616, 60182...</td>\n",
       "      <td>[67.56977831708917, 67.91869496224194, 67.9220...</td>\n",
       "      <td>[0.39999999999999997, 0.39999999999999997, 0.4...</td>\n",
       "      <td>[0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[4.53013149020025, 4.53013149020025, 4.5301314...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ENSMUST00000061070</th>\n",
       "      <td>GTTAGCGGGAGGGAGGGGAAGGCCTCGGGATCCCAGCCGCGCTGCC...</td>\n",
       "      <td>Pwwp2a</td>\n",
       "      <td>[11318502974, 11419560335, 2854890083, 7137225...</td>\n",
       "      <td>[2, 3, 3, 0, 2, 1, 2, 2, 2, 0, 2, 2, 2, 0, 2, ...</td>\n",
       "      <td>[191220510911276608, 912496256182954368, 51635...</td>\n",
       "      <td>[78.79240988069893, 78.60571212548945, 79.8938...</td>\n",
       "      <td>[0.7, 0.6666666666666666, 0.7, 0.7333333333333...</td>\n",
       "      <td>[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...</td>\n",
       "      <td>[0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...</td>\n",
       "      <td>[4.53013149020025, 4.53013149020025, 4.5301314...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                  seq    gene  \\\n",
       "ENSMUST00000109280  ATGCTATTGAATCAAACAGAAGAAGAGGCCAGTGATTGTGAGAATC...  Pwwp2a   \n",
       "ENSMUST00000061070  GTTAGCGGGAGGGAGGGGAAGGCCTCGGGATCCCAGCCGCGCTGCC...  Pwwp2a   \n",
       "\n",
       "                                                            seed_hash  \\\n",
       "ENSMUST00000109280  [4324520812, 1081130203, 8860217142, 221505428...   \n",
       "ENSMUST00000061070  [11318502974, 11419560335, 2854890083, 7137225...   \n",
       "\n",
       "                                                               intSeq  \\\n",
       "ENSMUST00000109280  [0, 3, 2, 1, 3, 0, 3, 3, 2, 0, 0, 3, 1, 0, 0, ...   \n",
       "ENSMUST00000061070  [2, 3, 3, 0, 2, 1, 2, 2, 2, 0, 2, 2, 2, 0, 2, ...   \n",
       "\n",
       "                                                                 hash  \\\n",
       "ENSMUST00000109280  [405895854273262464, 101473963568315616, 60182...   \n",
       "ENSMUST00000061070  [191220510911276608, 912496256182954368, 51635...   \n",
       "\n",
       "                                                                   tm  \\\n",
       "ENSMUST00000109280  [67.56977831708917, 67.91869496224194, 67.9220...   \n",
       "ENSMUST00000061070  [78.79240988069893, 78.60571212548945, 79.8938...   \n",
       "\n",
       "                                                                   gc  \\\n",
       "ENSMUST00000109280  [0.39999999999999997, 0.39999999999999997, 0.4...   \n",
       "ENSMUST00000061070  [0.7, 0.6666666666666666, 0.7, 0.7333333333333...   \n",
       "\n",
       "                                                               p_mask  \\\n",
       "ENSMUST00000109280  [0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 1, 1, 1, 1, 1, ...   \n",
       "ENSMUST00000061070  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, ...   \n",
       "\n",
       "                                                     background_score  \\\n",
       "ENSMUST00000109280  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "ENSMUST00000061070  [0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, 0.0, ...   \n",
       "\n",
       "                                                           gene_score  \n",
       "ENSMUST00000109280  [4.53013149020025, 4.53013149020025, 4.5301314...  \n",
       "ENSMUST00000061070  [4.53013149020025, 4.53013149020025, 4.5301314...  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_transcriptome_processed"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CTGCAAAAGAAAGGTGCAAAAAGGTTTGGG\n",
      "0.4333333333333333\n",
      "70.71254248499565\n",
      "0.0\n",
      "4.53013149020025\n"
     ]
    }
   ],
   "source": [
    "seq = gene_transcriptome_processed['seq'].iloc[0]\n",
    "start = probes[0]\n",
    "end = start+probe_len\n",
    "probe_seq = seq[start:end]\n",
    "print(probe_seq)\n",
    "print(gene_transcriptome_processed['gc'].iloc[0][start])\n",
    "print(gene_transcriptome_processed['tm'].iloc[0][start])\n",
    "print(gene_transcriptome_processed['background_score'].iloc[0][start])\n",
    "print(gene_transcriptome_processed['gene_score'].iloc[0][start])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Allen_1:10000.csv.gz\n"
     ]
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from scipy import sparse\n",
    "base = '/bigstore/binfo/mouse/Hippocampus/Allen/'\n",
    "raw_counts = []\n",
    "cells = []\n",
    "genes  = []\n",
    "for f in os.listdir(base):\n",
    "    print(f)\n",
    "    if f.split('.')[-1] =='gz':\n",
    "        temp = pd.read_csv(os.path.join(base,f))\n",
    "        cells.extend(temp.columns)\n",
    "        genes.append(temp.index)\n",
    "        raw_counts.append(sparse.csr_matrix(temp))\n",
    "        del temp\n",
    "rc = np.concatenate([A.A for A in raw_counts],axis=1)\n",
    "rc = pd.DataFrame(rc,index=genes[0],columns=cells)\n",
    "rc.to_csv(os.path.join(base,'raw_counts.csv'))\n",
    "del rc\n",
    "del raw_counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "rc.to_csv(os.path.join(base,'raw_counts.csv'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'rc' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-41-a01c3420f801>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0;32mdel\u001b[0m \u001b[0mrc\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      2\u001b[0m \u001b[0;32mdel\u001b[0m \u001b[0mraw_counts\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'rc' is not defined"
     ]
    }
   ],
   "source": [
    "del rc\n",
    "del raw_counts"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyspots [conda env:pyspots]",
   "language": "python",
   "name": "conda-env-pyspots-pyspots"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
