{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "from os import path\n",
    "import gtfparse\n",
    "import pyensembl\n",
    "pyensembl.EnsemblRelease(release=87)\n",
    "import gspread\n",
    "from  oauth2client.service_account  import ServiceAccountCredentials\n",
    "import gspread\n",
    "import sys\n",
    "if sys.version_info[0] < 3: \n",
    "    from StringIO import StringIO\n",
    "else:\n",
    "    from io import StringIO"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "https://www.twilio.com/blog/2017/02/an-easy-way-to-read-and-write-to-a-google-spreadsheet-in-python.html\n",
    "Go to the Google APIs Console.\n",
    "Create a new project.\n",
    "Click Enable API. Search for and enable the Google Drive API.\n",
    "Create credentials for a Web Server to access Application Data.\n",
    "Name the service account and grant it a Project Role of Editor.\n",
    "Download the JSON file.\n",
    "Copy the JSON file to your code directory\n",
    "Open JSON and find \"client_email\"\n",
    "share google sheet with this email"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 145,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "INFO:oauth2client.client:Refreshing access_token\n"
     ]
    }
   ],
   "source": [
    "scope = ['https://spreadsheets.google.com/feeds',\n",
    "        'https://www.googleapis.com/auth/drive']\n",
    "# Path to .json\n",
    "path = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/ProbeDesign-92ebdb6a17a7.json'\n",
    "creds = ServiceAccountCredentials.from_json_keyfile_name(path, scope)\n",
    "client = gspread.authorize(creds)\n",
    "# Start with a gene list \n",
    "expression_worksheet = client.open(\"Housekeeping_Gene_List\").sheet1\n",
    "expression = expression_worksheet.get_all_records()\n",
    "expression_df = pd.DataFrame(expression)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Barcode</th>\n",
       "      <th>Description</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Isoform %</th>\n",
       "      <th>Length</th>\n",
       "      <th>Oligos</th>\n",
       "      <th>Transcript_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td></td>\n",
       "      <td>phosphoglycerate kinase 1 [Source:MGI Symbol;A...</td>\n",
       "      <td>7.191000e+01</td>\n",
       "      <td>Pgk1</td>\n",
       "      <td>ENSMUSG00000062070</td>\n",
       "      <td></td>\n",
       "      <td>1825</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000081593</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td></td>\n",
       "      <td>succinate dehydrogenase complex, subunit A, fl...</td>\n",
       "      <td>5.106000e+01</td>\n",
       "      <td>Sdha</td>\n",
       "      <td>ENSMUSG00000021577</td>\n",
       "      <td></td>\n",
       "      <td>2900</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000022062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td></td>\n",
       "      <td>TATA box binding protein [Source:MGI Symbol;Ac...</td>\n",
       "      <td>9.768260e-10</td>\n",
       "      <td>Tbp</td>\n",
       "      <td>ENSMUSG00000014767</td>\n",
       "      <td></td>\n",
       "      <td>2067</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000118001</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td></td>\n",
       "      <td>H2A histone family, member Z [Source:MGI Symbo...</td>\n",
       "      <td>2.683060e+01</td>\n",
       "      <td>H2afz</td>\n",
       "      <td>ENSMUSG00000037894</td>\n",
       "      <td></td>\n",
       "      <td>810</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000174561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td></td>\n",
       "      <td>glyceraldehyde-3-phosphate dehydrogenase [Sour...</td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>Gapdh</td>\n",
       "      <td>ENSMUSG00000057666</td>\n",
       "      <td></td>\n",
       "      <td>1273</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000117757</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td></td>\n",
       "      <td>peptidylprolyl isomerase A [Source:MGI Symbol;...</td>\n",
       "      <td>1.121390e-05</td>\n",
       "      <td>Ppia</td>\n",
       "      <td>ENSMUSG00000071866</td>\n",
       "      <td></td>\n",
       "      <td>1450</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000213200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td></td>\n",
       "      <td>glucuronidase, beta [Source:MGI Symbol;Acc:MGI...</td>\n",
       "      <td>3.017270e+00</td>\n",
       "      <td>Gusb</td>\n",
       "      <td>ENSMUSG00000025534</td>\n",
       "      <td></td>\n",
       "      <td>2038</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000111308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td></td>\n",
       "      <td>heat shock protein 1B [Source:MGI Symbol;Acc:M...</td>\n",
       "      <td>1.166990e-01</td>\n",
       "      <td>Hsp70</td>\n",
       "      <td>ENSMUSG00000090877</td>\n",
       "      <td></td>\n",
       "      <td>2803</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000172753</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td></td>\n",
       "      <td>tyrosine 3-monooxygenase/tryptophan 5-monooxyg...</td>\n",
       "      <td>1.924170e+01</td>\n",
       "      <td>Ywhaz</td>\n",
       "      <td>ENSMUSG00000022285</td>\n",
       "      <td></td>\n",
       "      <td>2122</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000110362</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td></td>\n",
       "      <td>actin, beta [Source:MGI Symbol;Acc:MGI:87904]</td>\n",
       "      <td>2.575700e+02</td>\n",
       "      <td>Actb</td>\n",
       "      <td>ENSMUSG00000029580</td>\n",
       "      <td></td>\n",
       "      <td>1242</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000167721</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td></td>\n",
       "      <td>beta-2 microglobulin [Source:MGI Symbol;Acc:MG...</td>\n",
       "      <td>1.630490e+03</td>\n",
       "      <td>B2m</td>\n",
       "      <td>ENSMUSG00000060802</td>\n",
       "      <td></td>\n",
       "      <td>860</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000102476</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td></td>\n",
       "      <td>hexokinase 2 [Source:MGI Symbol;Acc:MGI:1315197]</td>\n",
       "      <td>2.262000e-01</td>\n",
       "      <td>Hk2</td>\n",
       "      <td>ENSMUSG00000000628</td>\n",
       "      <td></td>\n",
       "      <td>2792</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000170833</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td></td>\n",
       "      <td>hypoxanthine guanine phosphoribosyl transferas...</td>\n",
       "      <td>8.015580e+01</td>\n",
       "      <td>Hprt</td>\n",
       "      <td>ENSMUSG00000025630</td>\n",
       "      <td></td>\n",
       "      <td>1289</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000026723</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.485140e+02</td>\n",
       "      <td>Arf1</td>\n",
       "      <td>ENSMUSG00000048076</td>\n",
       "      <td></td>\n",
       "      <td>1800</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000061242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.283660e+02</td>\n",
       "      <td>Cox7a2l</td>\n",
       "      <td>ENSMUSG00000024248</td>\n",
       "      <td></td>\n",
       "      <td>1076</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000025095</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.498250e+01</td>\n",
       "      <td>D8Ertd738e</td>\n",
       "      <td>ENSMUSG00000019362</td>\n",
       "      <td></td>\n",
       "      <td>616</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000019506</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.387750e+01</td>\n",
       "      <td>Eif1</td>\n",
       "      <td>ENSMUSG00000035530</td>\n",
       "      <td></td>\n",
       "      <td>1322</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000049385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.151720e+02</td>\n",
       "      <td>Eif4g2</td>\n",
       "      <td>ENSMUSG00000005610</td>\n",
       "      <td></td>\n",
       "      <td>3548</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000161051</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>0.000000e+00</td>\n",
       "      <td>Eif5a</td>\n",
       "      <td>ENSMUSG00000078812</td>\n",
       "      <td></td>\n",
       "      <td>1373</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000070996</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6.519010e+01</td>\n",
       "      <td>Gabarap</td>\n",
       "      <td>ENSMUSG00000018567</td>\n",
       "      <td></td>\n",
       "      <td>1351</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000018711</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3.243840e+00</td>\n",
       "      <td>Grcc10</td>\n",
       "      <td>ENSMUSG00000072772</td>\n",
       "      <td></td>\n",
       "      <td>863</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000004389</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.023150e+02</td>\n",
       "      <td>Myeov2</td>\n",
       "      <td>ENSMUSG00000073616</td>\n",
       "      <td></td>\n",
       "      <td>438</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000097642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7.841120e+01</td>\n",
       "      <td>Ndufa2</td>\n",
       "      <td>ENSMUSG00000014294</td>\n",
       "      <td></td>\n",
       "      <td>572</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000014438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>3.380540e+00</td>\n",
       "      <td>Nedd8</td>\n",
       "      <td>ENSMUSG00000010376</td>\n",
       "      <td></td>\n",
       "      <td>824</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000010520</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>2.154730e+01</td>\n",
       "      <td>Pomp</td>\n",
       "      <td>ENSMUSG00000029649</td>\n",
       "      <td></td>\n",
       "      <td>1579</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000031654</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.252470e+01</td>\n",
       "      <td>Psma2</td>\n",
       "      <td>ENSMUSG00000015671</td>\n",
       "      <td></td>\n",
       "      <td>2911</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000170836</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.291480e+00</td>\n",
       "      <td>Psma3</td>\n",
       "      <td>ENSMUSG00000060073</td>\n",
       "      <td></td>\n",
       "      <td>2846</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000160027</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.106510e+01</td>\n",
       "      <td>Psma4</td>\n",
       "      <td>ENSMUSG00000032301</td>\n",
       "      <td></td>\n",
       "      <td>1165</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000034848</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>7.817650e+01</td>\n",
       "      <td>Psmb1</td>\n",
       "      <td>ENSMUSG00000014769</td>\n",
       "      <td></td>\n",
       "      <td>1758</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000014913</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6.330940e+01</td>\n",
       "      <td>Psmb2</td>\n",
       "      <td>ENSMUSG00000028837</td>\n",
       "      <td></td>\n",
       "      <td>859</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000030642</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>30</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>4.028090e+01</td>\n",
       "      <td>Psmb3</td>\n",
       "      <td>ENSMUSG00000069744</td>\n",
       "      <td></td>\n",
       "      <td>766</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000103147</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>1.444870e+01</td>\n",
       "      <td>Psmb4</td>\n",
       "      <td>ENSMUSG00000005779</td>\n",
       "      <td></td>\n",
       "      <td>1183</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000005923</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>32</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>8.373330e+01</td>\n",
       "      <td>Psmb6</td>\n",
       "      <td>ENSMUSG00000018286</td>\n",
       "      <td></td>\n",
       "      <td>798</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000018430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>33</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.593730e+01</td>\n",
       "      <td>Psmb7</td>\n",
       "      <td>ENSMUSG00000026750</td>\n",
       "      <td></td>\n",
       "      <td>1160</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000028083</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>34</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.867120e+01</td>\n",
       "      <td>Rab1b</td>\n",
       "      <td>ENSMUSG00000024870</td>\n",
       "      <td></td>\n",
       "      <td>1913</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000025804</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>35</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>6.738930e+01</td>\n",
       "      <td>Saraf</td>\n",
       "      <td>ENSMUSG00000031532</td>\n",
       "      <td></td>\n",
       "      <td>1860</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000033933</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>36</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.491520e-01</td>\n",
       "      <td>Ubl5</td>\n",
       "      <td>ENSMUSG00000084786</td>\n",
       "      <td></td>\n",
       "      <td>2114</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000129414</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td></td>\n",
       "      <td></td>\n",
       "      <td>5.592200e+01</td>\n",
       "      <td>Vcp</td>\n",
       "      <td>ENSMUSG00000028452</td>\n",
       "      <td></td>\n",
       "      <td>3316</td>\n",
       "      <td></td>\n",
       "      <td>ENSMUST00000030164</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Barcode                                        Description          FPKM  \\\n",
       "0           phosphoglycerate kinase 1 [Source:MGI Symbol;A...  7.191000e+01   \n",
       "1           succinate dehydrogenase complex, subunit A, fl...  5.106000e+01   \n",
       "2           TATA box binding protein [Source:MGI Symbol;Ac...  9.768260e-10   \n",
       "3           H2A histone family, member Z [Source:MGI Symbo...  2.683060e+01   \n",
       "4           glyceraldehyde-3-phosphate dehydrogenase [Sour...  0.000000e+00   \n",
       "5           peptidylprolyl isomerase A [Source:MGI Symbol;...  1.121390e-05   \n",
       "6           glucuronidase, beta [Source:MGI Symbol;Acc:MGI...  3.017270e+00   \n",
       "7           heat shock protein 1B [Source:MGI Symbol;Acc:M...  1.166990e-01   \n",
       "8           tyrosine 3-monooxygenase/tryptophan 5-monooxyg...  1.924170e+01   \n",
       "9               actin, beta [Source:MGI Symbol;Acc:MGI:87904]  2.575700e+02   \n",
       "10          beta-2 microglobulin [Source:MGI Symbol;Acc:MG...  1.630490e+03   \n",
       "11           hexokinase 2 [Source:MGI Symbol;Acc:MGI:1315197]  2.262000e-01   \n",
       "12          hypoxanthine guanine phosphoribosyl transferas...  8.015580e+01   \n",
       "13                                                             1.485140e+02   \n",
       "14                                                             1.283660e+02   \n",
       "15                                                             5.498250e+01   \n",
       "16                                                             4.387750e+01   \n",
       "17                                                             1.151720e+02   \n",
       "18                                                             0.000000e+00   \n",
       "19                                                             6.519010e+01   \n",
       "20                                                             3.243840e+00   \n",
       "21                                                             1.023150e+02   \n",
       "22                                                             7.841120e+01   \n",
       "23                                                             3.380540e+00   \n",
       "24                                                             2.154730e+01   \n",
       "25                                                             1.252470e+01   \n",
       "26                                                             1.291480e+00   \n",
       "27                                                             5.106510e+01   \n",
       "28                                                             7.817650e+01   \n",
       "29                                                             6.330940e+01   \n",
       "30                                                             4.028090e+01   \n",
       "31                                                             1.444870e+01   \n",
       "32                                                             8.373330e+01   \n",
       "33                                                             5.593730e+01   \n",
       "34                                                             5.867120e+01   \n",
       "35                                                             6.738930e+01   \n",
       "36                                                             5.491520e-01   \n",
       "37                                                             5.592200e+01   \n",
       "\n",
       "          Gene             Gene_ID Isoform %  Length Oligos  \\\n",
       "0         Pgk1  ENSMUSG00000062070              1825          \n",
       "1         Sdha  ENSMUSG00000021577              2900          \n",
       "2          Tbp  ENSMUSG00000014767              2067          \n",
       "3        H2afz  ENSMUSG00000037894               810          \n",
       "4        Gapdh  ENSMUSG00000057666              1273          \n",
       "5         Ppia  ENSMUSG00000071866              1450          \n",
       "6         Gusb  ENSMUSG00000025534              2038          \n",
       "7        Hsp70  ENSMUSG00000090877              2803          \n",
       "8        Ywhaz  ENSMUSG00000022285              2122          \n",
       "9         Actb  ENSMUSG00000029580              1242          \n",
       "10         B2m  ENSMUSG00000060802               860          \n",
       "11         Hk2  ENSMUSG00000000628              2792          \n",
       "12        Hprt  ENSMUSG00000025630              1289          \n",
       "13        Arf1  ENSMUSG00000048076              1800          \n",
       "14     Cox7a2l  ENSMUSG00000024248              1076          \n",
       "15  D8Ertd738e  ENSMUSG00000019362               616          \n",
       "16        Eif1  ENSMUSG00000035530              1322          \n",
       "17      Eif4g2  ENSMUSG00000005610              3548          \n",
       "18       Eif5a  ENSMUSG00000078812              1373          \n",
       "19     Gabarap  ENSMUSG00000018567              1351          \n",
       "20      Grcc10  ENSMUSG00000072772               863          \n",
       "21      Myeov2  ENSMUSG00000073616               438          \n",
       "22      Ndufa2  ENSMUSG00000014294               572          \n",
       "23       Nedd8  ENSMUSG00000010376               824          \n",
       "24        Pomp  ENSMUSG00000029649              1579          \n",
       "25       Psma2  ENSMUSG00000015671              2911          \n",
       "26       Psma3  ENSMUSG00000060073              2846          \n",
       "27       Psma4  ENSMUSG00000032301              1165          \n",
       "28       Psmb1  ENSMUSG00000014769              1758          \n",
       "29       Psmb2  ENSMUSG00000028837               859          \n",
       "30       Psmb3  ENSMUSG00000069744               766          \n",
       "31       Psmb4  ENSMUSG00000005779              1183          \n",
       "32       Psmb6  ENSMUSG00000018286               798          \n",
       "33       Psmb7  ENSMUSG00000026750              1160          \n",
       "34       Rab1b  ENSMUSG00000024870              1913          \n",
       "35       Saraf  ENSMUSG00000031532              1860          \n",
       "36        Ubl5  ENSMUSG00000084786              2114          \n",
       "37         Vcp  ENSMUSG00000028452              3316          \n",
       "\n",
       "         Transcript_ID  \n",
       "0   ENSMUST00000081593  \n",
       "1   ENSMUST00000022062  \n",
       "2   ENSMUST00000118001  \n",
       "3   ENSMUST00000174561  \n",
       "4   ENSMUST00000117757  \n",
       "5   ENSMUST00000213200  \n",
       "6   ENSMUST00000111308  \n",
       "7   ENSMUST00000172753  \n",
       "8   ENSMUST00000110362  \n",
       "9   ENSMUST00000167721  \n",
       "10  ENSMUST00000102476  \n",
       "11  ENSMUST00000170833  \n",
       "12  ENSMUST00000026723  \n",
       "13  ENSMUST00000061242  \n",
       "14  ENSMUST00000025095  \n",
       "15  ENSMUST00000019506  \n",
       "16  ENSMUST00000049385  \n",
       "17  ENSMUST00000161051  \n",
       "18  ENSMUST00000070996  \n",
       "19  ENSMUST00000018711  \n",
       "20  ENSMUST00000004389  \n",
       "21  ENSMUST00000097642  \n",
       "22  ENSMUST00000014438  \n",
       "23  ENSMUST00000010520  \n",
       "24  ENSMUST00000031654  \n",
       "25  ENSMUST00000170836  \n",
       "26  ENSMUST00000160027  \n",
       "27  ENSMUST00000034848  \n",
       "28  ENSMUST00000014913  \n",
       "29  ENSMUST00000030642  \n",
       "30  ENSMUST00000103147  \n",
       "31  ENSMUST00000005923  \n",
       "32  ENSMUST00000018430  \n",
       "33  ENSMUST00000028083  \n",
       "34  ENSMUST00000025804  \n",
       "35  ENSMUST00000033933  \n",
       "36  ENSMUST00000129414  \n",
       "37  ENSMUST00000030164  "
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "expression_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '11MMvEY4s8oUIGf5kyoA4QOKs-Br2myyNooNmSO7-nEU',\n",
       " 'updatedCells': 188,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedRange': 'Sheet1!G2:G189',\n",
       " 'updatedRows': 188}"
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# removing trnascript versions from worksheet\n",
    "# Select a range\n",
    "Transcript_ID = expression_worksheet.range('G2:G189')\n",
    "# Edit range\n",
    "for cell in Transcript_ID:\n",
    "    cell.value = cell.value.split('.')[0]\n",
    "# Update in batch\n",
    "expression_worksheet.update_cells(Transcript_ID)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "def col_cells(worksheet, col):\n",
    "    \"\"\"Returns a range of cells in a `worksheet`'s column `col`.\"\"\"\n",
    "    start_cell = worksheet.get_addr_int(1, col)\n",
    "    end_cell = worksheet.get_addr_int(worksheet.row_count, col)\n",
    "\n",
    "    return worksheet.range('%s:%s' % (start_cell, end_cell))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: '/home/zach/Documents/Untitled Folder/cbook_140MHD4_200MHD2.txt'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-10-438f90632833>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      7\u001b[0m             \u001b[0mcwords\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mappend\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0ml\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      8\u001b[0m     \u001b[0;32mreturn\u001b[0m \u001b[0mcwords\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 9\u001b[0;31m \u001b[0mcodewords\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mread_codebook\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'/home/zach/Documents/Untitled Folder/cbook_140MHD4_200MHD2.txt'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;31m# need to fill in\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     10\u001b[0m \u001b[0mnp\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mrandom\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mshuffle\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcodewords\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     11\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'Number of codewords in codebook -'\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcodewords\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-10-438f90632833>\u001b[0m in \u001b[0;36mread_codebook\u001b[0;34m(cbook_fname)\u001b[0m\n\u001b[1;32m      2\u001b[0m     \u001b[0;34m\"\"\"Read CSV of n-bit codewords.\"\"\"\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      3\u001b[0m     \u001b[0mcwords\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m[\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 4\u001b[0;31m     \u001b[0;32mwith\u001b[0m \u001b[0mopen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mcbook_fname\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0;34m'r'\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;32mas\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      5\u001b[0m         \u001b[0mcolumn_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadline\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mstrip\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      6\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0ml\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreadlines\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: '/home/zach/Documents/Untitled Folder/cbook_140MHD4_200MHD2.txt'"
     ]
    }
   ],
   "source": [
    "def read_codebook(cbook_fname):\n",
    "    \"\"\"Read CSV of n-bit codewords.\"\"\"\n",
    "    cwords = []\n",
    "    with open(cbook_fname, 'r') as f:\n",
    "        column_name = f.readline().strip()\n",
    "        for l in f.readlines():\n",
    "            cwords.append(l.strip())\n",
    "    return cwords\n",
    "codewords = read_codebook('/home/zach/Documents/Untitled Folder/cbook_140MHD4_200MHD2.txt') # need to fill in\n",
    "np.random.shuffle(codewords)\n",
    "print('Number of codewords in codebook -', len(codewords))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Need to download biomart file for the species of interest\n",
    "This can be done here http://uswest.ensembl.org/biomart/martview/121d069e69616b12e28fa129a1e20f1c "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def geneSymbol_to_ensembl(gene_symbol_list, biomart_download_fname,\n",
    "                          organism='mouse', min_length=1000):\n",
    "    \"\"\"\n",
    "    Look up the ensembl of gene symbols.\n",
    "    \"\"\"\n",
    "    annotations = []\n",
    "    with open(biomart_download_fname, 'r') as f:\n",
    "        transcript_df = pd.read_csv(f)\n",
    "        #genes = [i.value for i in gene_symbol_list if i != '']\n",
    "        genes = gene_symbol_list\n",
    "        transcript_df = transcript_df[transcript_df[u'Gene name'].isin(genes)]\n",
    "    for cell in gene_symbol_list:\n",
    "        gene = cell#.value\n",
    "        if gene == '':\n",
    "            continue\n",
    "\n",
    "        transcripts = transcript_df[transcript_df[u'Gene name'] == gene].drop_duplicates('Transcript stable ID')\n",
    "        transcripts = transcripts[transcripts[u'Transcript type']==u'protein_coding'].sort_values('Transcript length (including UTRs and CDS)', ascending=False)\n",
    "        \n",
    "        if len(transcripts)==0:\n",
    "            print('Failed finding: ', gene)\n",
    "            annotations.append((gene, None))\n",
    "        elif transcripts.iloc[0]['Transcript length (including UTRs and CDS)'] > min_length:\n",
    "            annotations.append((gene, transcripts))\n",
    "        else:\n",
    "            print('Gene too short: ', gene)\n",
    "            annotations.append((gene, transcripts))\n",
    "    return annotations\n",
    "#"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def ensymbl_to_geneinfo(Gene_Stable_IDs, biomart_download_fname_path,\n",
    "                          organism='mouse', min_length=1000):\n",
    "    annotations = []\n",
    "    with open(biomart_download_fname_path, 'r') as f:\n",
    "        transcript_df = pd.read_csv(f)\n",
    "        genes = [i for i in Gene_Stable_IDs if i != '']\n",
    "        transcript_df = transcript_df[transcript_df[u'Gene stable ID'].isin(genes)]\n",
    "    for cell in Gene_Stable_IDs:\n",
    "        gene = cell#.value\n",
    "        if gene == '':\n",
    "            continue\n",
    "        transcripts = transcript_df[transcript_df[u'Gene stable ID'] == gene].drop_duplicates('Transcript stable ID')\n",
    "        transcripts = transcripts[transcripts[u'Transcript type']==u'protein_coding'].sort_values('Transcript length (including UTRs and CDS)', ascending=True)\n",
    "        if len(transcripts)==0:\n",
    "            print('Failed finding: ', gene)\n",
    "            annotations.append((gene, None))\n",
    "        elif transcripts.iloc[0]['Transcript length (including UTRs and CDS)'] > min_length:\n",
    "            annotations.append((gene, transcripts))\n",
    "        else:\n",
    "            print('Gene too short: ', gene)\n",
    "            annotations.append((gene, transcripts))\n",
    "    return annotations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def ensymbl_to_geneinfo(Gene_Stable_IDs, biomart_download_fname_path,expression_path,\n",
    "                          organism='mouse', min_length=1000):\n",
    "    annotations = []\n",
    "    FPKM_df = pd.read_csv(open(expression_path, 'r'))\n",
    "    transcript_df = pd.read_csv(open(biomart_download_fname_path, 'r'))\n",
    "    genes = [i for i in Gene_Stable_IDs if i != '']\n",
    "    FPKM_df = FPKM_df[FPKM_df[u'gene_id'].isin(genes)]\n",
    "    for cell in Gene_Stable_IDs:\n",
    "        gene = cell#.value\n",
    "        if gene == '':\n",
    "            continue\n",
    "        gene_FPKM_df = FPKM_df[FPKM_df[u'gene_id'] == gene].drop_duplicates('transcript_id')\n",
    "        gene_FPKM_df = gene_FPKM_df[gene_FPKM_df[u'FPKM']>0].sort_values('FPKM', ascending=False)\n",
    "        if len(gene_FPKM_df)==0:\n",
    "            print('Failed finding: ', gene)\n",
    "            annotations.append((gene, None))\n",
    "        else:\n",
    "            tids = [i for i in gene_FPKM_df['transcript_id'] if i != '']\n",
    "            transcripts = transcript_df[transcript_df[u'Transcript stable ID'].isin(tids)]\n",
    "            expression = []\n",
    "            for tid in transcripts['Transcript stable ID']:\n",
    "                expression.append(gene_FPKM_df[gene_FPKM_df['transcript_id']==tid]['FPKM'].iloc[0])\n",
    "            transcripts['FPKM']=expression\n",
    "            transcripts = transcripts.sort_values('FPKM', ascending=False)\n",
    "            annotations.append((gene, transcripts))\n",
    "    return annotations\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(genes[50])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1TmiXNh_yku0GHz6Jw76f0VPZH_h87kqJrrUUtJPfahU',\n",
       " 'updatedRange': 'Sheet1!A2:A3000',\n",
       " 'updatedRows': 2999,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 2999}"
      ]
     },
     "execution_count": 140,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "biomart_download_fname = '/bigstore/binfo/mouse/mouse_ensemble.txt'\n",
    "expression_path = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/MERFISH_analysis-master/mouse/Zach/Cornea/Expression/no_versions_isoforms_tracking.csv'\n",
    "FPKM_df = pd.read_csv(open(expression_path, 'r'))\n",
    "transcript_df = pd.read_csv(open(biomart_download_fname, 'r'))\n",
    "\n",
    "gids = expression_worksheet.range('F2:F3000')\n",
    "tids = expression_worksheet.range('G2:G3000')\n",
    "fpkm = expression_worksheet.range('C2:C3000')\n",
    "length = expression_worksheet.range('D2:D3000')\n",
    "descript = expression_worksheet.range('B2:B3000')\n",
    "gene_name = expression_worksheet.range('A2:A3000')\n",
    "\n",
    "genes = [i.value for i in gids if len(i.value)>0]\n",
    "FPKM_df = FPKM_df[FPKM_df[u'gene_id'].isin(genes)]\n",
    "transcript_df = transcript_df[transcript_df[u'Gene stable ID'].isin(genes)]\n",
    "for i,gene in enumerate(genes):\n",
    "    gene = gene\n",
    "    gene_fpkm_df = FPKM_df[(FPKM_df['gene_id']==gene)&(FPKM_df['length']>500)].sort_values('FPKM', ascending=False)\n",
    "    tid = gene_fpkm_df['transcript_id'].iloc[0]\n",
    "    tids[i].value = tid\n",
    "    fpkm[i].value = str(gene_fpkm_df[gene_fpkm_df['transcript_id']==tid]['FPKM'].iloc[0])\n",
    "    length[i].value = str(gene_fpkm_df[gene_fpkm_df['transcript_id']==tid]['length'].iloc[0])\n",
    "    annotations = transcript_df[transcript_df[u'Transcript stable ID']==tid]\n",
    "    descript[i].value = annotations['Gene description'].iloc[0]\n",
    "    gene_name[i].value = annotations['Gene name'].iloc[0]\n",
    "expression_worksheet.update_cells(tids)\n",
    "expression_worksheet.update_cells(fpkm)\n",
    "expression_worksheet.update_cells(length)\n",
    "expression_worksheet.update_cells(descript)\n",
    "expression_worksheet.update_cells(gene_name)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/zach/miniconda3/envs/pyspots/lib/python3.6/site-packages/ipykernel_launcher.py:24: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: http://pandas.pydata.org/pandas-docs/stable/indexing.html#indexing-view-versus-copy\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1TmiXNh_yku0GHz6Jw76f0VPZH_h87kqJrrUUtJPfahU',\n",
       " 'updatedRange': 'Sheet1!C2:C3000',\n",
       " 'updatedRows': 2999,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 2999}"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add info to spreadsheet\n",
    "# WARNING - you must figure out column numbers and change bits for new gene sets\n",
    "biomart_download_fname = '/bigstore/binfo/mouse/mouse_ensemble.txt'\n",
    "expression_path = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/MERFISH_analysis-master/mouse/Zach/Cornea/Expression/no_versions_isoforms_tracking.csv'\n",
    "\n",
    "gene_symbol_list = expression_df['Gene_ID']\n",
    "gene_annotations = ensymbl_to_geneinfo(gene_symbol_list, biomart_download_fname,expression_path, organism='finch', min_length=1000)\n",
    "#gname = expression_worksheet.range('F2:F3000')\n",
    "tname = expression_worksheet.range('G2:G3000')\n",
    "length = expression_worksheet.range('D2:D3000')\n",
    "descript = expression_worksheet.range('B2:B3000')\n",
    "fpkm = expression_worksheet.range('C2:C3000')\n",
    "#gene = expression_worksheet.range('A2:A3000')\n",
    "loc = 0\n",
    "update = []\n",
    "for idx, g in enumerate(gene_annotations):\n",
    "    info = g[1]\n",
    "    if info is None:\n",
    "        continue\n",
    "    #Need to better select transcripts here\n",
    "    max_transcript = info['Transcript stable ID'].iloc[loc]\n",
    "    max_length = info['Transcript length (including UTRs and CDS)'].iloc[loc]\n",
    "    descript_val = info['Gene description'].iloc[loc]\n",
    "    fpkm_val = info['FPKM'].iloc[loc]\n",
    "    #gene_name = info['Gene name'].iloc[loc]\n",
    "    #gene_id = info['Gene stable ID'].iloc[loc]\n",
    "    tname[idx].value = max_transcript\n",
    "    length[idx].value = str(max_length)\n",
    "    descript[idx].value = str(descript_val)\n",
    "    fpkm[idx].value = str(fpkm_val)\n",
    "    #gene[idx].value = str(gene_name)\n",
    "    #gname[idx].value = str(gene_id)\n",
    "expression_worksheet.update_cells(tname)\n",
    "expression_worksheet.update_cells(length)\n",
    "expression_worksheet.update_cells(descript)\n",
    "expression_worksheet.update_cells(fpkm)\n",
    "#expression_worksheet.update_cells(gene)\n",
    "#expression_worksheet.update_cells(gname)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "def Trascriptome_append(transcript_IDs,file):\n",
    "    annotations = []\n",
    "    with open(file, 'r') as f:\n",
    "        transcriptome_df = pd.read_csv(f)\n",
    "        genes = [i for i in transcript_IDs if i != '']\n",
    "        transcriptome_df = transcriptome_df[transcriptome_df[u'transcript_id'].isin(genes)] \n",
    "    for cell in transcript_IDs:\n",
    "        gene = cell\n",
    "        if gene == '':\n",
    "            continue\n",
    "        transcripts = transcriptome_df[transcriptome_df[u'transcript_id'] == gene]\n",
    "        if len(transcripts)==0:\n",
    "            print('Failed finding: ', gene)\n",
    "            annotations.append((gene, None))\n",
    "        else:\n",
    "            annotations.append((gene, transcripts))\n",
    "    return annotations "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>transcript_id</th>\n",
       "      <th>class_code</th>\n",
       "      <th>nearest_ref_id</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_short_name</th>\n",
       "      <th>tss_id</th>\n",
       "      <th>locus</th>\n",
       "      <th>length</th>\n",
       "      <th>coverage</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>FPKM_conf_lo</th>\n",
       "      <th>FPKM_conf_hi</th>\n",
       "      <th>FPKM_status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>121192</th>\n",
       "      <td>ENSMUST00000086878</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000024172</td>\n",
       "      <td>St6gal2</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17:55445388-55499226</td>\n",
       "      <td>2606</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50083</th>\n",
       "      <td>ENSMUST00000203475</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000108147</td>\n",
       "      <td>RP24-446A8.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139575923-139577205</td>\n",
       "      <td>1282</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50084</th>\n",
       "      <td>ENSMUST00000170650</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000092164</td>\n",
       "      <td>Rergl</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139492972-139501976</td>\n",
       "      <td>958</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50085</th>\n",
       "      <td>ENSMUST00000203246</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000107679</td>\n",
       "      <td>RP24-446A8.2</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139555558-139557611</td>\n",
       "      <td>498</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50086</th>\n",
       "      <td>ENSMUST00000083948</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000065882</td>\n",
       "      <td>n-R5s168</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140319629-140319733</td>\n",
       "      <td>104</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50087</th>\n",
       "      <td>ENSMUST00000203859</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000108229</td>\n",
       "      <td>RP23-330J17.2</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140309070-140309740</td>\n",
       "      <td>670</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50088</th>\n",
       "      <td>ENSMUST00000203881</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000108047</td>\n",
       "      <td>RP23-78J15.7</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140069993-140070498</td>\n",
       "      <td>505</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50089</th>\n",
       "      <td>ENSMUST00000204065</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000107407</td>\n",
       "      <td>RP23-78J15.4</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140132066-140132508</td>\n",
       "      <td>442</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50090</th>\n",
       "      <td>ENSMUST00000204425</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000107572</td>\n",
       "      <td>RP23-330J17.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140295462-140296456</td>\n",
       "      <td>994</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50091</th>\n",
       "      <td>ENSMUST00000158933</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000089558</td>\n",
       "      <td>Gm24174</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140166534-140166645</td>\n",
       "      <td>111</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50092</th>\n",
       "      <td>ENSMUST00000188066</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030228</td>\n",
       "      <td>Pik3c2g</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139587220-139622487</td>\n",
       "      <td>721</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50093</th>\n",
       "      <td>ENSMUST00000190962</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030228</td>\n",
       "      <td>Pik3c2g</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139587236-139635967</td>\n",
       "      <td>1774</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50094</th>\n",
       "      <td>ENSMUST00000187618</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030228</td>\n",
       "      <td>Pik3c2g</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139599479-139636074</td>\n",
       "      <td>2023</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50095</th>\n",
       "      <td>ENSMUST00000186585</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030228</td>\n",
       "      <td>Pik3c2g</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139599482-139657231</td>\n",
       "      <td>1358</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50096</th>\n",
       "      <td>ENSMUST00000185968</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030228</td>\n",
       "      <td>Pik3c2g</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139614070-139657231</td>\n",
       "      <td>1627</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50097</th>\n",
       "      <td>ENSMUST00000032353</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030228</td>\n",
       "      <td>Pik3c2g</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139621887-139635740</td>\n",
       "      <td>1206</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100536</th>\n",
       "      <td>ENSMUST00000223134</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000113453</td>\n",
       "      <td>RP24-70J8.4</td>\n",
       "      <td>-</td>\n",
       "      <td>chr13:63621499-63627350</td>\n",
       "      <td>751</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50099</th>\n",
       "      <td>ENSMUST00000203885</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000107487</td>\n",
       "      <td>RP23-78J15.6</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140056400-140057589</td>\n",
       "      <td>1189</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.027574</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100535</th>\n",
       "      <td>ENSMUST00000221519</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000113453</td>\n",
       "      <td>RP24-70J8.4</td>\n",
       "      <td>-</td>\n",
       "      <td>chr13:63621499-63626881</td>\n",
       "      <td>1003</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50102</th>\n",
       "      <td>ENSMUST00000032356</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030230</td>\n",
       "      <td>Plcz1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139989672-140041457</td>\n",
       "      <td>2271</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50103</th>\n",
       "      <td>ENSMUST00000137148</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030230</td>\n",
       "      <td>Plcz1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139989721-140041431</td>\n",
       "      <td>2226</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50104</th>\n",
       "      <td>ENSMUST00000129986</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030230</td>\n",
       "      <td>Plcz1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139989723-140041412</td>\n",
       "      <td>2178</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50105</th>\n",
       "      <td>ENSMUST00000149931</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000030230</td>\n",
       "      <td>Plcz1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140020250-140041412</td>\n",
       "      <td>989</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50106</th>\n",
       "      <td>ENSMUST00000204342</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000041791</td>\n",
       "      <td>Capza3</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140041251-140042158</td>\n",
       "      <td>771</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50107</th>\n",
       "      <td>ENSMUST00000043797</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000041791</td>\n",
       "      <td>Capza3</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140041556-140042778</td>\n",
       "      <td>1222</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50108</th>\n",
       "      <td>ENSMUST00000204787</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000108251</td>\n",
       "      <td>RP23-78J15.3</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:140033160-140033532</td>\n",
       "      <td>372</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100534</th>\n",
       "      <td>ENSMUST00000221930</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000113700</td>\n",
       "      <td>RP24-70J8.6</td>\n",
       "      <td>-</td>\n",
       "      <td>chr13:63661796-63662161</td>\n",
       "      <td>365</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>129294</th>\n",
       "      <td>ENSMUST00000124446</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000031388</td>\n",
       "      <td>Naa10</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:73917860-73920465</td>\n",
       "      <td>963</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.043211</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50082</th>\n",
       "      <td>ENSMUST00000204038</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000108128</td>\n",
       "      <td>RP23-238K12.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139315596-139322576</td>\n",
       "      <td>1149</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50081</th>\n",
       "      <td>ENSMUST00000158278</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000088903</td>\n",
       "      <td>Gm25434</td>\n",
       "      <td>-</td>\n",
       "      <td>chr6:139196064-139196231</td>\n",
       "      <td>167</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134495</th>\n",
       "      <td>ENSMUST00000082401</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064350</td>\n",
       "      <td>mt-Ty</td>\n",
       "      <td>-</td>\n",
       "      <td>chrM:5259-5326</td>\n",
       "      <td>67</td>\n",
       "      <td>5291.61</td>\n",
       "      <td>1700.89</td>\n",
       "      <td>898.553</td>\n",
       "      <td>2503.230000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>107869</th>\n",
       "      <td>ENSMUST00000226722</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000022108</td>\n",
       "      <td>Itm2b</td>\n",
       "      <td>-</td>\n",
       "      <td>chr14:73362884-73368505</td>\n",
       "      <td>935</td>\n",
       "      <td>5341.89</td>\n",
       "      <td>1717.05</td>\n",
       "      <td>1644.980</td>\n",
       "      <td>1789.130000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>132204</th>\n",
       "      <td>ENSMUST00000112172</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000049775</td>\n",
       "      <td>Tmsb4x</td>\n",
       "      <td>-</td>\n",
       "      <td>chrX:167207092-167209315</td>\n",
       "      <td>768</td>\n",
       "      <td>5388.94</td>\n",
       "      <td>1732.18</td>\n",
       "      <td>1659.310</td>\n",
       "      <td>1805.040000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>699</th>\n",
       "      <td>ENSMUST00000189941</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000100862</td>\n",
       "      <td>RP23-8J15.3</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:24613973-24614651</td>\n",
       "      <td>678</td>\n",
       "      <td>5398.74</td>\n",
       "      <td>1735.33</td>\n",
       "      <td>1719.700</td>\n",
       "      <td>1750.950000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93701</th>\n",
       "      <td>ENSMUST00000197752</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000106344</td>\n",
       "      <td>Mir1892</td>\n",
       "      <td>-</td>\n",
       "      <td>chr12:54645932-54646012</td>\n",
       "      <td>80</td>\n",
       "      <td>5583.25</td>\n",
       "      <td>1794.63</td>\n",
       "      <td>1188.700</td>\n",
       "      <td>2400.570000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>108445</th>\n",
       "      <td>ENSMUST00000134040</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000060126</td>\n",
       "      <td>Tpt1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr14:75845899-75848297</td>\n",
       "      <td>661</td>\n",
       "      <td>5656.84</td>\n",
       "      <td>1818.29</td>\n",
       "      <td>1741.440</td>\n",
       "      <td>1895.140000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>82989</th>\n",
       "      <td>ENSMUST00000092162</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000069515</td>\n",
       "      <td>Lyz1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10:117287796-117292868</td>\n",
       "      <td>1263</td>\n",
       "      <td>5792.09</td>\n",
       "      <td>1861.76</td>\n",
       "      <td>1850.920</td>\n",
       "      <td>1872.600000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83425</th>\n",
       "      <td>ENSMUST00000177557</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000093816</td>\n",
       "      <td>Gm26445</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10:130592135-130592244</td>\n",
       "      <td>109</td>\n",
       "      <td>5895.96</td>\n",
       "      <td>1895.15</td>\n",
       "      <td>1658.500</td>\n",
       "      <td>2131.790000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117485</th>\n",
       "      <td>ENSMUST00000102084</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000076038</td>\n",
       "      <td>Gm22774</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17:6131835-6131944</td>\n",
       "      <td>109</td>\n",
       "      <td>6178.38</td>\n",
       "      <td>1985.93</td>\n",
       "      <td>1743.130</td>\n",
       "      <td>2228.720000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>31107</th>\n",
       "      <td>ENSMUST00000102081</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000094403</td>\n",
       "      <td>Mir692-2</td>\n",
       "      <td>-</td>\n",
       "      <td>chr4:125504748-125504857</td>\n",
       "      <td>109</td>\n",
       "      <td>6289.17</td>\n",
       "      <td>2021.54</td>\n",
       "      <td>1776.400</td>\n",
       "      <td>2266.680000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55427</th>\n",
       "      <td>ENSMUST00000174355</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000002985</td>\n",
       "      <td>Apoe</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:19696271-19699008</td>\n",
       "      <td>1104</td>\n",
       "      <td>6441.73</td>\n",
       "      <td>2070.58</td>\n",
       "      <td>1984.170</td>\n",
       "      <td>2156.990000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4744</th>\n",
       "      <td>ENSMUST00000188837</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000100405</td>\n",
       "      <td>RP24-282H12.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:123205148-123205265</td>\n",
       "      <td>117</td>\n",
       "      <td>6870.43</td>\n",
       "      <td>2208.37</td>\n",
       "      <td>2003.020</td>\n",
       "      <td>2413.730000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55432</th>\n",
       "      <td>ENSMUST00000172808</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000002985</td>\n",
       "      <td>Apoe</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:19696843-19698247</td>\n",
       "      <td>456</td>\n",
       "      <td>7027.86</td>\n",
       "      <td>2258.98</td>\n",
       "      <td>2159.930</td>\n",
       "      <td>2358.020000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>63644</th>\n",
       "      <td>ENSMUST00000066401</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000007891</td>\n",
       "      <td>Ctsd</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:142375916-142387857</td>\n",
       "      <td>1922</td>\n",
       "      <td>7581.21</td>\n",
       "      <td>2436.84</td>\n",
       "      <td>2335.840</td>\n",
       "      <td>2537.840000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37766</th>\n",
       "      <td>ENSMUST00000197766</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000105796</td>\n",
       "      <td>RP23-440C1.3</td>\n",
       "      <td>-</td>\n",
       "      <td>chr5:87737105-87737210</td>\n",
       "      <td>105</td>\n",
       "      <td>8603.67</td>\n",
       "      <td>2765.49</td>\n",
       "      <td>2474.370</td>\n",
       "      <td>3056.610000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>103349</th>\n",
       "      <td>ENSMUST00000184558</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000098694</td>\n",
       "      <td>Gm27506</td>\n",
       "      <td>-</td>\n",
       "      <td>chr14:13778071-13778151</td>\n",
       "      <td>80</td>\n",
       "      <td>8900.06</td>\n",
       "      <td>2860.76</td>\n",
       "      <td>2216.050</td>\n",
       "      <td>3505.470000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>77035</th>\n",
       "      <td>ENSMUST00000121411</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000081926</td>\n",
       "      <td>RP24-217J7.2</td>\n",
       "      <td>-</td>\n",
       "      <td>chr10:7038660-7038774</td>\n",
       "      <td>114</td>\n",
       "      <td>9354.27</td>\n",
       "      <td>3006.76</td>\n",
       "      <td>2751.690</td>\n",
       "      <td>3261.830000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134501</th>\n",
       "      <td>ENSMUST00000082402</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064351</td>\n",
       "      <td>mt-Co1</td>\n",
       "      <td>-</td>\n",
       "      <td>chrM:5327-6872</td>\n",
       "      <td>1545</td>\n",
       "      <td>9711.81</td>\n",
       "      <td>3121.68</td>\n",
       "      <td>2992.370</td>\n",
       "      <td>3251.000000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>109958</th>\n",
       "      <td>ENSMUST00000159216</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064373</td>\n",
       "      <td>Sepp1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr15:3270920-3280507</td>\n",
       "      <td>2085</td>\n",
       "      <td>10492.80</td>\n",
       "      <td>3372.73</td>\n",
       "      <td>3233.180</td>\n",
       "      <td>3512.280000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>125362</th>\n",
       "      <td>ENSMUST00000025563</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000024661</td>\n",
       "      <td>Fth1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr19:9982702-9985092</td>\n",
       "      <td>866</td>\n",
       "      <td>11416.40</td>\n",
       "      <td>3669.59</td>\n",
       "      <td>3650.300</td>\n",
       "      <td>3688.880000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1509</th>\n",
       "      <td>ENSMUST00000187117</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000101249</td>\n",
       "      <td>RP23-8J15.1</td>\n",
       "      <td>-</td>\n",
       "      <td>chr1:24615705-24616197</td>\n",
       "      <td>492</td>\n",
       "      <td>13177.20</td>\n",
       "      <td>4235.58</td>\n",
       "      <td>4204.270</td>\n",
       "      <td>4266.890000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134494</th>\n",
       "      <td>ENSMUST00000082400</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064349</td>\n",
       "      <td>mt-Tc</td>\n",
       "      <td>-</td>\n",
       "      <td>chrM:5191-5257</td>\n",
       "      <td>66</td>\n",
       "      <td>13411.80</td>\n",
       "      <td>4310.98</td>\n",
       "      <td>2938.610</td>\n",
       "      <td>5683.340000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134511</th>\n",
       "      <td>ENSMUST00000082407</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064356</td>\n",
       "      <td>mt-Atp8</td>\n",
       "      <td>-</td>\n",
       "      <td>chrM:7765-7969</td>\n",
       "      <td>204</td>\n",
       "      <td>16887.60</td>\n",
       "      <td>5428.22</td>\n",
       "      <td>5178.680</td>\n",
       "      <td>5677.760000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>55425</th>\n",
       "      <td>ENSMUST00000167646</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000002985</td>\n",
       "      <td>Apoe</td>\n",
       "      <td>-</td>\n",
       "      <td>chr7:19696266-19699052</td>\n",
       "      <td>728</td>\n",
       "      <td>18751.70</td>\n",
       "      <td>6027.37</td>\n",
       "      <td>5776.900</td>\n",
       "      <td>6277.850000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75775</th>\n",
       "      <td>ENSMUST00000184437</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000098973</td>\n",
       "      <td>Mir6236</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:110281286-110281409</td>\n",
       "      <td>123</td>\n",
       "      <td>19556.50</td>\n",
       "      <td>6286.09</td>\n",
       "      <td>5975.420</td>\n",
       "      <td>6596.760000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>117924</th>\n",
       "      <td>ENSMUST00000102183</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000076137</td>\n",
       "      <td>Gm26384</td>\n",
       "      <td>-</td>\n",
       "      <td>chr17:21284102-21284212</td>\n",
       "      <td>110</td>\n",
       "      <td>20635.90</td>\n",
       "      <td>6633.03</td>\n",
       "      <td>6227.550</td>\n",
       "      <td>7038.510000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>74819</th>\n",
       "      <td>ENSMUST00000102184</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000076138</td>\n",
       "      <td>Gm26377</td>\n",
       "      <td>-</td>\n",
       "      <td>chr9:78478527-78478635</td>\n",
       "      <td>108</td>\n",
       "      <td>23976.00</td>\n",
       "      <td>7706.64</td>\n",
       "      <td>7050.590</td>\n",
       "      <td>8362.700000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134493</th>\n",
       "      <td>ENSMUST00000082399</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064348</td>\n",
       "      <td>mt-Tn</td>\n",
       "      <td>-</td>\n",
       "      <td>chrM:5088-5159</td>\n",
       "      <td>71</td>\n",
       "      <td>26765.20</td>\n",
       "      <td>8603.20</td>\n",
       "      <td>7143.770</td>\n",
       "      <td>10062.600000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134518</th>\n",
       "      <td>ENSMUST00000082423</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064372</td>\n",
       "      <td>mt-Tp</td>\n",
       "      <td>-</td>\n",
       "      <td>chrM:15355-15422</td>\n",
       "      <td>67</td>\n",
       "      <td>100835.00</td>\n",
       "      <td>32411.40</td>\n",
       "      <td>28889.200</td>\n",
       "      <td>35933.600000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>134492</th>\n",
       "      <td>ENSMUST00000082398</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>ENSMUSG00000064347</td>\n",
       "      <td>mt-Ta</td>\n",
       "      <td>-</td>\n",
       "      <td>chrM:5017-5086</td>\n",
       "      <td>69</td>\n",
       "      <td>122539.00</td>\n",
       "      <td>39388.00</td>\n",
       "      <td>35912.700</td>\n",
       "      <td>42863.300000</td>\n",
       "      <td>OK</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>134520 rows × 13 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "             transcript_id class_code nearest_ref_id             gene_id  \\\n",
       "121192  ENSMUST00000086878          -              -  ENSMUSG00000024172   \n",
       "50083   ENSMUST00000203475          -              -  ENSMUSG00000108147   \n",
       "50084   ENSMUST00000170650          -              -  ENSMUSG00000092164   \n",
       "50085   ENSMUST00000203246          -              -  ENSMUSG00000107679   \n",
       "50086   ENSMUST00000083948          -              -  ENSMUSG00000065882   \n",
       "50087   ENSMUST00000203859          -              -  ENSMUSG00000108229   \n",
       "50088   ENSMUST00000203881          -              -  ENSMUSG00000108047   \n",
       "50089   ENSMUST00000204065          -              -  ENSMUSG00000107407   \n",
       "50090   ENSMUST00000204425          -              -  ENSMUSG00000107572   \n",
       "50091   ENSMUST00000158933          -              -  ENSMUSG00000089558   \n",
       "50092   ENSMUST00000188066          -              -  ENSMUSG00000030228   \n",
       "50093   ENSMUST00000190962          -              -  ENSMUSG00000030228   \n",
       "50094   ENSMUST00000187618          -              -  ENSMUSG00000030228   \n",
       "50095   ENSMUST00000186585          -              -  ENSMUSG00000030228   \n",
       "50096   ENSMUST00000185968          -              -  ENSMUSG00000030228   \n",
       "50097   ENSMUST00000032353          -              -  ENSMUSG00000030228   \n",
       "100536  ENSMUST00000223134          -              -  ENSMUSG00000113453   \n",
       "50099   ENSMUST00000203885          -              -  ENSMUSG00000107487   \n",
       "100535  ENSMUST00000221519          -              -  ENSMUSG00000113453   \n",
       "50102   ENSMUST00000032356          -              -  ENSMUSG00000030230   \n",
       "50103   ENSMUST00000137148          -              -  ENSMUSG00000030230   \n",
       "50104   ENSMUST00000129986          -              -  ENSMUSG00000030230   \n",
       "50105   ENSMUST00000149931          -              -  ENSMUSG00000030230   \n",
       "50106   ENSMUST00000204342          -              -  ENSMUSG00000041791   \n",
       "50107   ENSMUST00000043797          -              -  ENSMUSG00000041791   \n",
       "50108   ENSMUST00000204787          -              -  ENSMUSG00000108251   \n",
       "100534  ENSMUST00000221930          -              -  ENSMUSG00000113700   \n",
       "129294  ENSMUST00000124446          -              -  ENSMUSG00000031388   \n",
       "50082   ENSMUST00000204038          -              -  ENSMUSG00000108128   \n",
       "50081   ENSMUST00000158278          -              -  ENSMUSG00000088903   \n",
       "...                    ...        ...            ...                 ...   \n",
       "134495  ENSMUST00000082401          -              -  ENSMUSG00000064350   \n",
       "107869  ENSMUST00000226722          -              -  ENSMUSG00000022108   \n",
       "132204  ENSMUST00000112172          -              -  ENSMUSG00000049775   \n",
       "699     ENSMUST00000189941          -              -  ENSMUSG00000100862   \n",
       "93701   ENSMUST00000197752          -              -  ENSMUSG00000106344   \n",
       "108445  ENSMUST00000134040          -              -  ENSMUSG00000060126   \n",
       "82989   ENSMUST00000092162          -              -  ENSMUSG00000069515   \n",
       "83425   ENSMUST00000177557          -              -  ENSMUSG00000093816   \n",
       "117485  ENSMUST00000102084          -              -  ENSMUSG00000076038   \n",
       "31107   ENSMUST00000102081          -              -  ENSMUSG00000094403   \n",
       "55427   ENSMUST00000174355          -              -  ENSMUSG00000002985   \n",
       "4744    ENSMUST00000188837          -              -  ENSMUSG00000100405   \n",
       "55432   ENSMUST00000172808          -              -  ENSMUSG00000002985   \n",
       "63644   ENSMUST00000066401          -              -  ENSMUSG00000007891   \n",
       "37766   ENSMUST00000197766          -              -  ENSMUSG00000105796   \n",
       "103349  ENSMUST00000184558          -              -  ENSMUSG00000098694   \n",
       "77035   ENSMUST00000121411          -              -  ENSMUSG00000081926   \n",
       "134501  ENSMUST00000082402          -              -  ENSMUSG00000064351   \n",
       "109958  ENSMUST00000159216          -              -  ENSMUSG00000064373   \n",
       "125362  ENSMUST00000025563          -              -  ENSMUSG00000024661   \n",
       "1509    ENSMUST00000187117          -              -  ENSMUSG00000101249   \n",
       "134494  ENSMUST00000082400          -              -  ENSMUSG00000064349   \n",
       "134511  ENSMUST00000082407          -              -  ENSMUSG00000064356   \n",
       "55425   ENSMUST00000167646          -              -  ENSMUSG00000002985   \n",
       "75775   ENSMUST00000184437          -              -  ENSMUSG00000098973   \n",
       "117924  ENSMUST00000102183          -              -  ENSMUSG00000076137   \n",
       "74819   ENSMUST00000102184          -              -  ENSMUSG00000076138   \n",
       "134493  ENSMUST00000082399          -              -  ENSMUSG00000064348   \n",
       "134518  ENSMUST00000082423          -              -  ENSMUSG00000064372   \n",
       "134492  ENSMUST00000082398          -              -  ENSMUSG00000064347   \n",
       "\n",
       "       gene_short_name tss_id                      locus  length   coverage  \\\n",
       "121192         St6gal2      -    chr17:55445388-55499226    2606       0.00   \n",
       "50083     RP24-446A8.1      -   chr6:139575923-139577205    1282       0.00   \n",
       "50084            Rergl      -   chr6:139492972-139501976     958       0.00   \n",
       "50085     RP24-446A8.2      -   chr6:139555558-139557611     498       0.00   \n",
       "50086         n-R5s168      -   chr6:140319629-140319733     104       0.00   \n",
       "50087    RP23-330J17.2      -   chr6:140309070-140309740     670       0.00   \n",
       "50088     RP23-78J15.7      -   chr6:140069993-140070498     505       0.00   \n",
       "50089     RP23-78J15.4      -   chr6:140132066-140132508     442       0.00   \n",
       "50090    RP23-330J17.1      -   chr6:140295462-140296456     994       0.00   \n",
       "50091          Gm24174      -   chr6:140166534-140166645     111       0.00   \n",
       "50092          Pik3c2g      -   chr6:139587220-139622487     721       0.00   \n",
       "50093          Pik3c2g      -   chr6:139587236-139635967    1774       0.00   \n",
       "50094          Pik3c2g      -   chr6:139599479-139636074    2023       0.00   \n",
       "50095          Pik3c2g      -   chr6:139599482-139657231    1358       0.00   \n",
       "50096          Pik3c2g      -   chr6:139614070-139657231    1627       0.00   \n",
       "50097          Pik3c2g      -   chr6:139621887-139635740    1206       0.00   \n",
       "100536     RP24-70J8.4      -    chr13:63621499-63627350     751       0.00   \n",
       "50099     RP23-78J15.6      -   chr6:140056400-140057589    1189       0.00   \n",
       "100535     RP24-70J8.4      -    chr13:63621499-63626881    1003       0.00   \n",
       "50102            Plcz1      -   chr6:139989672-140041457    2271       0.00   \n",
       "50103            Plcz1      -   chr6:139989721-140041431    2226       0.00   \n",
       "50104            Plcz1      -   chr6:139989723-140041412    2178       0.00   \n",
       "50105            Plcz1      -   chr6:140020250-140041412     989       0.00   \n",
       "50106           Capza3      -   chr6:140041251-140042158     771       0.00   \n",
       "50107           Capza3      -   chr6:140041556-140042778    1222       0.00   \n",
       "50108     RP23-78J15.3      -   chr6:140033160-140033532     372       0.00   \n",
       "100534     RP24-70J8.6      -    chr13:63661796-63662161     365       0.00   \n",
       "129294           Naa10      -     chrX:73917860-73920465     963       0.00   \n",
       "50082    RP23-238K12.1      -   chr6:139315596-139322576    1149       0.00   \n",
       "50081          Gm25434      -   chr6:139196064-139196231     167       0.00   \n",
       "...                ...    ...                        ...     ...        ...   \n",
       "134495           mt-Ty      -             chrM:5259-5326      67    5291.61   \n",
       "107869           Itm2b      -    chr14:73362884-73368505     935    5341.89   \n",
       "132204          Tmsb4x      -   chrX:167207092-167209315     768    5388.94   \n",
       "699        RP23-8J15.3      -     chr1:24613973-24614651     678    5398.74   \n",
       "93701          Mir1892      -    chr12:54645932-54646012      80    5583.25   \n",
       "108445            Tpt1      -    chr14:75845899-75848297     661    5656.84   \n",
       "82989             Lyz1      -  chr10:117287796-117292868    1263    5792.09   \n",
       "83425          Gm26445      -  chr10:130592135-130592244     109    5895.96   \n",
       "117485         Gm22774      -      chr17:6131835-6131944     109    6178.38   \n",
       "31107         Mir692-2      -   chr4:125504748-125504857     109    6289.17   \n",
       "55427             Apoe      -     chr7:19696271-19699008    1104    6441.73   \n",
       "4744     RP24-282H12.1      -   chr1:123205148-123205265     117    6870.43   \n",
       "55432             Apoe      -     chr7:19696843-19698247     456    7027.86   \n",
       "63644             Ctsd      -   chr7:142375916-142387857    1922    7581.21   \n",
       "37766     RP23-440C1.3      -     chr5:87737105-87737210     105    8603.67   \n",
       "103349         Gm27506      -    chr14:13778071-13778151      80    8900.06   \n",
       "77035     RP24-217J7.2      -      chr10:7038660-7038774     114    9354.27   \n",
       "134501          mt-Co1      -             chrM:5327-6872    1545    9711.81   \n",
       "109958           Sepp1      -      chr15:3270920-3280507    2085   10492.80   \n",
       "125362            Fth1      -      chr19:9982702-9985092     866   11416.40   \n",
       "1509       RP23-8J15.1      -     chr1:24615705-24616197     492   13177.20   \n",
       "134494           mt-Tc      -             chrM:5191-5257      66   13411.80   \n",
       "134511         mt-Atp8      -             chrM:7765-7969     204   16887.60   \n",
       "55425             Apoe      -     chr7:19696266-19699052     728   18751.70   \n",
       "75775          Mir6236      -   chr9:110281286-110281409     123   19556.50   \n",
       "117924         Gm26384      -    chr17:21284102-21284212     110   20635.90   \n",
       "74819          Gm26377      -     chr9:78478527-78478635     108   23976.00   \n",
       "134493           mt-Tn      -             chrM:5088-5159      71   26765.20   \n",
       "134518           mt-Tp      -           chrM:15355-15422      67  100835.00   \n",
       "134492           mt-Ta      -             chrM:5017-5086      69  122539.00   \n",
       "\n",
       "            FPKM  FPKM_conf_lo  FPKM_conf_hi FPKM_status  \n",
       "121192      0.00         0.000      0.000000          OK  \n",
       "50083       0.00         0.000      0.000000          OK  \n",
       "50084       0.00         0.000      0.000000          OK  \n",
       "50085       0.00         0.000      0.000000          OK  \n",
       "50086       0.00         0.000      0.000000          OK  \n",
       "50087       0.00         0.000      0.000000          OK  \n",
       "50088       0.00         0.000      0.000000          OK  \n",
       "50089       0.00         0.000      0.000000          OK  \n",
       "50090       0.00         0.000      0.000000          OK  \n",
       "50091       0.00         0.000      0.000000          OK  \n",
       "50092       0.00         0.000      0.000000          OK  \n",
       "50093       0.00         0.000      0.000000          OK  \n",
       "50094       0.00         0.000      0.000000          OK  \n",
       "50095       0.00         0.000      0.000000          OK  \n",
       "50096       0.00         0.000      0.000000          OK  \n",
       "50097       0.00         0.000      0.000000          OK  \n",
       "100536      0.00         0.000      0.000000          OK  \n",
       "50099       0.00         0.000      0.027574          OK  \n",
       "100535      0.00         0.000      0.000000          OK  \n",
       "50102       0.00         0.000      0.000000          OK  \n",
       "50103       0.00         0.000      0.000000          OK  \n",
       "50104       0.00         0.000      0.000000          OK  \n",
       "50105       0.00         0.000      0.000000          OK  \n",
       "50106       0.00         0.000      0.000000          OK  \n",
       "50107       0.00         0.000      0.000000          OK  \n",
       "50108       0.00         0.000      0.000000          OK  \n",
       "100534      0.00         0.000      0.000000          OK  \n",
       "129294      0.00         0.000      0.043211          OK  \n",
       "50082       0.00         0.000      0.000000          OK  \n",
       "50081       0.00         0.000      0.000000          OK  \n",
       "...          ...           ...           ...         ...  \n",
       "134495   1700.89       898.553   2503.230000          OK  \n",
       "107869   1717.05      1644.980   1789.130000          OK  \n",
       "132204   1732.18      1659.310   1805.040000          OK  \n",
       "699      1735.33      1719.700   1750.950000          OK  \n",
       "93701    1794.63      1188.700   2400.570000          OK  \n",
       "108445   1818.29      1741.440   1895.140000          OK  \n",
       "82989    1861.76      1850.920   1872.600000          OK  \n",
       "83425    1895.15      1658.500   2131.790000          OK  \n",
       "117485   1985.93      1743.130   2228.720000          OK  \n",
       "31107    2021.54      1776.400   2266.680000          OK  \n",
       "55427    2070.58      1984.170   2156.990000          OK  \n",
       "4744     2208.37      2003.020   2413.730000          OK  \n",
       "55432    2258.98      2159.930   2358.020000          OK  \n",
       "63644    2436.84      2335.840   2537.840000          OK  \n",
       "37766    2765.49      2474.370   3056.610000          OK  \n",
       "103349   2860.76      2216.050   3505.470000          OK  \n",
       "77035    3006.76      2751.690   3261.830000          OK  \n",
       "134501   3121.68      2992.370   3251.000000          OK  \n",
       "109958   3372.73      3233.180   3512.280000          OK  \n",
       "125362   3669.59      3650.300   3688.880000          OK  \n",
       "1509     4235.58      4204.270   4266.890000          OK  \n",
       "134494   4310.98      2938.610   5683.340000          OK  \n",
       "134511   5428.22      5178.680   5677.760000          OK  \n",
       "55425    6027.37      5776.900   6277.850000          OK  \n",
       "75775    6286.09      5975.420   6596.760000          OK  \n",
       "117924   6633.03      6227.550   7038.510000          OK  \n",
       "74819    7706.64      7050.590   8362.700000          OK  \n",
       "134493   8603.20      7143.770  10062.600000          OK  \n",
       "134518  32411.40     28889.200  35933.600000          OK  \n",
       "134492  39388.00     35912.700  42863.300000          OK  \n",
       "\n",
       "[134520 rows x 13 columns]"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "file = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/MERFISH_analysis-master/mouse/Zach/Cornea/Expression/no_versions_isoforms_tracking.csv'\n",
    "transcriptome_df = pd.read_csv(open(file, 'r'))\n",
    "transcriptome_df.sort_values('FPKM', ascending=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '1TmiXNh_yku0GHz6Jw76f0VPZH_h87kqJrrUUtJPfahU',\n",
       " 'updatedRange': 'Sheet1!C2:C189',\n",
       " 'updatedRows': 188,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedCells': 188}"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add FPKM data\n",
    "transcript_IDs = expression_df['Transcript_ID']\n",
    "file = '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/MERFISH_analysis-master/mouse/Zach/Cornea/Expression/no_versions_isoforms_tracking.csv'\n",
    "Transcriptome_annotations = Trascriptome_append(transcript_IDs,file)\n",
    "FPKM = expression_worksheet.range('C2:C189')\n",
    "for idx, g in enumerate(Transcriptome_annotations):\n",
    "    info = g[1]\n",
    "    if info is None:\n",
    "        continue\n",
    "    fpkm = info['FPKM'].iloc[0]\n",
    "    FPKM[idx].value = fpkm\n",
    "expression_worksheet.update_cells(FPKM)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gene too short:  ENSMUSG00000046259\n",
      "Gene too short:  ENSMUSG00000048455\n",
      "Gene too short:  ENSMUSG00000056054\n",
      "Gene too short:  ENSMUSG00000044303\n",
      "Gene too short:  ENSMUSG00000049775\n",
      "Gene too short:  ENSMUSG00000001131\n",
      "Gene too short:  ENSMUSG00000056071\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '11MMvEY4s8oUIGf5kyoA4QOKs-Br2myyNooNmSO7-nEU',\n",
       " 'updatedCells': 188,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedRange': 'Sheet1!A2:A189',\n",
       " 'updatedRows': 188}"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Add info to spreadsheet\n",
    "# WARNING - you must figure out column numbers and change bits for new gene sets\n",
    "gene_annotations = ensymbl_to_geneinfo(expression_worksheet.range('F2:F205'), '/home/zach/Documents/Untitled Folder/mouse_gene_info2.txt', organism='mouse')\n",
    "gname = expression_worksheet.range('F2:F189')\n",
    "tname = expression_worksheet.range('G2:G189')\n",
    "length = expression_worksheet.range('D2:D189')\n",
    "descript = expression_worksheet.range('B2:B189')\n",
    "gene = expression_worksheet.range('A2:A189')\n",
    "update = []\n",
    "for idx, g in enumerate(gene_annotations):\n",
    "    info = g[1]\n",
    "    if info is None:\n",
    "        continue\n",
    "    max_transcript = info['Transcript stable ID'].iloc[0]\n",
    "    max_length = info['Transcript length (including UTRs and CDS)'].iloc[0]\n",
    "    descript_val = info['Gene description'].iloc[0]\n",
    "    gene_name = info['Gene name'].iloc[0]\n",
    "    tname[idx].value = max_transcript\n",
    "    length[idx].value = str(max_length)\n",
    "    descript[idx].value = str(descript_val)\n",
    "    gene[idx].value = str(gene_name)\n",
    "expression_worksheet.update_cells(tname)\n",
    "expression_worksheet.update_cells(length)\n",
    "expression_worksheet.update_cells(descript)\n",
    "expression_worksheet.update_cells(gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [],
   "source": [
    "readout_names = ['RS0095', 'RS0109', 'RS0175', 'RS0237', 'RS0307', 'RS0332', 'RS0384', 'RS0406', \n",
    "                'RS0451', 'RS0468', 'RS0548', 'RS64.0', 'RS156.0', 'RS278.0', 'RS313.0', 'RS643.0', \n",
    "                'RS740.0', 'RS810.0']\n",
    "def write_codebook(rows, fname, readout_names, codebook_style = '148MHD4'):\n",
    "    with open(fname, 'w') as f:\n",
    "        f.write('version'+','+str(1)+'\\n')\n",
    "        f.write('codebook_name'+','+codebook_style+'\\n')\n",
    "        f.write('bit_names,'+','.join(readout_names)+'\\n')\n",
    "        f.write('name, id, barcode\\n')\n",
    "        for row in rows:\n",
    "            f.write(','.join([row[0], row[1], row[2]+'\\n']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('Pgk1', 'ENSMUST00000081593', '1000110010000000'),\n",
       " ('Sdha', 'ENSMUST00000022062', '1000110010000000'),\n",
       " ('Tbp', 'ENSMUST00000118001', '1000110010000000'),\n",
       " ('H2afz', 'ENSMUST00000174561', '1000110010000000'),\n",
       " ('Gapdh', 'ENSMUST00000147954', '1000110010000000'),\n",
       " ('Ppia', 'ENSMUST00000090749', '1000110010000000'),\n",
       " ('Gusb', 'ENSMUST00000026613', '1000110010000000'),\n",
       " ('Hspa1b', 'ENSMUST00000172753', '1000110010000000'),\n",
       " ('Ywhaz', 'ENSMUST00000022894', '1000110010000000'),\n",
       " ('Actb', 'ENSMUST00000100497', '1000110010000000'),\n",
       " ('B2m', 'ENSMUST00000102476', '1000110010000000'),\n",
       " ('Hk2', 'ENSMUST00000168725', '1000110010000000'),\n",
       " ('Hprt', 'ENSMUST00000026723', '1000110010000000'),\n",
       " ('Arf1', 'ENSMUST00000061242', '1000110010000000'),\n",
       " ('Cox7a2l', 'ENSMUST00000025095', '1000110010000000'),\n",
       " ('D8Ertd738e', 'ENSMUST00000019506', '1000110010000000'),\n",
       " ('Eif1', 'ENSMUST00000049385', '1000110010000000'),\n",
       " ('Eif4g2', 'ENSMUST00000161051', '1000110010000000'),\n",
       " ('Eif5a', 'ENSMUST00000108608', '1000110010000000'),\n",
       " ('Gabarap', 'ENSMUST00000108592', '1000110010000000'),\n",
       " ('Grcc10', 'ENSMUST00000004389', '1000110010000000'),\n",
       " ('Cops9', 'ENSMUST00000148508', '1000110010000000'),\n",
       " ('Ndufa2', 'ENSMUST00000014438', '1000110010000000'),\n",
       " ('Nedd8', 'ENSMUST00000163750', '1000110010000000'),\n",
       " ('Pomp', 'ENSMUST00000031654', '1000110010000000'),\n",
       " ('Psma2', 'ENSMUST00000221168', '1000110010000000'),\n",
       " ('Psma3', 'ENSMUST00000071704', '1000110010000000'),\n",
       " ('Psma4', 'ENSMUST00000034848', '1000110010000000'),\n",
       " ('Psmb1', 'ENSMUST00000014913', '1000110010000000'),\n",
       " ('Psmb2', 'ENSMUST00000030642', '1000110010000000'),\n",
       " ('Psmb3', 'ENSMUST00000103147', '1000110010000000'),\n",
       " ('Psmb4', 'ENSMUST00000133412', '1000110010000000'),\n",
       " ('Psmb6', 'ENSMUST00000018430', '1000110010000000'),\n",
       " ('Psmb7', 'ENSMUST00000028083', '1000110010000000'),\n",
       " ('Rab1b', 'ENSMUST00000025804', '1000110010000000'),\n",
       " ('Saraf', 'ENSMUST00000033933', '1000110010000000'),\n",
       " ('Ubl5', 'ENSMUST00000162303', '1000110010000000'),\n",
       " ('Vcp', 'ENSMUST00000030164', '1000110010000000')]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Generate Codebook\n",
    "row_tuples = []\n",
    "\n",
    "for idx, row in expression_df.drop_duplicates('Transcript_ID').iterrows():\n",
    "    row_tuples.append((row['Gene'], row['Transcript_ID'].split('.')[0], str(row['Barcode'])))\n",
    "\n",
    "write_codebook(row_tuples, '/bigstore/GeneralStorage/Zach/MERFISH/Probe_Design/results/Housekeeping/Housekeeping_Codebook.txt', readout_names)\n",
    "row_tuples"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now design probes\n",
    "parse_merfish_oligos to add results to spreadsheet\n",
    "Filter genes with low probes and repeat\n",
    "Decide on final gene list and run agian\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_merfish_oligos(fname, counts_df = None,\n",
    "                         counts_df_column='FPKM', tid_column='transcript_id'):\n",
    "    \"\"\"\n",
    "    Bit hacky - should refactor and figure out how to handle missing isoform specificity info.\n",
    "    \"\"\"\n",
    "    from Bio import SeqIO\n",
    "    import pandas as pd\n",
    "    df = pd.DataFrame(columns=['experiemnt', 'pleft', 'ro1', 'ro2', 'ro3', \n",
    "                               'pright', 'seq', 'gene', 'tid', \n",
    "                              'start', 'length', 'gc', 'tm', 'specicity'])\n",
    "    readout_dict = {}\n",
    "    oligos = SeqIO.FastaIO.SimpleFastaParser(open(fname, 'r'))\n",
    "    rows = []\n",
    "    fpkms = []\n",
    "    for header, seq in oligos:\n",
    "        fields = header.split(' ')\n",
    "        experiment = str(fields[0])\n",
    "        primer_left = str(fields[1])\n",
    "        primer_seqL = seq[:20]\n",
    "        primer_seqR = seq[-20:]\n",
    "        readout1 = str(fields[2])\n",
    "        readout_dict[readout1] = seq[20+1:20+1+20]\n",
    "        isoSpecificity = 1\n",
    "        # Order is different if Readouts are RO1/RO2 - encoding - RO3 vs RO1 - encoding - RO2/RO3\n",
    "        # Check with if statement and handle accordingly.\n",
    "        if '__' not in fields[3]:\n",
    "            readout2 = str(fields[3])\n",
    "            encoding = str(fields[4])\n",
    "            readout3 = fields[5]\n",
    "            ro2_start_idx = 41\n",
    "            ro3_start_idx = 92\n",
    "            readout_dict[readout3] = seq[20+20+20+1+30+1:20+1+20+20+30+1+20]\n",
    "            readout_dict[readout2] = seq[20+1+20:20+1+20+20]\n",
    "            primer_right = fields[6]\n",
    "            gene, tid, start, length, gc, tm, specificity  = encoding.split('__')\n",
    "\n",
    "\n",
    "#             gene, tid, start, length, gc, tm, specificity, isoSpecificity  = encoding.split('__')\n",
    "            encoding_region = seq[20+1+20+20:20+20+1+20+30]\n",
    "        else:\n",
    "            encoding = fields[3]\n",
    "#             isoSpecificity = fields[4]\n",
    "            readout2 = fields[4]\n",
    "            readout3 = fields[5]\n",
    "            ro2_start_idx = 72\n",
    "            ro3_start_idx = 92\n",
    "            readout_dict[readout2] = seq[20+2+20+30:20+20+30+20+2]\n",
    "            readout_dict[readout3] = seq[20+2+20+30+20:20+20+30+20+20+2]\n",
    "            primer_right = fields[6]\n",
    "            gene, tid, start, length, gc, tm, specificity  = encoding.split('__')\n",
    "#             gene, tid, start, length, gc, tm, specificity, isoSpecificity  = encoding.split('__')\n",
    "        # IMPLEMENT READOUT DICT In If Else\n",
    "            encoding_region = seq[20+1+20:20+1+20+30]\n",
    "        \n",
    "        rows.append([experiment, primer_seqL, readout1, readout2, readout3, \n",
    "                       primer_seqR, encoding_region, seq, gene, tid, start,\n",
    "                     length, gc, tm, specificity, isoSpecificity, header])\n",
    "    df = pd.DataFrame(rows, columns=['experiment', 'pleft', 'ro1', 'ro2', 'ro3', \n",
    "                               'pright', 'encodingRegion', 'seq', 'gene', 'tid', \n",
    "                              'start', 'length', 'gc', 'tm', 'specificity', 'isoSpecificity', 'header'])\n",
    "    df = df.drop_duplicates(subset=['gene', 'encodingRegion'])\n",
    "#     if isinstance(counts_df, pd.DataFrame):\n",
    "#         for tid in df.tid.unique():\n",
    "#             fpkm = counts_df[counts_df[tid_column]==tid][counts_df_column]\n",
    "#             tid_idx = df[df.tid==tid].index\n",
    "#             for i in tid_idx:\n",
    "#                 df.set_value(i, counts_df_column, fpkm.values[0])\n",
    "#     df = df.convert_objects(convert_numeric=True)\n",
    "#     df.sort_values(['gene', 'specificity', 'isoSpecificity'], ascending=False, inplace=True)\n",
    "#     df['iso_off_spots'] = (df[counts_df_column] - df['isoSpecificity']*df[counts_df_column])/df['isoSpecificity']\n",
    "#     df['gene_off_spots'] = (df[counts_df_column] - df['specificity']*df[counts_df_column])/df['specificity']\n",
    "#     df = df.drop_duplicates('tid')\n",
    "    return df, primer_seqL, primer_seqR, readout_dict\n",
    "\n",
    "def trim_oligos_to_fit(oligo_df, multi_transcripts_cutoff = 148, min_oligos=48):\n",
    "    df2 = oligo_df.copy()\n",
    "    c = Counter(df2.gene)\n",
    "    high_count = {}\n",
    "    for g, count in c.items():\n",
    "        if count < min_oligos:\n",
    "            print(g, count)\n",
    "#             c.pop(g)\n",
    "            if g not in ['SNAI2', 'SNAI1', 'ORAI1', 'P2RY11', 'INPP1', 'ACTA2', 'PICK1']:\n",
    "                df2.drop(df2[df2.gene==g].index, inplace=True)\n",
    "#         if count>multi_transcripts_cutoff:\n",
    "#             high_count[g] = count\n",
    "#             ixes = list(df2[df2.gene==g].index)\n",
    "#             ixes = np.random.choice(ixes, size=multi_transcripts_cutoff, replace=False)\n",
    "#             df2.drop(ixes, inplace=True)\n",
    "    return df2\n",
    "\n",
    "# def balance_readouts(df, per_tid=64, fa_out='mergos.fa'):\n",
    "#     from itertools import repeat\n",
    "#     tids = df.groupby(group)\n",
    "#     f = open(fa_out, 'w')\n",
    "#     new_df = pd.DataFrame()\n",
    "#     counters = []\n",
    "#     for name, group in tids:\n",
    "#         r_used = pd.unique(np.concatenate((group.ro1.unique(),group.ro2.unique(),group.ro3.unique())))\n",
    "        \n",
    "def balance_readouts(df, primersL, primersR, readouts, per_tid=64, group='tid',\n",
    "                     fa_out='python_mergos.fa', sep='__'):\n",
    "    verbose=False\n",
    "    from itertools import repeat\n",
    "    tids = df.groupby(group)\n",
    "    f = open(fa_out, 'w')\n",
    "    new_df = pd.DataFrame()\n",
    "    counters = {}\n",
    "    for name, group in tids:\n",
    "        counts = Counter()\n",
    "        r_used = pd.unique(np.concatenate((group.ro1.unique(),group.ro2.unique(),group.ro3.unique())))\n",
    "        r_used = np.concatenate(list(repeat(r_used, 1000)))\n",
    "        oligo_index = group.index.tolist()\n",
    "        np.random.shuffle(oligo_index)\n",
    "        oligo_index = oligo_index[:per_tid]\n",
    "        base_idx = 0\n",
    "        c = Counter()\n",
    "        for i, idx in enumerate(oligo_index):\n",
    "            ro1_seq = ''\n",
    "            ro2_seq = ''\n",
    "            ro3_seq = ''\n",
    "            oligo = ''\n",
    "            ro1=''\n",
    "            ro2=''\n",
    "            ro3=''\n",
    "#             try:\n",
    "            ro1_seq = readouts[r_used[base_idx]]\n",
    "            ro1 = r_used[base_idx]\n",
    "        \n",
    "            ro2_seq = readouts[r_used[base_idx+1]]\n",
    "            ro2 = r_used[base_idx+1]\n",
    "            \n",
    "            ro3_seq = readouts[r_used[base_idx+2]]\n",
    "            ro3 = r_used[base_idx+2]\n",
    "            \n",
    "            c.update([ro1, ro2, ro3])\n",
    "            row = group.loc[idx]\n",
    "            rand = np.random.randint(0, high=2)\n",
    "            if (ro1 not in r_used) or (ro2 not in r_used) or (ro3 not in r_used):\n",
    "                print(row)\n",
    "            if rand:\n",
    "                oligo = row.pleft+'A'+ro1_seq+ro2_seq+row.encodingRegion+'A'+ro3_seq+row.pright\n",
    "\n",
    "#                     row.set_value(idx, 'oligo', row.pleft+row.ro1+row.ro2+'A'+row.encodingRegion+'A'+row.ro3+row.pright)\n",
    "            else:\n",
    "                oligo = row.pleft+'A'+ro1_seq+row.encodingRegion+'A'+ro2_seq+ro3_seq+row.pright\n",
    "            if (len(row.encodingRegion) != 30) or (len(ro1_seq) != 20):\n",
    "\n",
    "                print(len(row.encodingRegion), len(ro1_seq))\n",
    "#                     row.set_value(idx, 'oligo', row.pleft+row.ro1+'A'+row.encodingRegion+'A'+row.ro2+row.ro3+row.pright)\n",
    "            header = \">\"+row.gene+sep+row.tid+sep+str(row.start)+sep+ro1+sep+ro2+sep+ro3+sep+row.experiment+'\\n'\n",
    "            f.write(header)\n",
    "            f.write(oligo+'\\n')\n",
    "            base_idx += 3\n",
    "#             except Exception as e:\n",
    "#                 print(e)\n",
    "#                 continue\n",
    "        counters[name] = c\n",
    "        if len(c.keys())>4:\n",
    "            print(name)\n",
    "    f.close()\n",
    "    return new_df, fa_out, counters"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'spreadsheetId': '11MMvEY4s8oUIGf5kyoA4QOKs-Br2myyNooNmSO7-nEU',\n",
       " 'updatedCells': 188,\n",
       " 'updatedColumns': 1,\n",
       " 'updatedRange': 'Sheet1!E2:E189',\n",
       " 'updatedRows': 188}"
      ]
     },
     "execution_count": 133,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# update spreadsheet with Probe design data\n",
    "# Sequences, FPKM, number of probes\n",
    "df = parse_merfish_oligos('/bigstore/GeneralStorage/Rob/merfish/MERFISH_analysis-master/mouse/Cornea_Wound/Cornea_Wound_oligos.fasta')[0]\n",
    "from collections import Counter\n",
    "counts = Counter(df.gene)\n",
    "num_oligos = expression_worksheet.range('E2:E189')\n",
    "gnames = expression_worksheet.range('A2:A189')\n",
    "for k, v in counts.items():\n",
    "    idx = [i for i, g in enumerate(gnames) if g.value==k][0]\n",
    "    num_oligos[idx].value = v\n",
    "expression_worksheet.update_cells(num_oligos)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_list = []\n",
    "for i in range(len(gene_annotations)):\n",
    "    gene_list.append(gene_annotations[i][0])"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pyspots",
   "language": "python",
   "name": "pyspots"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
